<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Root Mean Squared Fluctuation (RMSF) · MIToS</title><meta name="title" content="Root Mean Squared Fluctuation (RMSF) · MIToS"/><meta property="og:title" content="Root Mean Squared Fluctuation (RMSF) · MIToS"/><meta property="twitter:title" content="Root Mean Squared Fluctuation (RMSF) · MIToS"/><meta name="description" content="Documentation for MIToS."/><meta property="og:description" content="Documentation for MIToS."/><meta property="twitter:description" content="Documentation for MIToS."/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">MIToS</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../Installation/">Installation</a></li><li><a class="tocitem" href="../Example/">Example</a></li><li><span class="tocitem">Modules</span><ul><li><a class="tocitem" href="../MSA/">MSA</a></li><li><a class="tocitem" href="../Information/">Information</a></li><li><a class="tocitem" href="../SIFTS/">SIFTS</a></li><li><a class="tocitem" href="../PDB/">PDB</a></li><li><a class="tocitem" href="../Pfam/">Pfam</a></li></ul></li><li><span class="tocitem">Cookbook</span><ul><li><a class="tocitem" href="../01_Change_B_factors/">Change B-factors</a></li><li><a class="tocitem" href="../02_Linking_structural_and_evolutionary_information/">Linking structural and evolutionary information</a></li><li class="is-active"><a class="tocitem" href>Root Mean Squared Fluctuation (RMSF)</a><ul class="internal"><li><a class="tocitem" href="#Problem-description"><span>Problem description</span></a></li><li><a class="tocitem" href="#MIToS-solution"><span>MIToS solution</span></a></li></ul></li></ul></li><li><span class="tocitem">API</span><ul><li><a class="tocitem" href="../MSA_API/">MSA</a></li><li><a class="tocitem" href="../Information_API/">Information</a></li><li><a class="tocitem" href="../SIFTS_API/">SIFTS</a></li><li><a class="tocitem" href="../PDB_API/">PDB</a></li><li><a class="tocitem" href="../Pfam_API/">Pfam</a></li><li><a class="tocitem" href="../Utils_API/">Utils</a></li></ul></li><li><a class="tocitem" href="../Scripts/">Scripts</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Cookbook</a></li><li class="is-active"><a href>Root Mean Squared Fluctuation (RMSF)</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Root Mean Squared Fluctuation (RMSF)</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/diegozea/MIToS.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/diegozea/MIToS.jl/blob/master/docs/src/cookbook/03_RMSF.jl" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Root-Mean-Squared-Fluctuation-(RMSF)"><a class="docs-heading-anchor" href="#Root-Mean-Squared-Fluctuation-(RMSF)">Root Mean Squared Fluctuation (RMSF)</a><a id="Root-Mean-Squared-Fluctuation-(RMSF)-1"></a><a class="docs-heading-anchor-permalink" href="#Root-Mean-Squared-Fluctuation-(RMSF)" title="Permalink"></a></h1><p>md # <a href="https://mybinder.org/v2/gh/diegozea/MIToS.jl/gh-pages?filepath=devcookbook/notebooks/03_RMSF.ipynb"><img src="https://mybinder.org/badge_logo.svg" alt/></a> md # <a href="https://nbviewer.jupyter.org/github/diegozea/MIToS.jl/blob/gh-pages/devcookbook/notebooks/03_RMSF.ipynb"><img src="https://img.shields.io/badge/show-nbviewer-579ACA.svg" alt/></a></p><h2 id="Problem-description"><a class="docs-heading-anchor" href="#Problem-description">Problem description</a><a id="Problem-description-1"></a><a class="docs-heading-anchor-permalink" href="#Problem-description" title="Permalink"></a></h2><p>The <a href="https://en.wikipedia.org/wiki/Mean_squared_displacement">Root Mean Squared Fluctuation (RMSF)</a> is a common way to measure residue flexibility in a structural ensemble. It is a measure of how far is the residue moving from its average position in the group of structures. Usually, we represent a residue position with the spatial coordinates of its alpha carbon.</p><p>The protein structures should be previously superimposed to calculate the RMSF, for example, by using the <code>superimpose</code> function of the <a href="../PDB/#Module-PDB"><code>PDB</code> module of <code>MIToS</code></a>. In this example, we are going to measure the RMSF of each residue from an NMR ensemble using the <code>rmsf</code> function.</p><p>The structure superimposition could be the most complicated step of the process, depending on the input data. In particular, it structures come from different PDB structures or homologous proteins can require the use of external programs, as <a href="https://ub.cbm.uam.es/software/online/mamothmult.php">MAMMOTH-mult</a> or <a href="https://lcb.infotech.monash.edu/mustang/">MUSTANG</a> among others, tailored for this task.</p><p>In this case, we are going to use an NMR ensemble. Therefore, we are not going to need to superimpose the structures as NMR models have the same protein sequence and are, usually, well-aligned.</p><h2 id="MIToS-solution"><a class="docs-heading-anchor" href="#MIToS-solution">MIToS solution</a><a id="MIToS-solution-1"></a><a class="docs-heading-anchor-permalink" href="#MIToS-solution" title="Permalink"></a></h2><pre><code class="language-julia hljs">import MIToS
using MIToS.PDB
using Plots</code></pre><p>Lets read the NMR ensemble:</p><pre><code class="language-julia hljs">pdb_file = abspath(pathof(MIToS), &quot;..&quot;, &quot;..&quot;, &quot;test&quot;, &quot;data&quot;, &quot;1AS5.pdb&quot;)
pdb_res = read_file(pdb_file, PDBFile, occupancyfilter = true)</code></pre><p>We set <code>occupancyfilter</code> to <code>true</code> to ensure that we have one single set of coordinates for each atom. That filter isn&#39;t essential for NMR structures, but It can avoid multiple alpha carbons in crystallographic structures with disordered atoms. We can get an idea of the alpha carbon positions by plotting these residues:</p><pre><code class="language-julia hljs">scatter(pdb_res, legend = false)</code></pre><img src="474542ef.svg" alt="Example block output"/><p>As we saw in the previous plot, the structure doesn&#39;t need to be superimposed. Now, we are going to separate each model into different vectors, storing each vector into a <code>Dict</code>:</p><pre><code class="language-julia hljs">models = Dict{String,Vector{PDBResidue}}()
for res in pdb_res
    push!(get!(models, res.id.model, []), res)
end</code></pre><p>Then, we simply need to collect all the PDB models in the values of the <code>Dict</code>, to get the vector of <code>PDBResidues</code> vectors required to calculate the RMSF.</p><pre><code class="language-julia hljs">pdb_models = collect(values(models))</code></pre><p>And, finally, call the <code>rmsf</code> function on the list of structures. It is important that all the vectors has the same number of <code>PDBResidue</code>s. This function assumes that the nth element of each vector corresponds to the same residue:</p><pre><code class="language-julia hljs">RMSF = rmsf(pdb_models)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">21-element Vector{Float64}:
 0.6008049320091593
 0.6562466637427343
 0.5868752127208325
 0.7275663200011392
 1.6512606331440922
 1.3117607364118795
 0.6982103470842509
 0.7109607828100489
 0.6927992280655076
 0.8222912705808504
 ⋮
 0.5339744031929865
 0.5653746094224534
 0.7416600691443928
 0.7898229264221448
 0.8414943371097368
 0.8242922195831439
 1.004681790419235
 1.4029641961626411
 3.0733292325145656</code></pre><p>This return the vector of RMSF values for each residue, calculated using the coordinates of the alpha carbons. You can plot this vector to get an idea of the which are the most flexible position in your structure:</p><pre><code class="language-julia hljs">plot(RMSF, legend = false, xlab = &quot;Residue&quot;, ylab = &quot;RMSF [Å]&quot;)</code></pre><img src="fcc14d89.svg" alt="Example block output"/><hr/><p><em>This page was generated using <a href="https://github.com/fredrikekre/Literate.jl">Literate.jl</a>.</em></p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../02_Linking_structural_and_evolutionary_information/">« Linking structural and evolutionary information</a><a class="docs-footer-nextpage" href="../MSA_API/">MSA »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.5.0 on <span class="colophon-date" title="Friday 28 June 2024 10:52">Friday 28 June 2024</span>. Using Julia version 1.10.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
