<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>PDB · MIToS</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link href="assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><h1>MIToS</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="search.html"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="index.html">Home</a></li><li><a class="toctext" href="Installation.html">Installation</a></li><li><a class="toctext" href="Example.html">Example</a></li><li><span class="toctext">Modules</span><ul><li><a class="toctext" href="MSA.html">MSA</a></li><li><a class="toctext" href="Information.html">Information</a></li><li><a class="toctext" href="SIFTS.html">SIFTS</a></li><li><a class="toctext" href="PDB.html">PDB</a></li><li><a class="toctext" href="Pfam.html">Pfam</a></li></ul></li><li><a class="toctext" href="Scripts.html">Scripts</a></li><li><span class="toctext">API</span><ul><li><a class="toctext" href="MSA_API.html">MSA</a></li><li><a class="toctext" href="Information_API.html">Information</a></li><li><a class="toctext" href="SIFTS_API.html">SIFTS</a></li><li class="current"><a class="toctext" href="PDB_API.html">PDB</a><ul class="internal"><li><a class="toctext" href="#Contents-1">Contents</a></li><li><a class="toctext" href="#Types-1">Types</a></li><li><a class="toctext" href="#Constants-1">Constants</a></li><li><a class="toctext" href="#Macros-1">Macros</a></li><li><a class="toctext" href="#Methods-and-functions-1">Methods and functions</a></li></ul></li><li><a class="toctext" href="Pfam_API.html">Pfam</a></li><li><a class="toctext" href="Utils_API.html">Utils</a></li></ul></li></ul></nav><article id="docs"><header><nav><ul><li>API</li><li><a href="PDB_API.html">PDB</a></li></ul><a class="edit-page" href="https://github.com/diegozea/MIToS.jl/blob/master/docs/src/PDB_API.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>PDB</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="PDB-1" href="#PDB-1">PDB</a></h1><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB" href="#MIToS.PDB"><code>MIToS.PDB</code></a> — <span class="docstring-category">Module</span>.</div><div><p>The module <code>PDB</code> defines types and methods to work with protein structures inside Julia. It is useful to link structural and sequential information, and needed for measure the predictive performance at protein contact prediction of mutual information scores.</p><p><strong>Features</strong></p><ul><li><p>Read and parse PDF and PDBML files</p></li><li><p>Calculate distance and contacts between atoms or residues</p></li><li><p>Determine interaction between residues</p></li></ul><pre><code class="language-julia">using MIToS.PDB</code></pre></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDB.jl#L1-L15">source</a></section><h2><a class="nav-anchor" id="Contents-1" href="#Contents-1">Contents</a></h2><ul><li><a href="PDB_API.html#PDB-1">PDB</a></li><ul><li><a href="PDB_API.html#Contents-1">Contents</a></li><li><a href="PDB_API.html#Types-1">Types</a></li><li><a href="PDB_API.html#Constants-1">Constants</a></li><li><a href="PDB_API.html#Macros-1">Macros</a></li><li><a href="PDB_API.html#Methods-and-functions-1">Methods and functions</a></li></ul></ul><h2><a class="nav-anchor" id="Types-1" href="#Types-1">Types</a></h2><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.Coordinates" href="#MIToS.PDB.Coordinates"><code>MIToS.PDB.Coordinates</code></a> — <span class="docstring-category">Type</span>.</div><div><p>A <code>Coordinates</code> object is a fixed size vector with the coordinates x,y,z.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L24">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.PDBAtom" href="#MIToS.PDB.PDBAtom"><code>MIToS.PDB.PDBAtom</code></a> — <span class="docstring-category">Type</span>.</div><div><p>A <code>PDBAtom</code> object contains the information from a PDB atom, without information of the residue. It has the following fields that you can access at any moment for query purposes:</p><pre><code class="language-none">- `coordinates` : x,y,z coordinates, e.g. `Coordinates(109.641,73.162,42.7)`.
- `atom` : Atom name, e.g. `&quot;CA&quot;`.
- `element` : Element type of the atom, e.g. `&quot;C&quot;`.
- `occupancy` : A float number with the occupancy, e.g. `1.0`.
- `B` : B factor as a string, e.g. `&quot;23.60&quot;`.</code></pre></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L34-L43">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.PDBFile" href="#MIToS.PDB.PDBFile"><code>MIToS.PDB.PDBFile</code></a> — <span class="docstring-category">Type</span>.</div><div><p><code>PDBFile &lt;: Format</code></p><p>Protein Data Bank (PDB) format. It provides a standard representation for macromolecular structure data derived from X-ray diffraction and NMR studies.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBParser.jl#L1-L7">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.PDBML" href="#MIToS.PDB.PDBML"><code>MIToS.PDB.PDBML</code></a> — <span class="docstring-category">Type</span>.</div><div><p><code>PDBML &lt;: Format</code></p><p>Protein Data Bank Markup Language (PDBML), a representation of PDB data in XML format.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBMLParser.jl#L1-L5">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.PDBResidue" href="#MIToS.PDB.PDBResidue"><code>MIToS.PDB.PDBResidue</code></a> — <span class="docstring-category">Type</span>.</div><div><p>A <code>PDBResidue</code> object contains all the information about a PDB residue. It has the following fields that you can access at any moment for query purposes:</p><pre><code class="language-none">- `id` : A `PDBResidueIdentifier` object.
- `atoms` : A vector of `PDBAtom`s.</code></pre></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L52-L58">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.PDBResidueIdentifier" href="#MIToS.PDB.PDBResidueIdentifier"><code>MIToS.PDB.PDBResidueIdentifier</code></a> — <span class="docstring-category">Type</span>.</div><div><p>A <code>PDBResidueIdentifier</code> object contains the information needed to identity PDB residues. It has the following fields that you can access at any moment for query purposes:</p><pre><code class="language-none">- `PDBe_number` : It&#39;s only used when a PDBML is readed (PDBe number as a string).
- `number` : PDB residue number, it includes insertion codes, e.g. `&quot;34A&quot;`.
- `name` : Three letter residue name in PDB, e.g. `&quot;LYS&quot;`.
- `group` : It can be `&quot;ATOM&quot;` or `&quot;HETATM&quot;`.
- `model` : The model number as a string, e.g. `&quot;1&quot;`.
- `chain` : The chain as a string, e.g. `&quot;A&quot;`.</code></pre></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L4-L14">source</a></section><h2><a class="nav-anchor" id="Constants-1" href="#Constants-1">Constants</a></h2><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.covalentradius" href="#MIToS.PDB.covalentradius"><code>MIToS.PDB.covalentradius</code></a> — <span class="docstring-category">Constant</span>.</div><div><p>Covalent radius in Å of each element from the Additional file 1 of PICCOLO [1]. Hydrogen was updated using the value on Table 2 from Cordero et. al. [2].</p><ol><li><p>Bickerton, G. R., Higueruelo, A. P., &amp; Blundell, T. L. (2011).</p></li></ol><p>Comprehensive, atomic-level characterization of structurally characterized protein-protein interactions: the PICCOLO database. BMC bioinformatics, 12(1), 313.</p><ol><li><p>Cordero, B., Gómez, V., Platero-Prats, A. E., Revés, M.,</p></li></ol><p>Echeverría, J., Cremades, E., ... &amp; Alvarez, S. (2008). Covalent radii revisited. Dalton Transactions, (21), 2832-2838.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/AtomsData.jl#L1-L12">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.vanderwaalsradius" href="#MIToS.PDB.vanderwaalsradius"><code>MIToS.PDB.vanderwaalsradius</code></a> — <span class="docstring-category">Constant</span>.</div><div><p>van der Waals radius in Å from the Additional file 1 of Bickerton et. al. 2011</p><ul><li><p>Bickerton, G. R., Higueruelo, A. P., &amp; Blundell, T. L. (2011).</p></li></ul><p>Comprehensive, atomic-level characterization of structurally characterized protein-protein interactions: the PICCOLO database. BMC bioinformatics, 12(1), 313.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/AtomsData.jl#L21-L27">source</a></section><h2><a class="nav-anchor" id="Macros-1" href="#Macros-1">Macros</a></h2><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.@atoms-Tuple{Any,Symbol,Any,Symbol,Any,Symbol,Any,Symbol,Any,Symbol,Any}" href="#MIToS.PDB.@atoms-Tuple{Any,Symbol,Any,Symbol,Any,Symbol,Any,Symbol,Any,Symbol,Any}"><code>MIToS.PDB.@atoms</code></a> — <span class="docstring-category">Macro</span>.</div><div><p><code>@atoms ... model ... chain ... group ... residue ... atom ...</code></p><p>These return a vector of <code>PDBAtom</code>s with the selected subset of atoms. You can use the type <code>All</code> to avoid filtering that option.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L248-L253">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.@residues-Tuple{Any,Symbol,Any,Symbol,Any,Symbol,Any,Symbol,Any}" href="#MIToS.PDB.@residues-Tuple{Any,Symbol,Any,Symbol,Any,Symbol,Any,Symbol,Any}"><code>MIToS.PDB.@residues</code></a> — <span class="docstring-category">Macro</span>.</div><div><p><code>@residues ... model ... chain ... group ... residue ...</code></p><p>These return a new vector with the selected subset of residues from a list of residues. You can use the type <code>All</code> to avoid filtering that option.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L168-L173">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.@residuesdict-Tuple{Any,Symbol,Any,Symbol,Any,Symbol,Any,Symbol,Any}" href="#MIToS.PDB.@residuesdict-Tuple{Any,Symbol,Any,Symbol,Any,Symbol,Any,Symbol,Any}"><code>MIToS.PDB.@residuesdict</code></a> — <span class="docstring-category">Macro</span>.</div><div><p><code>@residuesdict ... model ... chain ... group ... residue ...</code></p><p>These return a dictionary (using PDB residue numbers as keys) with the selected subset of residues. You can use the type <code>All</code> to avoid filtering that option.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L205-L210">source</a></section><h2><a class="nav-anchor" id="Methods-and-functions-1" href="#Methods-and-functions-1">Methods and functions</a></h2><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Base.angle-Tuple{MIToS.PDB.Coordinates,MIToS.PDB.Coordinates,MIToS.PDB.Coordinates}" href="#Base.angle-Tuple{MIToS.PDB.Coordinates,MIToS.PDB.Coordinates,MIToS.PDB.Coordinates}"><code>Base.angle</code></a> — <span class="docstring-category">Method</span>.</div><div><p><code>angle(a::Coordinates, b::Coordinates, c::Coordinates)</code></p><p>Angle (in degrees) at <code>b</code> between <code>a-b</code> and <code>b-c</code></p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L108-L112">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Base.any-Tuple{Function,MIToS.PDB.PDBResidue,MIToS.PDB.PDBResidue,Function}" href="#Base.any-Tuple{Function,MIToS.PDB.PDBResidue,MIToS.PDB.PDBResidue,Function}"><code>Base.any</code></a> — <span class="docstring-category">Method</span>.</div><div><p><code>any(f::Function, a::PDBResidue, b::PDBResidue, criteria::Function)</code></p><p>Test if the function <code>f</code> is true for any pair of atoms between the residues <code>a</code> and <code>b</code>. This function only test atoms that returns <code>true</code> for the fuction <code>criteria</code>.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Interaction.jl#L32-L37">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Base.any-Tuple{Function,MIToS.PDB.PDBResidue,MIToS.PDB.PDBResidue}" href="#Base.any-Tuple{Function,MIToS.PDB.PDBResidue,MIToS.PDB.PDBResidue}"><code>Base.any</code></a> — <span class="docstring-category">Method</span>.</div><div><p><code>any(f::Function, a::PDBResidue, b::PDBResidue)</code></p><p>Test if the function <code>f</code> is true for any pair of atoms between the residues <code>a</code> and <code>b</code></p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L444-L448">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Base.parse-Tuple{LightXML.XMLDocument,Type{MIToS.PDB.PDBML}}" href="#Base.parse-Tuple{LightXML.XMLDocument,Type{MIToS.PDB.PDBML}}"><code>Base.parse</code></a> — <span class="docstring-category">Method</span>.</div><div><p><code>parse(pdbml, ::Type{PDBML}; chain=All, model=All, group=All, atomname=All, onlyheavy=false, label=true, occupancyfilter=false)</code></p><p>Reads a <code>LightXML.XMLDocument</code> representing a pdb file. Returns a list of <code>PDBResidue</code>s (view <code>MIToS.PDB.PDBResidues</code>). Setting <code>chain</code>, <code>model</code>, <code>group</code>, <code>atomname</code> and <code>onlyheavy</code> values can be used to select of a subset of all residues. If not set, all residues are returned. If the keyword argument <code>label</code> (default: <code>true</code>) is <code>false</code>,the <strong>auth_</strong> attributes will be use instead of the <strong>label_</strong> attributes for <code>chain</code>, <code>atom</code> and residue <code>name</code> fields. The <strong>auth_</strong> attributes are alternatives provided by an author in order to match the identification/values used in the publication that describes the structure. If the keyword argument <code>occupancyfilter</code> (default: <code>false</code>) is <code>true</code>, only the atoms with the best occupancy are returned.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBMLParser.jl#L40-L53">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Base.parse-Tuple{Union{IO, String},Type{MIToS.PDB.PDBFile}}" href="#Base.parse-Tuple{Union{IO, String},Type{MIToS.PDB.PDBFile}}"><code>Base.parse</code></a> — <span class="docstring-category">Method</span>.</div><div><p><code>parse(io, ::Type{PDBFile}; chain=All, model=All, group=All, atomname=All, onlyheavy=false, occupancyfilter=false)</code></p><p>Reads a text file of a PDB entry. Returns a list of <code>PDBResidue</code> (view <code>MIToS.PDB.PDBResidues</code>). Setting <code>chain</code>, <code>model</code>, <code>group</code>, <code>atomname</code> and <code>onlyheavy</code> values can be used to select of a subset of all residues. Group can be <code>&quot;ATOM&quot;</code> or <code>&quot;HETATM&quot;</code>. If not set, all residues are returned. If the keyword argument <code>occupancyfilter</code> (default: <code>false</code>) is <code>true</code>, only the atoms with the best occupancy are returned.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBParser.jl#L30-L40">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Base.print" href="#Base.print"><code>Base.print</code></a> — <span class="docstring-category">Function</span>.</div><div><p><code>print(io, res, format::Type{PDBFile})</code> <code>print(res, format::Type{PDBFile})</code></p><p>Print a <code>PDBResidue</code> or a vector of <code>PDBResidue</code>s in PDB format.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBParser.jl#L296-L301">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.CAmatrix-Tuple{AbstractArray{MIToS.PDB.PDBResidue,1}}" href="#MIToS.PDB.CAmatrix-Tuple{AbstractArray{MIToS.PDB.PDBResidue,1}}"><code>MIToS.PDB.CAmatrix</code></a> — <span class="docstring-category">Method</span>.</div><div><p>Returns a matrix with the x, y and z coordinates of the Cα with best occupancy for each <code>PDBResidue</code> of the ATOM group. If a residue doesn&#39;t have a Cα, its Cα coordinates are NaNs.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Kabsch.jl#L79-L82">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.aromatic-Tuple{MIToS.PDB.PDBResidue,MIToS.PDB.PDBResidue}" href="#MIToS.PDB.aromatic-Tuple{MIToS.PDB.PDBResidue,MIToS.PDB.PDBResidue}"><code>MIToS.PDB.aromatic</code></a> — <span class="docstring-category">Method</span>.</div><div><p>There&#39;s an aromatic interaction if centriods are at 6.0 Å or less.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Interaction.jl#L165-L167">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.aromaticsulphur-Tuple{MIToS.PDB.PDBAtom,MIToS.PDB.PDBAtom,Any,Any}" href="#MIToS.PDB.aromaticsulphur-Tuple{MIToS.PDB.PDBAtom,MIToS.PDB.PDBAtom,Any,Any}"><code>MIToS.PDB.aromaticsulphur</code></a> — <span class="docstring-category">Method</span>.</div><div><p>Returns <code>true</code> if an sulphur and an aromatic atoms are 5.3 Å or less&quot;</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Interaction.jl#L129-L131">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.atoms" href="#MIToS.PDB.atoms"><code>MIToS.PDB.atoms</code></a> — <span class="docstring-category">Function</span>.</div><div><p><code>atoms(residue_list, model, chain, group, residue, atom)</code></p><p>These return a vector of <code>PDBAtom</code>s with the selected subset of atoms. You can use the type <code>All</code> (default value of the positional arguments) to avoid filtering a that level.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L228-L233">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.bestoccupancy-Tuple{Array{MIToS.PDB.PDBAtom,1}}" href="#MIToS.PDB.bestoccupancy-Tuple{Array{MIToS.PDB.PDBAtom,1}}"><code>MIToS.PDB.bestoccupancy</code></a> — <span class="docstring-category">Method</span>.</div><div><p>Takes a <code>Vector</code> of <code>PDBAtom</code>s and returns a <code>Vector</code> of the <code>PDBAtom</code>s with best occupancy.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L344-L346">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.center!-Tuple{Array{Float64,2}}" href="#MIToS.PDB.center!-Tuple{Array{Float64,2}}"><code>MIToS.PDB.center!</code></a> — <span class="docstring-category">Method</span>.</div><div><p><code>center!(A::Matrix{Float64})</code></p><p>Takes a set of points <code>A</code> as an NxD matrix (N: number of points, D: dimension). Translates <code>A</code> in place so that its centroid is at the origin of coordinates</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Kabsch.jl#L23-L28">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.centeredcoordinates" href="#MIToS.PDB.centeredcoordinates"><code>MIToS.PDB.centeredcoordinates</code></a> — <span class="docstring-category">Function</span>.</div><div><p>Returns a <code>Matrix{Float64}</code> with the centered coordinates of all the atoms in <code>residues</code>. An optional positional argument <code>CA</code> (default: <code>true</code>) defines if only Cα carbons should be used to center the matrix.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Kabsch.jl#L125-L129">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.centeredresidues" href="#MIToS.PDB.centeredresidues"><code>MIToS.PDB.centeredresidues</code></a> — <span class="docstring-category">Function</span>.</div><div><p>Returns a new <code>Vector{PDBResidue}</code> with the <code>PDBResidue</code>s having centered coordinates. An optional positional argument <code>CA</code> (default: <code>true</code>) defines if only Cα carbons should be used to center the matrix.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Kabsch.jl#L136-L140">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.change_coordinates" href="#MIToS.PDB.change_coordinates"><code>MIToS.PDB.change_coordinates</code></a> — <span class="docstring-category">Function</span>.</div><div><p><code>change_coordinates(residue::PDBResidue, coordinates::Matrix{Float64}, offset::Int=1)</code></p><p>Returns a new <code>PDBResidues</code> with (x,y,z) from a coordinates <code>Matrix{Float64}</code> You can give an <code>offset</code> indicating in wich matrix row starts the (x,y,z) coordinates of the residue.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Kabsch.jl#L159-L165">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.change_coordinates-Tuple{AbstractArray{MIToS.PDB.PDBResidue,1},Array{Float64,2}}" href="#MIToS.PDB.change_coordinates-Tuple{AbstractArray{MIToS.PDB.PDBResidue,1},Array{Float64,2}}"><code>MIToS.PDB.change_coordinates</code></a> — <span class="docstring-category">Method</span>.</div><div><p><code>change_coordinates(residues::AbstractVector{PDBResidue}, coordinates::Matrix{Float64})</code></p><p>Returns a new <code>Vector{PDBResidues}</code> with (x,y,z) from a coordinates <code>Matrix{Float64}</code></p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Kabsch.jl#L175-L179">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.change_coordinates-Tuple{MIToS.PDB.PDBAtom,MIToS.PDB.Coordinates}" href="#MIToS.PDB.change_coordinates-Tuple{MIToS.PDB.PDBAtom,MIToS.PDB.Coordinates}"><code>MIToS.PDB.change_coordinates</code></a> — <span class="docstring-category">Method</span>.</div><div><p><code>change_coordinates(atom::PDBAtom, coordinates::Coordinates)</code></p><p>Returns a new <code>PDBAtom</code> but with a new <code>coordinates</code></p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Kabsch.jl#L146-L150">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.check_atoms_for_interactions-Tuple{MIToS.PDB.PDBResidue}" href="#MIToS.PDB.check_atoms_for_interactions-Tuple{MIToS.PDB.PDBResidue}"><code>MIToS.PDB.check_atoms_for_interactions</code></a> — <span class="docstring-category">Method</span>.</div><div><p>This function takes a <code>PDBResidue</code> and returns <code>true</code> only if all the atoms can be used for checking interactions.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/AtomsData.jl#L430-L433">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.contact-Tuple{Array{MIToS.PDB.PDBResidue,1},AbstractFloat}" href="#MIToS.PDB.contact-Tuple{Array{MIToS.PDB.PDBResidue,1},AbstractFloat}"><code>MIToS.PDB.contact</code></a> — <span class="docstring-category">Method</span>.</div><div><p><code>contact(residues::Vector{PDBResidue}, limit::AbstractFloat; criteria::String=&quot;All&quot;)</code></p><p>If <code>contact</code> takes a <code>Vector{PDBResidue}</code>, It returns a matrix with all the pairwise comparisons (contact map).</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L568-L573">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.contact-Tuple{MIToS.PDB.Coordinates,MIToS.PDB.Coordinates,AbstractFloat}" href="#MIToS.PDB.contact-Tuple{MIToS.PDB.Coordinates,MIToS.PDB.Coordinates,AbstractFloat}"><code>MIToS.PDB.contact</code></a> — <span class="docstring-category">Method</span>.</div><div><p><code>contact(a::Coordinates, b::Coordinates, limit::AbstractFloat)</code></p><p>It returns true if the distance is less or equal to the limit. It doesn&#39;t call <code>sqrt</code> because it does <code>squared_distance(a,b) &lt;= limit^2</code>.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L94-L99">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.contact-Tuple{MIToS.PDB.PDBResidue,MIToS.PDB.PDBResidue,AbstractFloat}" href="#MIToS.PDB.contact-Tuple{MIToS.PDB.PDBResidue,MIToS.PDB.PDBResidue,AbstractFloat}"><code>MIToS.PDB.contact</code></a> — <span class="docstring-category">Method</span>.</div><div><p><code>contact(A::PDBResidue, B::PDBResidue, limit::AbstractFloat; criteria::String=&quot;All&quot;)</code></p><p>Returns <code>true</code> if the residues <code>A</code> and <code>B</code> are at contact distance (<code>limit</code>). The available distance <code>criteria</code> are: <code>Heavy</code>, <code>All</code>, <code>CA</code>, <code>CB</code> (<code>CA</code> for <code>GLY</code>)</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L462-L467">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.coordinatesmatrix-Tuple{MIToS.PDB.PDBResidue}" href="#MIToS.PDB.coordinatesmatrix-Tuple{MIToS.PDB.PDBResidue}"><code>MIToS.PDB.coordinatesmatrix</code></a> — <span class="docstring-category">Method</span>.</div><div><p>Returns a matrix with the x, y, z coordinates of each atom in each <code>PDBResidue</code></p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Kabsch.jl#L109">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.covalent-Tuple{MIToS.PDB.PDBAtom,MIToS.PDB.PDBAtom,Any,Any}" href="#MIToS.PDB.covalent-Tuple{MIToS.PDB.PDBAtom,MIToS.PDB.PDBAtom,Any,Any}"><code>MIToS.PDB.covalent</code></a> — <span class="docstring-category">Method</span>.</div><div><p>Returns <code>true</code> if the distance between atoms is less than the sum of the <code>covalentradius</code> of each atom.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Interaction.jl#L96-L99">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.distance-Tuple{Array{MIToS.PDB.PDBResidue,1}}" href="#MIToS.PDB.distance-Tuple{Array{MIToS.PDB.PDBResidue,1}}"><code>MIToS.PDB.distance</code></a> — <span class="docstring-category">Method</span>.</div><div><p><code>distance(residues::Vector{PDBResidue}; criteria::String=&quot;All&quot;)</code></p><p>If <code>distance</code> takes a <code>Vector{PDBResidue}</code> returns a <code>PairwiseListMatrix{Float64, false}</code> with all the pairwise comparisons (distance matrix).</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L583-L588">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.distance-Tuple{MIToS.PDB.Coordinates,MIToS.PDB.Coordinates}" href="#MIToS.PDB.distance-Tuple{MIToS.PDB.Coordinates,MIToS.PDB.Coordinates}"><code>MIToS.PDB.distance</code></a> — <span class="docstring-category">Method</span>.</div><div><p>It calculates the squared euclidean distance.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L81">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.disulphide-Tuple{MIToS.PDB.PDBAtom,MIToS.PDB.PDBAtom,Any,Any}" href="#MIToS.PDB.disulphide-Tuple{MIToS.PDB.PDBAtom,MIToS.PDB.PDBAtom,Any,Any}"><code>MIToS.PDB.disulphide</code></a> — <span class="docstring-category">Method</span>.</div><div><p>Returns <code>true</code> if two <code>CYS</code>&#39;s <code>S</code> are at 2.08 Å or less</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Interaction.jl#L114">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.downloadpdb-Union{Tuple{String}, Tuple{T}} where T&lt;:MIToS.Utils.Format" href="#MIToS.PDB.downloadpdb-Union{Tuple{String}, Tuple{T}} where T&lt;:MIToS.Utils.Format"><code>MIToS.PDB.downloadpdb</code></a> — <span class="docstring-category">Method</span>.</div><div><p>It downloads a gzipped PDB file from PDB database. It requires a four character <code>pdbcode</code>. Its default <code>format</code> is <code>PDBML</code> (PDB XML) and It uses the <code>baseurl</code> &quot;http://www.rcsb.org/pdb/files/&quot;. <code>filename</code> is the path/name of the output file. This function calls <code>MIToS.Utils.download_file</code> that calls <code>Requests.get_streaming</code>. You can use keyword arguments from <code>Requests.get_streaming</code> (e.g. <code>header</code>).</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBMLParser.jl#L211-L219">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.findCB-Tuple{MIToS.PDB.PDBResidue}" href="#MIToS.PDB.findCB-Tuple{MIToS.PDB.PDBResidue}"><code>MIToS.PDB.findCB</code></a> — <span class="docstring-category">Method</span>.</div><div><p>Returns a vector of indices for <code>CB</code> (<code>CA</code> for <code>GLY</code>)</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L309">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.findatoms-Tuple{Array{MIToS.PDB.PDBAtom,1},String}" href="#MIToS.PDB.findatoms-Tuple{Array{MIToS.PDB.PDBAtom,1},String}"><code>MIToS.PDB.findatoms</code></a> — <span class="docstring-category">Method</span>.</div><div><p><code>findatoms(res::PDBResidue, atom::String)</code></p><p>Returns a index vector of the atoms with the given <code>atom</code> name.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L298-L302">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.findheavy-Tuple{Array{MIToS.PDB.PDBAtom,1}}" href="#MIToS.PDB.findheavy-Tuple{Array{MIToS.PDB.PDBAtom,1}}"><code>MIToS.PDB.findheavy</code></a> — <span class="docstring-category">Method</span>.</div><div><p>Returns a list with the index of the heavy atoms (all atoms except hydrogen) in the <code>PDBResidue</code></p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L288-L291">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.getCA-Tuple{MIToS.PDB.PDBResidue}" href="#MIToS.PDB.getCA-Tuple{MIToS.PDB.PDBResidue}"><code>MIToS.PDB.getCA</code></a> — <span class="docstring-category">Method</span>.</div><div><p>Returns the Cα with best occupancy in the <code>PDBResidue</code>.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Kabsch.jl#L70">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.getpdbdescription-Tuple{String}" href="#MIToS.PDB.getpdbdescription-Tuple{String}"><code>MIToS.PDB.getpdbdescription</code></a> — <span class="docstring-category">Method</span>.</div><div><p>Access general information about a PDB entry (e.g., Header information) using the RESTful interface of the PDB database (describePDB). Returns a Dict for the four character <code>pdbcode</code>.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBMLParser.jl#L260-L264">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.hydrogenbond-Tuple{MIToS.PDB.PDBResidue,MIToS.PDB.PDBResidue}" href="#MIToS.PDB.hydrogenbond-Tuple{MIToS.PDB.PDBResidue,MIToS.PDB.PDBResidue}"><code>MIToS.PDB.hydrogenbond</code></a> — <span class="docstring-category">Method</span>.</div><div><p>This function only works if there are hydrogens in the structure. The criteria for a hydrogen bond are:</p><ul><li><p>d(Ai, Aj) &lt; 3.9Å</p></li><li><p>d(Ah, Aacc) &lt; 2.5Å</p></li><li><p>θ(Adon, Ah, Aacc) &gt; 90°</p></li><li><p>θ(Adon, Aacc, Aacc-antecedent) &gt; 90°</p></li><li><p>θ(Ah, Aacc, Aacc-antecedent) &gt; 90°</p></li></ul><p>Where Ah is the donated hydrogen atom, Adon is the hydrogen bond donor atom, Aacc is the hydrogen bond acceptor atom and Aacc-antecednt is the atom antecedent to the hydrogen bond acceptor atom.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Interaction.jl#L265-L278">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.hydrophobic-Tuple{MIToS.PDB.PDBAtom,MIToS.PDB.PDBAtom,Any,Any}" href="#MIToS.PDB.hydrophobic-Tuple{MIToS.PDB.PDBAtom,MIToS.PDB.PDBAtom,Any,Any}"><code>MIToS.PDB.hydrophobic</code></a> — <span class="docstring-category">Method</span>.</div><div><p>There&#39;s an hydrophobic interaction if two hydrophobic atoms are at 5.0 Å or less.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Interaction.jl#L201-L203">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.ionic-Tuple{MIToS.PDB.PDBAtom,MIToS.PDB.PDBAtom,Any,Any}" href="#MIToS.PDB.ionic-Tuple{MIToS.PDB.PDBAtom,MIToS.PDB.PDBAtom,Any,Any}"><code>MIToS.PDB.ionic</code></a> — <span class="docstring-category">Method</span>.</div><div><p>There&#39;s an ionic interaction if a cationic and an anionic atoms are at 6.0 Å or less.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Interaction.jl#L183-L185">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.isanionic-Tuple{MIToS.PDB.PDBAtom,String}" href="#MIToS.PDB.isanionic-Tuple{MIToS.PDB.PDBAtom,String}"><code>MIToS.PDB.isanionic</code></a> — <span class="docstring-category">Method</span>.</div><div><p>Returns true if the atom, e.g. <code>(&quot;GLU&quot;,&quot;CD&quot;)</code>, is an anionic atom in the residue.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Interaction.jl#L17-L19">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.isaromatic-Tuple{MIToS.PDB.PDBAtom,String}" href="#MIToS.PDB.isaromatic-Tuple{MIToS.PDB.PDBAtom,String}"><code>MIToS.PDB.isaromatic</code></a> — <span class="docstring-category">Method</span>.</div><div><p>Returns true if the atom, e.g. <code>(&quot;HIS&quot;,&quot;CG&quot;)</code>, is an aromatic atom in the residue.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Interaction.jl#L7-L9">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.isatom-Tuple{MIToS.PDB.PDBAtom,Any}" href="#MIToS.PDB.isatom-Tuple{MIToS.PDB.PDBAtom,Any}"><code>MIToS.PDB.isatom</code></a> — <span class="docstring-category">Method</span>.</div><div><p>It tests if the atom has the indicated atom name.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L149-L151">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.iscationic-Tuple{MIToS.PDB.PDBAtom,String}" href="#MIToS.PDB.iscationic-Tuple{MIToS.PDB.PDBAtom,String}"><code>MIToS.PDB.iscationic</code></a> — <span class="docstring-category">Method</span>.</div><div><p>Returns true if the atom, e.g. <code>(&quot;ARG&quot;,&quot;NE&quot;)</code>, is a cationic atom in the residue.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Interaction.jl#L12-L14">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.ishbondacceptor-Tuple{MIToS.PDB.PDBAtom,String}" href="#MIToS.PDB.ishbondacceptor-Tuple{MIToS.PDB.PDBAtom,String}"><code>MIToS.PDB.ishbondacceptor</code></a> — <span class="docstring-category">Method</span>.</div><div><p>Returns true if the atom, e.g. <code>(&quot;ARG&quot;,&quot;O&quot;)</code>, is an acceptor in H bonds.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Interaction.jl#L27-L29">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.ishbonddonor-Tuple{MIToS.PDB.PDBAtom,String}" href="#MIToS.PDB.ishbonddonor-Tuple{MIToS.PDB.PDBAtom,String}"><code>MIToS.PDB.ishbonddonor</code></a> — <span class="docstring-category">Method</span>.</div><div><p>Returns true if the atom, e.g. <code>(&quot;ARG&quot;,&quot;N&quot;)</code>, is a donor in H bonds.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Interaction.jl#L22-L24">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.isresidue-Tuple{MIToS.PDB.PDBResidueIdentifier,Any,Any,Any,Any}" href="#MIToS.PDB.isresidue-Tuple{MIToS.PDB.PDBResidueIdentifier,Any,Any,Any,Any}"><code>MIToS.PDB.isresidue</code></a> — <span class="docstring-category">Method</span>.</div><div><p>It tests if the PDB residue has the indicated model, chain, group and residue number.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L138-L140">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.kabsch-Tuple{Array{Float64,2},Array{Float64,2}}" href="#MIToS.PDB.kabsch-Tuple{Array{Float64,2},Array{Float64,2}}"><code>MIToS.PDB.kabsch</code></a> — <span class="docstring-category">Method</span>.</div><div><p><code>kabsch(A::Matrix{Float64}, B::Matrix{Float64})</code></p><p>This function takes two sets of points, <code>A</code> (refrence) and <code>B</code> as NxD matrices, where D is the dimension and N is the number of points. Assumes that the centroids of <code>A</code> and <code>B</code> are at the origin of coordinates. You can call <code>center!</code> on each matrix before calling <code>kabsch</code> to center the matrices in the <code>(0.0, 0.0, 0.0)</code>. Rotates <code>B</code> so that <code>rmsd(A,B)</code> is minimized. Returns the rotation matrix. You should do <code>B * RotationMatrix</code> to get the rotated B.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Kabsch.jl#L3-L13">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.mean_coordinates-Tuple{AbstractArray{Array{Float64,2},1}}" href="#MIToS.PDB.mean_coordinates-Tuple{AbstractArray{Array{Float64,2},1}}"><code>MIToS.PDB.mean_coordinates</code></a> — <span class="docstring-category">Method</span>.</div><div><p>Calculates the average/mean position of each atom in a set of structure. The function takes a vector (<code>AbstractVector</code>) of vectors (<code>Vector{PDBResidue}</code>) or matrices (<code>Matrix{Float64}</code>) as first argument. As second (optional) argument this function can take an <code>AbstractVector{Float64}</code> of matrix/structure weights to return a weighted mean. When a Vector{PDBResidue} is used, if the keyword argument <code>calpha</code> is <code>false</code> the RMSF is calculated for all the atoms. By default only alpha carbons are used (default: <code>calpha=true</code>).</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Kabsch.jl#L271-L279">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.pication-Tuple{MIToS.PDB.PDBAtom,MIToS.PDB.PDBAtom,Any,Any}" href="#MIToS.PDB.pication-Tuple{MIToS.PDB.PDBAtom,MIToS.PDB.PDBAtom,Any,Any}"><code>MIToS.PDB.pication</code></a> — <span class="docstring-category">Method</span>.</div><div><p>There&#39;s a Π-Cation interaction if a cationic and an aromatic atoms are at 6.0 Å or less</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Interaction.jl#L147-L149">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.proximitymean-Union{Tuple{Array{MIToS.PDB.PDBResidue,1},AbstractArray{T,1},T}, Tuple{Array{MIToS.PDB.PDBResidue,1},AbstractArray{T,1}}, Tuple{T}} where T&lt;:AbstractFloat" href="#MIToS.PDB.proximitymean-Union{Tuple{Array{MIToS.PDB.PDBResidue,1},AbstractArray{T,1},T}, Tuple{Array{MIToS.PDB.PDBResidue,1},AbstractArray{T,1}}, Tuple{T}} where T&lt;:AbstractFloat"><code>MIToS.PDB.proximitymean</code></a> — <span class="docstring-category">Method</span>.</div><div><p><code>proximitymean</code> calculates the proximity mean/average for each residue as the average score (from a <code>scores</code> list) of all the residues within a certain physical distance to a given amino acid. The score of that residue is not included in the mean unless you set <code>include</code> to <code>true</code>. The default values are 6.05 for the distance threshold/<code>limit</code> and <code>&quot;Heavy&quot;</code> for the <code>criteria</code> keyword argument. This function allows to calculate pMI (proximity mutual information) and pC (proximity conservation) as in Buslje et. al. 2010.</p><p>Buslje, Cristina Marino, Elin Teppa, Tomas Di Doménico, José María Delfino, and Morten Nielsen. <em>Networks of high mutual information define the structural proximity of catalytic sites: implications for catalytic residue identification.</em> PLoS Comput Biol 6, no. 11 (2010): e1000978.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L601-L612">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.residuepairsmatrix-Union{Tuple{Array{MIToS.PDB.PDBResidue,1},Type{T},Type{Val{diagonal}},T}, Tuple{T}, Tuple{diagonal}} where diagonal where T" href="#MIToS.PDB.residuepairsmatrix-Union{Tuple{Array{MIToS.PDB.PDBResidue,1},Type{T},Type{Val{diagonal}},T}, Tuple{T}, Tuple{diagonal}} where diagonal where T"><code>MIToS.PDB.residuepairsmatrix</code></a> — <span class="docstring-category">Method</span>.</div><div><p>It creates a <code>NamedArray</code> containing a <code>PairwiseListMatrix</code> where each element (column, row) is identified with a <code>PDBResidue</code> from the input vector. You can indicate the value type of the matrix (default to <code>Float64</code>), if the list should have the diagonal values (default to <code>Val{false}</code>) and the diagonal values (default to <code>NaN</code>).</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L532-L537">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.residues-Union{Tuple{AbstractArray{MIToS.PDB.PDBResidue,N},Any,Any,Any,Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue,N},Any,Any,Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue,N},Any,Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue,N},Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue,N}}, Tuple{N}} where N" href="#MIToS.PDB.residues-Union{Tuple{AbstractArray{MIToS.PDB.PDBResidue,N},Any,Any,Any,Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue,N},Any,Any,Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue,N},Any,Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue,N},Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue,N}}, Tuple{N}} where N"><code>MIToS.PDB.residues</code></a> — <span class="docstring-category">Method</span>.</div><div><p><code>residues(residue_list, model, chain, group, residue)</code></p><p>These return a new vector with the selected subset of residues from a list of residues. You can use the type <code>All</code> (default value) to avoid filtering a that level.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L157-L162">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.residuesdict-Union{Tuple{AbstractArray{MIToS.PDB.PDBResidue,N},Any,Any,Any,Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue,N},Any,Any,Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue,N},Any,Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue,N},Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue,N}}, Tuple{N}} where N" href="#MIToS.PDB.residuesdict-Union{Tuple{AbstractArray{MIToS.PDB.PDBResidue,N},Any,Any,Any,Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue,N},Any,Any,Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue,N},Any,Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue,N},Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue,N}}, Tuple{N}} where N"><code>MIToS.PDB.residuesdict</code></a> — <span class="docstring-category">Method</span>.</div><div><p><code>residuesdict(residue_list, model, chain, group, residue)</code></p><p>These return a dictionary (using PDB residue numbers as keys) with the selected subset of residues. You can use the type <code>All</code> (default value) to avoid filtering a that level.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L188-L193">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.rmsd-Tuple{AbstractArray{MIToS.PDB.PDBResidue,1},AbstractArray{MIToS.PDB.PDBResidue,1}}" href="#MIToS.PDB.rmsd-Tuple{AbstractArray{MIToS.PDB.PDBResidue,1},AbstractArray{MIToS.PDB.PDBResidue,1}}"><code>MIToS.PDB.rmsd</code></a> — <span class="docstring-category">Method</span>.</div><div><p><code>rmsd(A::AbstractVector{PDBResidue}, B::AbstractVector{PDBResidue}; superimposed::Bool=false)</code></p><p>Returns the Cα RMSD value between two PDB structures: <code>A</code> and <code>B</code>. If the structures are already superimposed between them, use <code>superimposed=true</code> to avoid a new superimposition (<code>superimposed</code> is <code>false</code> by default).</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Kabsch.jl#L51-L57">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.rmsd-Tuple{Array{Float64,2},Array{Float64,2}}" href="#MIToS.PDB.rmsd-Tuple{Array{Float64,2},Array{Float64,2}}"><code>MIToS.PDB.rmsd</code></a> — <span class="docstring-category">Method</span>.</div><div><p><code>rmsd(A::Matrix{Float64}, B::Matrix{Float64})</code></p><p>Return RMSD between two sets of points <code>A</code> and <code>B</code>, given as NxD matrices (N: number of points, D: dimension).</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Kabsch.jl#L35-L40">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.rmsf-Tuple{AbstractArray{Array{Float64,2},1}}" href="#MIToS.PDB.rmsf-Tuple{AbstractArray{Array{Float64,2},1}}"><code>MIToS.PDB.rmsf</code></a> — <span class="docstring-category">Method</span>.</div><div><p>Calculates the RMSF (Root Mean-Square-Fluctuation) between an atom and its average position in a set of structures. The function takes a vector (<code>AbstractVector</code>) of vectors (<code>Vector{PDBResidue}</code>) or matrices (<code>Matrix{Float64}</code>) as first argument. As second (optional) argument this function can take an <code>AbstractVector{Float64}</code> of matrix/structure weights to return the root weighted mean-square-fluctuation around the weighted mean structure. When a Vector{PDBResidue} is used, if the keyword argument <code>calpha</code> is <code>false</code> the RMSF is calculated for all the atoms. By default only alpha carbons are used (default: <code>calpha=true</code>).</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Kabsch.jl#L297-L306">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.selectbestoccupancy-Tuple{Array{MIToS.PDB.PDBAtom,1},Array{Int64,1}}" href="#MIToS.PDB.selectbestoccupancy-Tuple{Array{MIToS.PDB.PDBAtom,1},Array{Int64,1}}"><code>MIToS.PDB.selectbestoccupancy</code></a> — <span class="docstring-category">Method</span>.</div><div><p>Takes a <code>PDBResidue</code> and a <code>Vector</code> of atom indices. Returns the index value of the <code>Vector</code> with maximum occupancy.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L318-L321">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.squared_distance-Tuple{MIToS.PDB.PDBAtom,MIToS.PDB.PDBAtom}" href="#MIToS.PDB.squared_distance-Tuple{MIToS.PDB.PDBAtom,MIToS.PDB.PDBAtom}"><code>MIToS.PDB.squared_distance</code></a> — <span class="docstring-category">Method</span>.</div><div><p>It calculates the squared euclidean distance, i.e. it doesn&#39;t spend time in <code>sqrt</code></p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L78">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.squared_distance-Tuple{MIToS.PDB.PDBResidue,MIToS.PDB.PDBResidue}" href="#MIToS.PDB.squared_distance-Tuple{MIToS.PDB.PDBResidue,MIToS.PDB.PDBResidue}"><code>MIToS.PDB.squared_distance</code></a> — <span class="docstring-category">Method</span>.</div><div><p><code>squared_distance(A::PDBResidue, B::PDBResidue; criteria::String=&quot;All&quot;)</code></p><p>Returns the squared distance between the residues <code>A</code> and <code>B</code>. The available <code>criteria</code> are: <code>Heavy</code>, <code>All</code>, <code>CA</code>, <code>CB</code> (<code>CA</code> for <code>GLY</code>)</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/PDBResidues.jl#L388-L393">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.superimpose-Tuple{Array{MIToS.PDB.PDBResidue,1},Array{MIToS.PDB.PDBResidue,1}}" href="#MIToS.PDB.superimpose-Tuple{Array{MIToS.PDB.PDBResidue,1},Array{MIToS.PDB.PDBResidue,1}}"><code>MIToS.PDB.superimpose</code></a> — <span class="docstring-category">Method</span>.</div><div><p>This function takes <code>A::Vector{PDBResidue}</code> (reference) and <code>B::Vector{PDBResidue}</code>. Translates <code>A</code> and <code>B</code> to the origin of coordinates, and rotates <code>B</code> so that <code>rmsd(A,B)</code> is minimized with the Kabsch algorithm (using only their α carbons). Returns the rotated and translated versions of <code>A</code> and <code>B</code>, and the RMSD value.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Kabsch.jl#L207-L213">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.vanderwaals-Tuple{MIToS.PDB.PDBAtom,MIToS.PDB.PDBAtom,Any,Any}" href="#MIToS.PDB.vanderwaals-Tuple{MIToS.PDB.PDBAtom,MIToS.PDB.PDBAtom,Any,Any}"><code>MIToS.PDB.vanderwaals</code></a> — <span class="docstring-category">Method</span>.</div><div><p>Test if two atoms or residues are in van der Waals contact using: <code>distance(a,b) &lt;= 0.5 + vanderwaalsradius[a] + vanderwaalsradius[b]</code>. It returns distance <code>&lt;= 0.5</code> if the atoms aren&#39;t in <code>vanderwaalsradius</code>.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Interaction.jl#L62-L66">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MIToS.PDB.vanderwaalsclash-Tuple{MIToS.PDB.PDBAtom,MIToS.PDB.PDBAtom,Any,Any}" href="#MIToS.PDB.vanderwaalsclash-Tuple{MIToS.PDB.PDBAtom,MIToS.PDB.PDBAtom,Any,Any}"><code>MIToS.PDB.vanderwaalsclash</code></a> — <span class="docstring-category">Method</span>.</div><div><p>Returns <code>true</code> if the distance between the atoms is less than the sum of the <code>vanderwaalsradius</code> of the atoms. If the atoms aren&#39;t on the list (i.e. <code>OXT</code>), the <code>vanderwaalsradius</code> of the element is used. If there is not data in the dict, distance <code>0.0</code> is used.</p></div><a class="source-link" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/ca36d75659a6defbd88ac2dd57b1765eb4bbc140/src/PDB/Interaction.jl#L78-L83">source</a></section><footer><hr/><a class="previous" href="SIFTS_API.html"><span class="direction">Previous</span><span class="title">SIFTS</span></a><a class="next" href="Pfam_API.html"><span class="direction">Next</span><span class="title">Pfam</span></a></footer></article></body></html>
