<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>References · MIToS</title><meta name="title" content="References · MIToS"/><meta property="og:title" content="References · MIToS"/><meta property="twitter:title" content="References · MIToS"/><meta name="description" content="Documentation for MIToS."/><meta property="og:description" content="Documentation for MIToS."/><meta property="twitter:description" content="Documentation for MIToS."/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script><link href="../assets/extra_styles.css" rel="stylesheet" type="text/css"/><link href="../assets/citations.css" rel="stylesheet" type="text/css"/></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">MIToS</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../Installation/">Installation</a></li><li><a class="tocitem" href="../Example/">Example</a></li><li><span class="tocitem">Modules</span><ul><li><a class="tocitem" href="../MSA/">MSA</a></li><li><a class="tocitem" href="../Information/">Information</a></li><li><a class="tocitem" href="../SIFTS/">SIFTS</a></li><li><a class="tocitem" href="../PDB/">PDB</a></li><li><a class="tocitem" href="../Pfam/">Pfam</a></li></ul></li><li><span class="tocitem">Cookbook</span><ul><li><a class="tocitem" href="../01_Change_B_factors/">Change B-factors</a></li><li><a class="tocitem" href="../02_Linking_structural_and_evolutionary_information/">Linking structural and evolutionary information</a></li><li><a class="tocitem" href="../03_RMSF/">Root Mean Squared Fluctuation (RMSF)</a></li></ul></li><li><span class="tocitem">API</span><ul><li><a class="tocitem" href="../MSA_API/">MSA</a></li><li><a class="tocitem" href="../Information_API/">Information</a></li><li><a class="tocitem" href="../SIFTS_API/">SIFTS</a></li><li><a class="tocitem" href="../PDB_API/">PDB</a></li><li><a class="tocitem" href="../Pfam_API/">Pfam</a></li><li><a class="tocitem" href="../Utils_API/">Utils</a></li></ul></li><li><a class="tocitem" href="../Scripts/">MIToS&#39; Scripts</a></li><li class="is-active"><a class="tocitem" href>References</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>References</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>References</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/diegozea/MIToS.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/diegozea/MIToS.jl/blob/master/docs/src/References.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="References"><a class="docs-heading-anchor" href="#References">References</a><a id="References-1"></a><a class="docs-heading-anchor-permalink" href="#References" title="Permalink"></a></h1><div class="citation canonical"><dl><dt>[1]</dt><dd><div class="1093" id="10">D. J. Zea, D. Anfossi, M. Nielsen and C. Marino-Buslje. <em>MIToS. jl: mutual information tools for protein sequence analysis in the Julia language</em>. <a href="https://doi.org/10.1093/bioinformatics/btw646">Bioinformatics <strong>33</strong>, 564–565</a> (2017).</div></dd><dt>[2]</dt><dd><div class="1002 5560010313" id="10">U. Hobohm, M. Scharf, R. Schneider and C. Sander. <em>Selection of representative protein data sets</em>. <a href="https://doi.org/10.1002/pro.5560010313">Protein Science <strong>1</strong>, 409–417</a> (1992).</div></dd><dt>[3]</dt><dd><div class="1093" id="10">C. M. Buslje, J. Santos, J. M. Delfino and M. Nielsen. <em>Correction for phylogeny, small number of observations and data redundancy improves the identification of coevolving amino acid pairs using mutual information</em>. <a href="https://doi.org/10.1093/bioinformatics/btp135">Bioinformatics <strong>25</strong>, 1125–1131</a> (2009).</div></dd><dt>[4]</dt><dd><div class="1093 17 3389" id="10">S. F. Altschul, T. L. Madden, A. A. Schäffer, J. Zhang, Z. Zhang, W. Miller and D. J. Lipman. <em>Gapped BLAST and PSI-BLAST: a new generation of protein database search programs</em>. <a href="https://doi.org/10.1093/nar/25.17.3389">Nucleic acids research <strong>25</strong>, 3389–3402</a> (1997).</div></dd><dt>[5]</dt><dd><div class="1093" id="10">S. D. Dunn, L. M. Wahl and G. B. Gloor. <em>Mutual information without the influence of phylogeny or entropy dramatically improves residue contact prediction</em>. <a href="https://doi.org/10.1093/bioinformatics/btm604">Bioinformatics <strong>24</strong>, 333–340</a> (2008).</div></dd><dt>[6]</dt><dd><div class="1093" id="10">S. Velankar, J. M. Dana, J. Jacobsen, G. van Ginkel, P. J. Gane, J. Luo, T. J. Oldfield, C. O’Donovan, M.-J. Martin and G. J. Kleywegt. <em>SIFTS: Structure Integration with Function, Taxonomy and Sequences resource</em>. <a href="https://doi.org/10.1093/nar/gks1258">Nucleic Acids Research <strong>41</strong>, D483-D489</a> (2012).</div></dd><dt>[7]</dt><dd><div class="2144" id="10">P. Stothard. <em>The sequence manipulation suite: JavaScript programs for analyzing and formatting protein and DNA sequences</em>. <a href="https://doi.org/10.2144/00286ir01">Biotechniques <strong>28</strong>, 1102–1104</a> (2000).</div></dd><dt>[8]</dt><dd><div class="1093" id="10">B. J. Grant, A. P. Rodrigues, K. M. ElSawy, J. A. McCammon and L. S. Caves. <em>Bio3d: an R package for the comparative analysis of protein structures</em>. <a href="https://doi.org/10.1093/bioinformatics/btl461">Bioinformatics <strong>22</strong>, 2695–2696</a> (2006).</div></dd><dt>[9]</dt><dd><div class="1017" id="10">W. Perks. <em>Some observations on inverse probability including a new indifference rule</em>. <a href="https://doi.org/10.1017/S0020268100012270">Journal of the Institute of Actuaries <strong>73</strong>, 285–334</a> (1947).</div></dd><dt>[10]</dt><dd><div class="1214" id="10">S. Trybula. <em>Some problems of simultaneous minimax estimation</em>. <a href="https://doi.org/10.1214/aoms/1177706722">The Annals of Mathematical Statistics <strong>29</strong>, 245–253</a> (1958).</div></dd><dt>[11]</dt><dd><div class="1098 1946 0056" id="10">H. Jeffreys. <em>An invariant form for the prior probability in estimation problems</em>. <a href="https://doi.org/10.1098/rspa.1946.0056">Proceedings of the Royal Society of London. Series A. Mathematical and Physical Sciences <strong>186</strong>, 453–461</a> (1946).</div></dd><dt>[12]</dt><dd><div class="1371 pcbi 1000978" id="10">C. Marino Buslje, E. Teppa, T. Di Doménico, J. M. Delfino and M. Nielsen. <em>Networks of high mutual information define the structural proximity of catalytic sites: implications for catalytic residue identification</em>. <a href="https://doi.org/10.1371/journal.pcbi.1000978">PLoS computational biology <strong>6</strong>, e1000978</a> (2010).</div></dd></dl></div></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../Scripts/">« MIToS&#39; Scripts</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.5.0 on <span class="colophon-date" title="Tuesday 30 July 2024 16:11">Tuesday 30 July 2024</span>. Using Julia version 1.10.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
