<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>PDB · MIToS</title><meta name="title" content="PDB · MIToS"/><meta property="og:title" content="PDB · MIToS"/><meta property="twitter:title" content="PDB · MIToS"/><meta name="description" content="Documentation for MIToS."/><meta property="og:description" content="Documentation for MIToS."/><meta property="twitter:description" content="Documentation for MIToS."/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">MIToS</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../Installation/">Installation</a></li><li><a class="tocitem" href="../Example/">Example</a></li><li><span class="tocitem">Modules</span><ul><li><a class="tocitem" href="../MSA/">MSA</a></li><li><a class="tocitem" href="../Information/">Information</a></li><li><a class="tocitem" href="../SIFTS/">SIFTS</a></li><li><a class="tocitem" href="../PDB/">PDB</a></li><li><a class="tocitem" href="../Pfam/">Pfam</a></li></ul></li><li><span class="tocitem">Cookbook</span><ul><li><a class="tocitem" href="../01_Change_B_factors/">Change B-factors</a></li><li><a class="tocitem" href="../02_Linking_structural_and_evolutionary_information/">Linking structural and evolutionary information</a></li><li><a class="tocitem" href="../03_RMSF/">Root Mean Squared Fluctuation (RMSF)</a></li></ul></li><li><span class="tocitem">API</span><ul><li><a class="tocitem" href="../MSA_API/">MSA</a></li><li><a class="tocitem" href="../Information_API/">Information</a></li><li><a class="tocitem" href="../SIFTS_API/">SIFTS</a></li><li class="is-active"><a class="tocitem" href>PDB</a><ul class="internal"><li><a class="tocitem" href="#Contents"><span>Contents</span></a></li><li><a class="tocitem" href="#Types"><span>Types</span></a></li><li><a class="tocitem" href="#Constants"><span>Constants</span></a></li><li><a class="tocitem" href="#Macros"><span>Macros</span></a></li><li><a class="tocitem" href="#Methods-and-functions"><span>Methods and functions</span></a></li></ul></li><li><a class="tocitem" href="../Pfam_API/">Pfam</a></li><li><a class="tocitem" href="../Utils_API/">Utils</a></li></ul></li><li><a class="tocitem" href="../Scripts/">Scripts</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">API</a></li><li class="is-active"><a href>PDB</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>PDB</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/diegozea/MIToS.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/diegozea/MIToS.jl/blob/master/docs/src/PDB_API.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="PDB"><a class="docs-heading-anchor" href="#PDB">PDB</a><a id="PDB-1"></a><a class="docs-heading-anchor-permalink" href="#PDB" title="Permalink"></a></h1><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB" href="#MIToS.PDB"><code>MIToS.PDB</code></a> — <span class="docstring-category">Module</span></header><section><div><p>The module <code>PDB</code> defines types and methods to work with protein structures inside Julia. It is useful to link structural and sequential information, and needed for measure the predictive performance at protein contact prediction of mutual information scores.</p><p><strong>Features</strong></p><ul><li>Read and parse PDF and PDBML files</li><li>Calculate distance and contacts between atoms or residues</li><li>Determine interaction between residues</li></ul><pre><code class="language-julia hljs">using MIToS.PDB</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDB.jl#L1-L15">source</a></section></article><h2 id="Contents"><a class="docs-heading-anchor" href="#Contents">Contents</a><a id="Contents-1"></a><a class="docs-heading-anchor-permalink" href="#Contents" title="Permalink"></a></h2><ul><li><a href="#PDB">PDB</a></li><li class="no-marker"><ul><li><a href="#Contents">Contents</a></li><li><a href="#Types">Types</a></li><li><a href="#Constants">Constants</a></li><li><a href="#Macros">Macros</a></li><li><a href="#Methods-and-functions">Methods and functions</a></li></ul></li></ul><h2 id="Types"><a class="docs-heading-anchor" href="#Types">Types</a><a id="Types-1"></a><a class="docs-heading-anchor-permalink" href="#Types" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.Coordinates" href="#MIToS.PDB.Coordinates"><code>MIToS.PDB.Coordinates</code></a> — <span class="docstring-category">Type</span></header><section><div><p>A <code>Coordinates</code> object is a fixed size vector with the coordinates x,y,z.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L24">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.PDBAtom" href="#MIToS.PDB.PDBAtom"><code>MIToS.PDB.PDBAtom</code></a> — <span class="docstring-category">Type</span></header><section><div><p>A <code>PDBAtom</code> object contains the information from a PDB atom, without information of the residue. It has the following fields that you can access at any moment for query purposes:</p><pre><code class="nohighlight hljs">- `coordinates` : x,y,z coordinates, e.g. `Coordinates(109.641,73.162,42.7)`.
- `atom` : Atom name, e.g. `&quot;CA&quot;`.
- `element` : Element type of the atom, e.g. `&quot;C&quot;`.
- `occupancy` : A float number with the occupancy, e.g. `1.0`.
- `B` : B factor as a string, e.g. `&quot;23.60&quot;`.</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L35-L44">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.PDBFile" href="#MIToS.PDB.PDBFile"><code>MIToS.PDB.PDBFile</code></a> — <span class="docstring-category">Type</span></header><section><div><p><code>PDBFile &lt;: FileFormat</code></p><p>Protein Data Bank (PDB) format. It provides a standard representation for macromolecular structure data derived from X-ray diffraction and NMR studies.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBParser.jl#L1-L7">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.PDBML" href="#MIToS.PDB.PDBML"><code>MIToS.PDB.PDBML</code></a> — <span class="docstring-category">Type</span></header><section><div><p><code>PDBML &lt;: FileFormat</code></p><p>Protein Data Bank Markup Language (PDBML), a representation of PDB data in XML format.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBMLParser.jl#L1-L5">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.PDBResidue" href="#MIToS.PDB.PDBResidue"><code>MIToS.PDB.PDBResidue</code></a> — <span class="docstring-category">Type</span></header><section><div><p>A <code>PDBResidue</code> object contains all the information about a PDB residue. It has the following fields that you can access at any moment for query purposes:</p><pre><code class="nohighlight hljs">- `id` : A `PDBResidueIdentifier` object.
- `atoms` : A vector of `PDBAtom`s.</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L53-L59">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.PDBResidueIdentifier" href="#MIToS.PDB.PDBResidueIdentifier"><code>MIToS.PDB.PDBResidueIdentifier</code></a> — <span class="docstring-category">Type</span></header><section><div><p>A <code>PDBResidueIdentifier</code> object contains the information needed to identity PDB residues. It has the following fields that you can access at any moment for query purposes:</p><pre><code class="nohighlight hljs">- `PDBe_number` : It&#39;s only used when a PDBML is readed (PDBe number as a string).
- `number` : PDB residue number, it includes insertion codes, e.g. `&quot;34A&quot;`.
- `name` : Three letter residue name in PDB, e.g. `&quot;LYS&quot;`.
- `group` : It can be `&quot;ATOM&quot;` or `&quot;HETATM&quot;`.
- `model` : The model number as a string, e.g. `&quot;1&quot;`.
- `chain` : The chain as a string, e.g. `&quot;A&quot;`.</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L4-L14">source</a></section></article><h2 id="Constants"><a class="docs-heading-anchor" href="#Constants">Constants</a><a id="Constants-1"></a><a class="docs-heading-anchor-permalink" href="#Constants" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.covalentradius" href="#MIToS.PDB.covalentradius"><code>MIToS.PDB.covalentradius</code></a> — <span class="docstring-category">Constant</span></header><section><div><p>Covalent radius in Å of each element from the Additional file 1 of PICCOLO [1]. Hydrogen was updated using the value on Table 2 from Cordero et. al. [2].</p><ol><li>Bickerton, G. R., Higueruelo, A. P., &amp; Blundell, T. L. (2011).</li></ol><p>Comprehensive, atomic-level characterization of structurally characterized protein-protein interactions: the PICCOLO database. BMC bioinformatics, 12(1), 313.</p><ol><li>Cordero, B., Gómez, V., Platero-Prats, A. E., Revés, M.,</li></ol><p>Echeverría, J., Cremades, E., ... &amp; Alvarez, S. (2008). Covalent radii revisited. Dalton Transactions, (21), 2832-2838.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/AtomsData.jl#L1-L12">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.vanderwaalsradius" href="#MIToS.PDB.vanderwaalsradius"><code>MIToS.PDB.vanderwaalsradius</code></a> — <span class="docstring-category">Constant</span></header><section><div><p>van der Waals radius in Å from the Additional file 1 of Bickerton et. al. 2011</p><ul><li>Bickerton, G. R., Higueruelo, A. P., &amp; Blundell, T. L. (2011).</li></ul><p>Comprehensive, atomic-level characterization of structurally characterized protein-protein interactions: the PICCOLO database. BMC bioinformatics, 12(1), 313.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/AtomsData.jl#L21-L27">source</a></section></article><h2 id="Macros"><a class="docs-heading-anchor" href="#Macros">Macros</a><a id="Macros-1"></a><a class="docs-heading-anchor-permalink" href="#Macros" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.@atoms-Tuple{Any, Symbol, Any, Symbol, Any, Symbol, Any, Symbol, Any, Symbol, Any}" href="#MIToS.PDB.@atoms-Tuple{Any, Symbol, Any, Symbol, Any, Symbol, Any, Symbol, Any, Symbol, Any}"><code>MIToS.PDB.@atoms</code></a> — <span class="docstring-category">Macro</span></header><section><div><p><code>@atoms ... model ... chain ... group ... residue ... atom ...</code></p><p>These return a vector of <code>PDBAtom</code>s with the selected subset of atoms. You can use the type <code>All</code> to avoid filtering that option.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L249-L254">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.@residues-Tuple{Any, Symbol, Any, Symbol, Any, Symbol, Any, Symbol, Any}" href="#MIToS.PDB.@residues-Tuple{Any, Symbol, Any, Symbol, Any, Symbol, Any, Symbol, Any}"><code>MIToS.PDB.@residues</code></a> — <span class="docstring-category">Macro</span></header><section><div><p><code>@residues ... model ... chain ... group ... residue ...</code></p><p>These return a new vector with the selected subset of residues from a list of residues. You can use the type <code>All</code> to avoid filtering that option.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L169-L174">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.@residuesdict-Tuple{Any, Symbol, Any, Symbol, Any, Symbol, Any, Symbol, Any}" href="#MIToS.PDB.@residuesdict-Tuple{Any, Symbol, Any, Symbol, Any, Symbol, Any, Symbol, Any}"><code>MIToS.PDB.@residuesdict</code></a> — <span class="docstring-category">Macro</span></header><section><div><p><code>@residuesdict ... model ... chain ... group ... residue ...</code></p><p>These return a dictionary (using PDB residue numbers as keys) with the selected subset of residues. You can use the type <code>All</code> to avoid filtering that option.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L206-L211">source</a></section></article><h2 id="Methods-and-functions"><a class="docs-heading-anchor" href="#Methods-and-functions">Methods and functions</a><a id="Methods-and-functions-1"></a><a class="docs-heading-anchor-permalink" href="#Methods-and-functions" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.angle-Tuple{MIToS.PDB.Coordinates, MIToS.PDB.Coordinates, MIToS.PDB.Coordinates}" href="#Base.angle-Tuple{MIToS.PDB.Coordinates, MIToS.PDB.Coordinates, MIToS.PDB.Coordinates}"><code>Base.angle</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>angle(a::Coordinates, b::Coordinates, c::Coordinates)</code></p><p>Angle (in degrees) at <code>b</code> between <code>a-b</code> and <code>b-c</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L109-L113">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.any-Tuple{Function, MIToS.PDB.PDBResidue, MIToS.PDB.PDBResidue, Function}" href="#Base.any-Tuple{Function, MIToS.PDB.PDBResidue, MIToS.PDB.PDBResidue, Function}"><code>Base.any</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>any(f::Function, a::PDBResidue, b::PDBResidue, criteria::Function)</code></p><p>Test if the function <code>f</code> is true for any pair of atoms between the residues <code>a</code> and <code>b</code>. This function only test atoms that returns <code>true</code> for the fuction <code>criteria</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Interaction.jl#L32-L37">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.any-Tuple{Function, MIToS.PDB.PDBResidue, MIToS.PDB.PDBResidue}" href="#Base.any-Tuple{Function, MIToS.PDB.PDBResidue, MIToS.PDB.PDBResidue}"><code>Base.any</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>any(f::Function, a::PDBResidue, b::PDBResidue)</code></p><p>Test if the function <code>f</code> is true for any pair of atoms between the residues <code>a</code> and <code>b</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L445-L449">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.parse-Tuple{LightXML.XMLDocument, Type{MIToS.PDB.PDBML}}" href="#Base.parse-Tuple{LightXML.XMLDocument, Type{MIToS.PDB.PDBML}}"><code>Base.parse</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>parse(pdbml, ::Type{PDBML}; chain=All, model=All, group=All, atomname=All, onlyheavy=false, label=true, occupancyfilter=false)</code></p><p>Reads a <code>LightXML.XMLDocument</code> representing a pdb file. Returns a list of <code>PDBResidue</code>s (view <code>MIToS.PDB.PDBResidues</code>). Setting <code>chain</code>, <code>model</code>, <code>group</code>, <code>atomname</code> and <code>onlyheavy</code> values can be used to select of a subset of all residues. If not set, all residues are returned. If the keyword argument <code>label</code> (default: <code>true</code>) is <code>false</code>,the <strong>auth_</strong> attributes will be use instead of the <strong>label_</strong> attributes for <code>chain</code>, <code>atom</code> and residue <code>name</code> fields. The <strong>auth_</strong> attributes are alternatives provided by an author in order to match the identification/values used in the publication that describes the structure. If the keyword argument <code>occupancyfilter</code> (default: <code>false</code>) is <code>true</code>, only the atoms with the best occupancy are returned.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBMLParser.jl#L40-L53">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.parse-Tuple{Union{IO, String}, Type{MIToS.PDB.PDBFile}}" href="#Base.parse-Tuple{Union{IO, String}, Type{MIToS.PDB.PDBFile}}"><code>Base.parse</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>parse(io, ::Type{PDBFile}; chain=All, model=All, group=All, atomname=All, onlyheavy=false, occupancyfilter=false)</code></p><p>Reads a text file of a PDB entry. Returns a list of <code>PDBResidue</code> (view <code>MIToS.PDB.PDBResidues</code>). Setting <code>chain</code>, <code>model</code>, <code>group</code>, <code>atomname</code> and <code>onlyheavy</code> values can be used to select of a subset of all residues. Group can be <code>&quot;ATOM&quot;</code> or <code>&quot;HETATM&quot;</code>. If not set, all residues are returned. If the keyword argument <code>occupancyfilter</code> (default: <code>false</code>) is <code>true</code>, only the atoms with the best occupancy are returned.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBParser.jl#L30-L40">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.print" href="#Base.print"><code>Base.print</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>print(io, res, format::Type{PDBFile})</code> <code>print(res, format::Type{PDBFile})</code></p><p>Print a <code>PDBResidue</code> or a vector of <code>PDBResidue</code>s in PDB format.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBParser.jl#L304-L309">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.CAmatrix-Tuple{AbstractVector{MIToS.PDB.PDBResidue}}" href="#MIToS.PDB.CAmatrix-Tuple{AbstractVector{MIToS.PDB.PDBResidue}}"><code>MIToS.PDB.CAmatrix</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Returns a matrix with the x, y and z coordinates of the Cα with best occupancy for each <code>PDBResidue</code> of the ATOM group. If a residue doesn&#39;t have a Cα, its Cα coordinates are NaNs.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Kabsch.jl#L91-L94">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.aromatic-Tuple{MIToS.PDB.PDBResidue, MIToS.PDB.PDBResidue}" href="#MIToS.PDB.aromatic-Tuple{MIToS.PDB.PDBResidue, MIToS.PDB.PDBResidue}"><code>MIToS.PDB.aromatic</code></a> — <span class="docstring-category">Method</span></header><section><div><p>There&#39;s an aromatic interaction if centriods are at 6.0 Å or less.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Interaction.jl#L165-L167">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.aromaticsulphur-Tuple{MIToS.PDB.PDBAtom, MIToS.PDB.PDBAtom, Any, Any}" href="#MIToS.PDB.aromaticsulphur-Tuple{MIToS.PDB.PDBAtom, MIToS.PDB.PDBAtom, Any, Any}"><code>MIToS.PDB.aromaticsulphur</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Returns <code>true</code> if an sulphur and an aromatic atoms are 5.3 Å or less&quot;</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Interaction.jl#L129-L131">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.atoms" href="#MIToS.PDB.atoms"><code>MIToS.PDB.atoms</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>atoms(residue_list, model, chain, group, residue, atom)</code></p><p>These return a vector of <code>PDBAtom</code>s with the selected subset of atoms. You can use the type <code>All</code> (default value of the positional arguments) to avoid filtering a that level.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L229-L234">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.bestoccupancy-Tuple{Vector{MIToS.PDB.PDBAtom}}" href="#MIToS.PDB.bestoccupancy-Tuple{Vector{MIToS.PDB.PDBAtom}}"><code>MIToS.PDB.bestoccupancy</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Takes a <code>Vector</code> of <code>PDBAtom</code>s and returns a <code>Vector</code> of the <code>PDBAtom</code>s with best occupancy.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L345-L347">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.center!-Tuple{AbstractMatrix{Float64}}" href="#MIToS.PDB.center!-Tuple{AbstractMatrix{Float64}}"><code>MIToS.PDB.center!</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>center!(A::AbstractMatrix{Float64})</code></p><p>Takes a set of points <code>A</code> as an NxD matrix (N: number of points, D: dimension). Translates <code>A</code> in place so that its centroid is at the origin of coordinates</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Kabsch.jl#L23-L28">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.centeredcoordinates" href="#MIToS.PDB.centeredcoordinates"><code>MIToS.PDB.centeredcoordinates</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Returns a <code>Matrix{Float64}</code> with the centered coordinates of all the atoms in <code>residues</code>. An optional positional argument <code>CA</code> (default: <code>true</code>) defines if only Cα carbons should be used to center the matrix.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Kabsch.jl#L139-L143">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.centeredresidues" href="#MIToS.PDB.centeredresidues"><code>MIToS.PDB.centeredresidues</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Returns a new <code>Vector{PDBResidue}</code> with the <code>PDBResidue</code>s having centered coordinates. An optional positional argument <code>CA</code> (default: <code>true</code>) defines if only Cα carbons should be used to center the matrix.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Kabsch.jl#L150-L154">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.change_coordinates" href="#MIToS.PDB.change_coordinates"><code>MIToS.PDB.change_coordinates</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>change_coordinates(residue::PDBResidue, coordinates::AbstractMatrix{Float64}, offset::Int=1)</code></p><p>Returns a new <code>PDBResidues</code> with (x,y,z) from a coordinates <code>AbstractMatrix{Float64}</code> You can give an <code>offset</code> indicating in wich matrix row starts the (x,y,z) coordinates of the residue.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Kabsch.jl#L173-L179">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.change_coordinates-Tuple{AbstractVector{MIToS.PDB.PDBResidue}, AbstractMatrix{Float64}}" href="#MIToS.PDB.change_coordinates-Tuple{AbstractVector{MIToS.PDB.PDBResidue}, AbstractMatrix{Float64}}"><code>MIToS.PDB.change_coordinates</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>change_coordinates(residues::AbstractVector{PDBResidue}, coordinates::AbstractMatrix{Float64})</code></p><p>Returns a new <code>Vector{PDBResidues}</code> with (x,y,z) from a coordinates <code>Matrix{Float64}</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Kabsch.jl#L189-L193">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.change_coordinates-Tuple{MIToS.PDB.PDBAtom, MIToS.PDB.Coordinates}" href="#MIToS.PDB.change_coordinates-Tuple{MIToS.PDB.PDBAtom, MIToS.PDB.Coordinates}"><code>MIToS.PDB.change_coordinates</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>change_coordinates(atom::PDBAtom, coordinates::Coordinates)</code></p><p>Returns a new <code>PDBAtom</code> but with a new <code>coordinates</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Kabsch.jl#L160-L164">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.check_atoms_for_interactions-Tuple{MIToS.PDB.PDBResidue}" href="#MIToS.PDB.check_atoms_for_interactions-Tuple{MIToS.PDB.PDBResidue}"><code>MIToS.PDB.check_atoms_for_interactions</code></a> — <span class="docstring-category">Method</span></header><section><div><p>This function takes a <code>PDBResidue</code> and returns <code>true</code> only if all the atoms can be used for checking interactions.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/AtomsData.jl#L430-L433">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.contact-Tuple{MIToS.PDB.Coordinates, MIToS.PDB.Coordinates, AbstractFloat}" href="#MIToS.PDB.contact-Tuple{MIToS.PDB.Coordinates, MIToS.PDB.Coordinates, AbstractFloat}"><code>MIToS.PDB.contact</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>contact(a::Coordinates, b::Coordinates, limit::AbstractFloat)</code></p><p>It returns true if the distance is less or equal to the limit. It doesn&#39;t call <code>sqrt</code> because it does <code>squared_distance(a,b) &lt;= limit^2</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L95-L100">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.contact-Tuple{MIToS.PDB.PDBResidue, MIToS.PDB.PDBResidue, AbstractFloat}" href="#MIToS.PDB.contact-Tuple{MIToS.PDB.PDBResidue, MIToS.PDB.PDBResidue, AbstractFloat}"><code>MIToS.PDB.contact</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>contact(A::PDBResidue, B::PDBResidue, limit::AbstractFloat; criteria::String=&quot;All&quot;)</code></p><p>Returns <code>true</code> if the residues <code>A</code> and <code>B</code> are at contact distance (<code>limit</code>). The available distance <code>criteria</code> are: <code>Heavy</code>, <code>All</code>, <code>CA</code>, <code>CB</code> (<code>CA</code> for <code>GLY</code>)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L463-L468">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.contact-Tuple{Vector{MIToS.PDB.PDBResidue}, AbstractFloat}" href="#MIToS.PDB.contact-Tuple{Vector{MIToS.PDB.PDBResidue}, AbstractFloat}"><code>MIToS.PDB.contact</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>contact(residues::Vector{PDBResidue}, limit::AbstractFloat; criteria::String=&quot;All&quot;)</code></p><p>If <code>contact</code> takes a <code>Vector{PDBResidue}</code>, It returns a matrix with all the pairwise comparisons (contact map).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L569-L574">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.coordinatesmatrix-Tuple{MIToS.PDB.PDBResidue}" href="#MIToS.PDB.coordinatesmatrix-Tuple{MIToS.PDB.PDBResidue}"><code>MIToS.PDB.coordinatesmatrix</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Returns a matrix with the x, y, z coordinates of each atom in each <code>PDBResidue</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Kabsch.jl#L121">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.covalent-Tuple{MIToS.PDB.PDBAtom, MIToS.PDB.PDBAtom, Any, Any}" href="#MIToS.PDB.covalent-Tuple{MIToS.PDB.PDBAtom, MIToS.PDB.PDBAtom, Any, Any}"><code>MIToS.PDB.covalent</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Returns <code>true</code> if the distance between atoms is less than the sum of the <code>covalentradius</code> of each atom.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Interaction.jl#L96-L99">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.distance-Tuple{MIToS.PDB.Coordinates, MIToS.PDB.Coordinates}" href="#MIToS.PDB.distance-Tuple{MIToS.PDB.Coordinates, MIToS.PDB.Coordinates}"><code>MIToS.PDB.distance</code></a> — <span class="docstring-category">Method</span></header><section><div><p>It calculates the squared euclidean distance.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L82">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.distance-Tuple{Vector{MIToS.PDB.PDBResidue}}" href="#MIToS.PDB.distance-Tuple{Vector{MIToS.PDB.PDBResidue}}"><code>MIToS.PDB.distance</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>distance(residues::Vector{PDBResidue}; criteria::String=&quot;All&quot;)</code></p><p>If <code>distance</code> takes a <code>Vector{PDBResidue}</code> returns a <code>PairwiseListMatrix{Float64, false}</code> with all the pairwise comparisons (distance matrix).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L584-L589">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.disulphide-Tuple{MIToS.PDB.PDBAtom, MIToS.PDB.PDBAtom, Any, Any}" href="#MIToS.PDB.disulphide-Tuple{MIToS.PDB.PDBAtom, MIToS.PDB.PDBAtom, Any, Any}"><code>MIToS.PDB.disulphide</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Returns <code>true</code> if two <code>CYS</code>&#39;s <code>S</code> are at 2.08 Å or less</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Interaction.jl#L114">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.download_alphafold_structure-Union{Tuple{String}, Tuple{T}} where T&lt;:MIToS.Utils.FileFormat" href="#MIToS.PDB.download_alphafold_structure-Union{Tuple{String}, Tuple{T}} where T&lt;:MIToS.Utils.FileFormat"><code>MIToS.PDB.download_alphafold_structure</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">download_alphafold_structure(uniprot_id::String; format::Type{T}=PDBFile) where T&lt;:FileFormat</code></pre><p>This function downloads the structure file (PDB or CIF) for a given UniProt ID from AlphaFoldDB. The <code>uniprot_id</code> parameter specifies the UniProt ID of the protein. The <code>format</code> parameter specifies the file format to download, with the default being PDBFile.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/AlphaFoldDB.jl#L29-L35">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.downloadpdb-Union{Tuple{String}, Tuple{T}} where T&lt;:MIToS.Utils.FileFormat" href="#MIToS.PDB.downloadpdb-Union{Tuple{String}, Tuple{T}} where T&lt;:MIToS.Utils.FileFormat"><code>MIToS.PDB.downloadpdb</code></a> — <span class="docstring-category">Method</span></header><section><div><p>It downloads a gzipped PDB file from PDB database. It requires a four character <code>pdbcode</code>. Its default <code>format</code> is <code>PDBML</code> (PDB XML) and It uses the <code>baseurl</code>  &quot;http://www.rcsb.org/pdb/files/&quot;. <code>filename</code> is the path/name of the output file. This function calls <code>MIToS.Utils.download_file</code> that calls <code>download</code> from the <em>HTTP.jl</em>  package. You can use keyword arguments from <code>HTTP.request</code>. Use the <code>headers</code> keyword argument to pass a <code>Dict{String, String}</code> with the  header information.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBMLParser.jl#L211-L221">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.downloadpdbheader-Tuple{String}" href="#MIToS.PDB.downloadpdbheader-Tuple{String}"><code>MIToS.PDB.downloadpdbheader</code></a> — <span class="docstring-category">Method</span></header><section><div><p>It downloads a JSON file containing the PDB header information.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBMLParser.jl#L290-L292">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.findCB-Tuple{MIToS.PDB.PDBResidue}" href="#MIToS.PDB.findCB-Tuple{MIToS.PDB.PDBResidue}"><code>MIToS.PDB.findCB</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Returns a vector of indices for <code>CB</code> (<code>CA</code> for <code>GLY</code>)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L310">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.findatoms-Tuple{Vector{MIToS.PDB.PDBAtom}, String}" href="#MIToS.PDB.findatoms-Tuple{Vector{MIToS.PDB.PDBAtom}, String}"><code>MIToS.PDB.findatoms</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>findatoms(res::PDBResidue, atom::String)</code></p><p>Returns a index vector of the atoms with the given <code>atom</code> name.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L299-L303">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.findheavy-Tuple{Vector{MIToS.PDB.PDBAtom}}" href="#MIToS.PDB.findheavy-Tuple{Vector{MIToS.PDB.PDBAtom}}"><code>MIToS.PDB.findheavy</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Returns a list with the index of the heavy atoms (all atoms except hydrogen) in the <code>PDBResidue</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L289-L292">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.getCA-Tuple{MIToS.PDB.PDBResidue}" href="#MIToS.PDB.getCA-Tuple{MIToS.PDB.PDBResidue}"><code>MIToS.PDB.getCA</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Returns the Cα with best occupancy in the <code>PDBResidue</code>.  If the <code>PDBResidue</code> has no Cα, <code>missing</code> is returned.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Kabsch.jl#L70-L73">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.getpdbdescription-Tuple{String}" href="#MIToS.PDB.getpdbdescription-Tuple{String}"><code>MIToS.PDB.getpdbdescription</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Access general information about a PDB entry (e.g., Header information) using the GraphQL interface of the PDB database. It parses the JSON answer into a <code>JSON3.Object</code> that  can be used as a dictionary.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBMLParser.jl#L300-L304">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.hydrogenbond-Tuple{MIToS.PDB.PDBResidue, MIToS.PDB.PDBResidue}" href="#MIToS.PDB.hydrogenbond-Tuple{MIToS.PDB.PDBResidue, MIToS.PDB.PDBResidue}"><code>MIToS.PDB.hydrogenbond</code></a> — <span class="docstring-category">Method</span></header><section><div><p>This function only works if there are hydrogens in the structure. The criteria for a hydrogen bond are:</p><ul><li>d(Ai, Aj) &lt; 3.9Å</li><li>d(Ah, Aacc) &lt; 2.5Å</li><li>θ(Adon, Ah, Aacc) &gt; 90°</li><li>θ(Adon, Aacc, Aacc-antecedent) &gt; 90°</li><li>θ(Ah, Aacc, Aacc-antecedent) &gt; 90°</li></ul><p>Where Ah is the donated hydrogen atom, Adon is the hydrogen bond donor atom, Aacc is the hydrogen bond acceptor atom and Aacc-antecednt is the atom antecedent to the hydrogen bond acceptor atom.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Interaction.jl#L265-L278">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.hydrophobic-Tuple{MIToS.PDB.PDBAtom, MIToS.PDB.PDBAtom, Any, Any}" href="#MIToS.PDB.hydrophobic-Tuple{MIToS.PDB.PDBAtom, MIToS.PDB.PDBAtom, Any, Any}"><code>MIToS.PDB.hydrophobic</code></a> — <span class="docstring-category">Method</span></header><section><div><p>There&#39;s an hydrophobic interaction if two hydrophobic atoms are at 5.0 Å or less.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Interaction.jl#L201-L203">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.ionic-Tuple{MIToS.PDB.PDBAtom, MIToS.PDB.PDBAtom, Any, Any}" href="#MIToS.PDB.ionic-Tuple{MIToS.PDB.PDBAtom, MIToS.PDB.PDBAtom, Any, Any}"><code>MIToS.PDB.ionic</code></a> — <span class="docstring-category">Method</span></header><section><div><p>There&#39;s an ionic interaction if a cationic and an anionic atoms are at 6.0 Å or less.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Interaction.jl#L183-L185">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.is_aminoacid-Tuple{MIToS.PDB.PDBResidue}" href="#MIToS.PDB.is_aminoacid-Tuple{MIToS.PDB.PDBResidue}"><code>MIToS.PDB.is_aminoacid</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">is_aminoacid(residue::PDBResidue)
is_aminoacid(residue_id::PDBResidueIdentifier)</code></pre><p>This function returns <code>true</code> if the PDB residue is an amino acid residue. It checks if the  residue&#39;s three-letter name exists in the <code>MIToS.Utils.THREE2ONE</code> dictionary, and  returns <code>false</code> otherwise.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Sequences.jl#L5-L12">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.isanionic-Tuple{MIToS.PDB.PDBAtom, String}" href="#MIToS.PDB.isanionic-Tuple{MIToS.PDB.PDBAtom, String}"><code>MIToS.PDB.isanionic</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Returns true if the atom, e.g. <code>(&quot;GLU&quot;,&quot;CD&quot;)</code>, is an anionic atom in the residue.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Interaction.jl#L17-L19">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.isaromatic-Tuple{MIToS.PDB.PDBAtom, String}" href="#MIToS.PDB.isaromatic-Tuple{MIToS.PDB.PDBAtom, String}"><code>MIToS.PDB.isaromatic</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Returns true if the atom, e.g. <code>(&quot;HIS&quot;,&quot;CG&quot;)</code>, is an aromatic atom in the residue.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Interaction.jl#L7-L9">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.isatom-Tuple{MIToS.PDB.PDBAtom, Any}" href="#MIToS.PDB.isatom-Tuple{MIToS.PDB.PDBAtom, Any}"><code>MIToS.PDB.isatom</code></a> — <span class="docstring-category">Method</span></header><section><div><p>It tests if the atom has the indicated atom name.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L150-L152">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.iscationic-Tuple{MIToS.PDB.PDBAtom, String}" href="#MIToS.PDB.iscationic-Tuple{MIToS.PDB.PDBAtom, String}"><code>MIToS.PDB.iscationic</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Returns true if the atom, e.g. <code>(&quot;ARG&quot;,&quot;NE&quot;)</code>, is a cationic atom in the residue.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Interaction.jl#L12-L14">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.ishbondacceptor-Tuple{MIToS.PDB.PDBAtom, String}" href="#MIToS.PDB.ishbondacceptor-Tuple{MIToS.PDB.PDBAtom, String}"><code>MIToS.PDB.ishbondacceptor</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Returns true if the atom, e.g. <code>(&quot;ARG&quot;,&quot;O&quot;)</code>, is an acceptor in H bonds.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Interaction.jl#L27-L29">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.ishbonddonor-Tuple{MIToS.PDB.PDBAtom, String}" href="#MIToS.PDB.ishbonddonor-Tuple{MIToS.PDB.PDBAtom, String}"><code>MIToS.PDB.ishbonddonor</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Returns true if the atom, e.g. <code>(&quot;ARG&quot;,&quot;N&quot;)</code>, is a donor in H bonds.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Interaction.jl#L22-L24">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.isresidue-Tuple{MIToS.PDB.PDBResidueIdentifier, Vararg{Any, 4}}" href="#MIToS.PDB.isresidue-Tuple{MIToS.PDB.PDBResidueIdentifier, Vararg{Any, 4}}"><code>MIToS.PDB.isresidue</code></a> — <span class="docstring-category">Method</span></header><section><div><p>It tests if the PDB residue has the indicated model, chain, group and residue number.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L139-L141">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.kabsch-Tuple{AbstractMatrix{Float64}, AbstractMatrix{Float64}}" href="#MIToS.PDB.kabsch-Tuple{AbstractMatrix{Float64}, AbstractMatrix{Float64}}"><code>MIToS.PDB.kabsch</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>kabsch(A::AbstractMatrix{Float64}, B::AbstractMatrix{Float64})</code></p><p>This function takes two sets of points, <code>A</code> (refrence) and <code>B</code> as NxD matrices, where D is the dimension and N is the number of points. Assumes that the centroids of <code>A</code> and <code>B</code> are at the origin of coordinates. You can call <code>center!</code> on each matrix before calling <code>kabsch</code> to center the matrices in the <code>(0.0, 0.0, 0.0)</code>. Rotates <code>B</code> so that <code>rmsd(A,B)</code> is minimized. Returns the rotation matrix. You should do <code>B * RotationMatrix</code> to get the rotated B.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Kabsch.jl#L3-L13">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.mean_coordinates-Union{Tuple{AbstractVector{T}}, Tuple{T}} where T&lt;:AbstractMatrix{Float64}" href="#MIToS.PDB.mean_coordinates-Union{Tuple{AbstractVector{T}}, Tuple{T}} where T&lt;:AbstractMatrix{Float64}"><code>MIToS.PDB.mean_coordinates</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Calculates the average/mean position of each atom in a set of structure. The function takes a vector (<code>AbstractVector</code>) of vectors (<code>AbstractVector{PDBResidue}</code>) or matrices (<code>AbstractMatrix{Float64}</code>) as first argument. As second (optional) argument this function can take an <code>AbstractVector{Float64}</code> of matrix/structure weights to return a weighted mean. When a AbstractVector{PDBResidue} is used, if the keyword argument <code>calpha</code> is <code>false</code> the RMSF is calculated for all the atoms. By default only alpha carbons are used (default: <code>calpha=true</code>).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Kabsch.jl#L338-L346">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.modelled_sequences-Union{Tuple{AbstractArray{MIToS.PDB.PDBResidue, N}}, Tuple{N}} where N" href="#MIToS.PDB.modelled_sequences-Union{Tuple{AbstractArray{MIToS.PDB.PDBResidue, N}}, Tuple{N}} where N"><code>MIToS.PDB.modelled_sequences</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">modelled_sequences(residue_list::AbstractArray{PDBResidue,N}; 
    model::Union{String,Type{All}}=All, chain::Union{String,Type{All}}=All, 
    group::Union{String,Regex,Type{All}}=All) where N</code></pre><p>This function returns an <code>OrderedDict</code> where each key is a named tuple (containing the  model and chain identifiers), and each value is the protein sequence corresponding to  the modelled residues in those chains. Therefore, the obtained sequences do not contain  missing residues. All modelled residues are included by default, but those that don&#39;t  satisfy specified criteria based on the <code>model</code>, <code>chain</code>, or <code>group</code> keyword arguments  are excluded. One-letter residue names are obtained from the <code>MIToS.Utils.THREE2ONE</code>  dictionary for all residue names that return <code>true</code> for <code>is_aminoacid</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Sequences.jl#L28-L40">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.pication-Tuple{MIToS.PDB.PDBAtom, MIToS.PDB.PDBAtom, Any, Any}" href="#MIToS.PDB.pication-Tuple{MIToS.PDB.PDBAtom, MIToS.PDB.PDBAtom, Any, Any}"><code>MIToS.PDB.pication</code></a> — <span class="docstring-category">Method</span></header><section><div><p>There&#39;s a Π-Cation interaction if a cationic and an aromatic atoms are at 6.0 Å or less</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Interaction.jl#L147-L149">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.proximitymean-Union{Tuple{T}, Tuple{Vector{MIToS.PDB.PDBResidue}, AbstractVector{T}}, Tuple{Vector{MIToS.PDB.PDBResidue}, AbstractVector{T}, T}} where T&lt;:AbstractFloat" href="#MIToS.PDB.proximitymean-Union{Tuple{T}, Tuple{Vector{MIToS.PDB.PDBResidue}, AbstractVector{T}}, Tuple{Vector{MIToS.PDB.PDBResidue}, AbstractVector{T}, T}} where T&lt;:AbstractFloat"><code>MIToS.PDB.proximitymean</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>proximitymean</code> calculates the proximity mean/average for each residue as the average score (from a <code>scores</code> list) of all the residues within a certain physical distance to a given amino acid. The score of that residue is not included in the mean unless you set <code>include</code> to <code>true</code>. The default values are 6.05 for the distance threshold/<code>limit</code> and <code>&quot;Heavy&quot;</code> for the <code>criteria</code> keyword argument. This function allows to calculate pMI (proximity mutual information) and pC (proximity conservation) as in Buslje et. al. 2010.</p><p>Buslje, Cristina Marino, Elin Teppa, Tomas Di Doménico, José María Delfino, and Morten Nielsen. <em>Networks of high mutual information define the structural proximity of catalytic sites: implications for catalytic residue identification.</em> PLoS Comput Biol 6, no. 11 (2010): e1000978.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L602-L613">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.query_alphafolddb-Tuple{String}" href="#MIToS.PDB.query_alphafolddb-Tuple{String}"><code>MIToS.PDB.query_alphafolddb</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">query_alphafolddb(uniprot_id::String)</code></pre><p>This function queries the AlphaFoldDB API to retrieve structure information for  a given <code>uniprot_id</code>. This function returns the structure information as a <code>JSON3.Object</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/AlphaFoldDB.jl#L1-L7">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.residuepairsmatrix-Union{Tuple{diagonal}, Tuple{T}, Tuple{Vector{MIToS.PDB.PDBResidue}, Type{T}, Type{Val{diagonal}}, T}} where {T, diagonal}" href="#MIToS.PDB.residuepairsmatrix-Union{Tuple{diagonal}, Tuple{T}, Tuple{Vector{MIToS.PDB.PDBResidue}, Type{T}, Type{Val{diagonal}}, T}} where {T, diagonal}"><code>MIToS.PDB.residuepairsmatrix</code></a> — <span class="docstring-category">Method</span></header><section><div><p>It creates a <code>NamedArray</code> containing a <code>PairwiseListMatrix</code> where each element (column, row) is identified with a <code>PDBResidue</code> from the input vector. You can indicate the value type of the matrix (default to <code>Float64</code>), if the list should have the diagonal values (default to <code>Val{false}</code>) and the diagonal values (default to <code>NaN</code>).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L533-L538">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.residues-Union{Tuple{AbstractArray{MIToS.PDB.PDBResidue, N}}, Tuple{N}, Tuple{AbstractArray{MIToS.PDB.PDBResidue, N}, Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue, N}, Any, Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue, N}, Any, Any, Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue, N}, Vararg{Any, 4}}} where N" href="#MIToS.PDB.residues-Union{Tuple{AbstractArray{MIToS.PDB.PDBResidue, N}}, Tuple{N}, Tuple{AbstractArray{MIToS.PDB.PDBResidue, N}, Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue, N}, Any, Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue, N}, Any, Any, Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue, N}, Vararg{Any, 4}}} where N"><code>MIToS.PDB.residues</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>residues(residue_list, model, chain, group, residue)</code></p><p>These return a new vector with the selected subset of residues from a list of residues. You can use the type <code>All</code> (default value) to avoid filtering a that level.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L158-L163">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.residuesdict-Union{Tuple{AbstractArray{MIToS.PDB.PDBResidue, N}}, Tuple{N}, Tuple{AbstractArray{MIToS.PDB.PDBResidue, N}, Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue, N}, Any, Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue, N}, Any, Any, Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue, N}, Vararg{Any, 4}}} where N" href="#MIToS.PDB.residuesdict-Union{Tuple{AbstractArray{MIToS.PDB.PDBResidue, N}}, Tuple{N}, Tuple{AbstractArray{MIToS.PDB.PDBResidue, N}, Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue, N}, Any, Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue, N}, Any, Any, Any}, Tuple{AbstractArray{MIToS.PDB.PDBResidue, N}, Vararg{Any, 4}}} where N"><code>MIToS.PDB.residuesdict</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>residuesdict(residue_list, model, chain, group, residue)</code></p><p>These return a dictionary (using PDB residue numbers as keys) with the selected subset of residues. You can use the type <code>All</code> (default value) to avoid filtering a that level.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L189-L194">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.rmsd-Tuple{AbstractMatrix{Float64}, AbstractMatrix{Float64}}" href="#MIToS.PDB.rmsd-Tuple{AbstractMatrix{Float64}, AbstractMatrix{Float64}}"><code>MIToS.PDB.rmsd</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>rmsd(A::AbstractMatrix{Float64}, B::AbstractMatrix{Float64})</code></p><p>Return RMSD between two sets of points <code>A</code> and <code>B</code>, given as NxD matrices (N: number of points, D: dimension).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Kabsch.jl#L35-L40">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.rmsd-Tuple{AbstractVector{MIToS.PDB.PDBResidue}, AbstractVector{MIToS.PDB.PDBResidue}}" href="#MIToS.PDB.rmsd-Tuple{AbstractVector{MIToS.PDB.PDBResidue}, AbstractVector{MIToS.PDB.PDBResidue}}"><code>MIToS.PDB.rmsd</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>rmsd(A::AbstractVector{PDBResidue}, B::AbstractVector{PDBResidue}; superimposed::Bool=false)</code></p><p>Returns the Cα RMSD value between two PDB structures: <code>A</code> and <code>B</code>. If the structures are already superimposed between them, use <code>superimposed=true</code> to avoid a new superimposition (<code>superimposed</code> is <code>false</code> by default).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Kabsch.jl#L51-L57">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.rmsf-Union{Tuple{AbstractVector{T}}, Tuple{T}} where T&lt;:AbstractMatrix{Float64}" href="#MIToS.PDB.rmsf-Union{Tuple{AbstractVector{T}}, Tuple{T}} where T&lt;:AbstractMatrix{Float64}"><code>MIToS.PDB.rmsf</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Calculates the RMSF (Root Mean-Square-Fluctuation) between an atom and its average position in a set of structures. The function takes a vector (<code>AbstractVector</code>) of vectors (<code>AbstractVector{PDBResidue}</code>) or matrices (<code>AbstractMatrix{Float64}</code>) as first argument. As second (optional) argument this function can take an <code>AbstractVector{Float64}</code> of matrix/structure weights to return the root weighted mean-square-fluctuation around the weighted mean structure. When a Vector{PDBResidue} is used, if the keyword argument <code>calpha</code> is <code>false</code> the RMSF is calculated for all the atoms. By default only alpha carbons are used (default: <code>calpha=true</code>).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Kabsch.jl#L371-L380">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.selectbestoccupancy-Tuple{Vector{MIToS.PDB.PDBAtom}, Vector{Int64}}" href="#MIToS.PDB.selectbestoccupancy-Tuple{Vector{MIToS.PDB.PDBAtom}, Vector{Int64}}"><code>MIToS.PDB.selectbestoccupancy</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Takes a <code>PDBResidue</code> and a <code>Vector</code> of atom indices. Returns the index value of the <code>Vector</code> with maximum occupancy.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L319-L322">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.squared_distance-Tuple{MIToS.PDB.PDBAtom, MIToS.PDB.PDBAtom}" href="#MIToS.PDB.squared_distance-Tuple{MIToS.PDB.PDBAtom, MIToS.PDB.PDBAtom}"><code>MIToS.PDB.squared_distance</code></a> — <span class="docstring-category">Method</span></header><section><div><p>It calculates the squared euclidean distance, i.e. it doesn&#39;t spend time in <code>sqrt</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L79">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.squared_distance-Tuple{MIToS.PDB.PDBResidue, MIToS.PDB.PDBResidue}" href="#MIToS.PDB.squared_distance-Tuple{MIToS.PDB.PDBResidue, MIToS.PDB.PDBResidue}"><code>MIToS.PDB.squared_distance</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>squared_distance(A::PDBResidue, B::PDBResidue; criteria::String=&quot;All&quot;)</code></p><p>Returns the squared distance between the residues <code>A</code> and <code>B</code>. The available <code>criteria</code> are: <code>Heavy</code>, <code>All</code>, <code>CA</code>, <code>CB</code> (<code>CA</code> for <code>GLY</code>)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/PDBResidues.jl#L389-L394">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.superimpose" href="#MIToS.PDB.superimpose"><code>MIToS.PDB.superimpose</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">Asuper, Bsuper, RMSD = superimpose(A, B, matches=nothing)</code></pre><p>This function takes <code>A::AbstractVector{PDBResidue}</code> (reference) and <code>B::AbstractVector{PDBResidue}</code>. Translates <code>A</code> and <code>B</code> to the origin of coordinates, and rotates <code>B</code> so that <code>rmsd(A,B)</code> is minimized with the Kabsch algorithm (using only their α carbons). Returns the rotated and translated versions of <code>A</code> and <code>B</code>, and the RMSD value.</p><p>Optionally provide <code>matches</code> which iterates over matched index pairs in <code>A</code> and <code>B</code>, e.g., <code>matches = [(3, 5), (4, 6), ...]</code>. The alignment will be constructed using just the matching residues.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Kabsch.jl#L270-L282">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.vanderwaals-Tuple{MIToS.PDB.PDBAtom, MIToS.PDB.PDBAtom, Any, Any}" href="#MIToS.PDB.vanderwaals-Tuple{MIToS.PDB.PDBAtom, MIToS.PDB.PDBAtom, Any, Any}"><code>MIToS.PDB.vanderwaals</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Test if two atoms or residues are in van der Waals contact using: <code>distance(a,b) &lt;= 0.5 + vanderwaalsradius[a] + vanderwaalsradius[b]</code>. It returns distance <code>&lt;= 0.5</code> if the atoms aren&#39;t in <code>vanderwaalsradius</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Interaction.jl#L62-L66">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.PDB.vanderwaalsclash-Tuple{MIToS.PDB.PDBAtom, MIToS.PDB.PDBAtom, Any, Any}" href="#MIToS.PDB.vanderwaalsclash-Tuple{MIToS.PDB.PDBAtom, MIToS.PDB.PDBAtom, Any, Any}"><code>MIToS.PDB.vanderwaalsclash</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Returns <code>true</code> if the distance between the atoms is less than the sum of the <code>vanderwaalsradius</code> of the atoms. If the atoms aren&#39;t on the list (i.e. <code>OXT</code>), the <code>vanderwaalsradius</code> of the element is used. If there is not data in the dict, distance <code>0.0</code> is used.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/24eca41cc049f426926a28f72c995ec2d31cae8a/src/PDB/Interaction.jl#L78-L83">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../SIFTS_API/">« SIFTS</a><a class="docs-footer-nextpage" href="../Pfam_API/">Pfam »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.4.1 on <span class="colophon-date" title="Tuesday 11 June 2024 09:48">Tuesday 11 June 2024</span>. Using Julia version 1.10.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
