<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>MSA · MIToS</title><meta name="title" content="MSA · MIToS"/><meta property="og:title" content="MSA · MIToS"/><meta property="twitter:title" content="MSA · MIToS"/><meta name="description" content="Documentation for MIToS."/><meta property="og:description" content="Documentation for MIToS."/><meta property="twitter:description" content="Documentation for MIToS."/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">MIToS</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../Installation/">Installation</a></li><li><a class="tocitem" href="../Example/">Example</a></li><li><span class="tocitem">Modules</span><ul><li><a class="tocitem" href="../MSA/">MSA</a></li><li><a class="tocitem" href="../Information/">Information</a></li><li><a class="tocitem" href="../SIFTS/">SIFTS</a></li><li><a class="tocitem" href="../PDB/">PDB</a></li><li><a class="tocitem" href="../Pfam/">Pfam</a></li></ul></li><li><span class="tocitem">Cookbook</span><ul><li><a class="tocitem" href="../01_Change_B_factors/">Change B-factors</a></li><li><a class="tocitem" href="../02_Linking_structural_and_evolutionary_information/">Linking structural and evolutionary information</a></li><li><a class="tocitem" href="../03_RMSF/">Root Mean Squared Fluctuation (RMSF)</a></li></ul></li><li><span class="tocitem">API</span><ul><li class="is-active"><a class="tocitem" href>MSA</a><ul class="internal"><li><a class="tocitem" href="#Contents"><span>Contents</span></a></li><li><a class="tocitem" href="#Types"><span>Types</span></a></li><li><a class="tocitem" href="#Constants"><span>Constants</span></a></li><li><a class="tocitem" href="#Macros"><span>Macros</span></a></li><li><a class="tocitem" href="#Methods-and-functions"><span>Methods and functions</span></a></li></ul></li><li><a class="tocitem" href="../Information_API/">Information</a></li><li><a class="tocitem" href="../SIFTS_API/">SIFTS</a></li><li><a class="tocitem" href="../PDB_API/">PDB</a></li><li><a class="tocitem" href="../Pfam_API/">Pfam</a></li><li><a class="tocitem" href="../Utils_API/">Utils</a></li></ul></li><li><a class="tocitem" href="../Scripts/">Scripts</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">API</a></li><li class="is-active"><a href>MSA</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>MSA</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/diegozea/MIToS.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/diegozea/MIToS.jl/blob/master/docs/src/MSA_API.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="MSA"><a class="docs-heading-anchor" href="#MSA">MSA</a><a id="MSA-1"></a><a class="docs-heading-anchor-permalink" href="#MSA" title="Permalink"></a></h1><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA" href="#MIToS.MSA"><code>MIToS.MSA</code></a> — <span class="docstring-category">Module</span></header><section><div><p>The MSA module of MIToS has utilities for working with Multiple Sequence Alignments of protein Sequences (MSA).</p><p><strong>Features</strong></p><ul><li>Read and write MSAs in <code>Stockholm</code>, <code>FASTA</code>, <code>A3M</code>, <code>PIR</code>, or <code>Raw</code> format</li><li>Handle MSA annotations</li><li>Edit the MSA, e.g. delete columns or sequences, change sequence order, shuffling...</li><li>Keep track of positions and annotations after modifications on the MSA</li><li>Describe a MSA, e.g. mean percent identity, sequence coverage, gap percentage...</li></ul><pre><code class="language-julia hljs">using MIToS.MSA</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MSA.jl#L1-L15">source</a></section></article><h2 id="Contents"><a class="docs-heading-anchor" href="#Contents">Contents</a><a id="Contents-1"></a><a class="docs-heading-anchor-permalink" href="#Contents" title="Permalink"></a></h2><ul><li><a href="#MSA">MSA</a></li><li class="no-marker"><ul><li><a href="#Contents">Contents</a></li><li><a href="#Types">Types</a></li><li><a href="#Constants">Constants</a></li><li><a href="#Macros">Macros</a></li><li><a href="#Methods-and-functions">Methods and functions</a></li></ul></li></ul><h2 id="Types"><a class="docs-heading-anchor" href="#Types">Types</a><a id="Types-1"></a><a class="docs-heading-anchor-permalink" href="#Types" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.AbstractAlignedObject" href="#MIToS.MSA.AbstractAlignedObject"><code>MIToS.MSA.AbstractAlignedObject</code></a> — <span class="docstring-category">Type</span></header><section><div><p>MIToS MSA and aligned sequences (aligned objects) are subtypes of <code>AbstractMatrix{Residue}</code>, because MSAs and sequences are stored as <code>Matrix</code> of <code>Residue</code>s.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MultipleSequenceAlignment.jl#L6-L9">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.AbstractAlignedSequence" href="#MIToS.MSA.AbstractAlignedSequence"><code>MIToS.MSA.AbstractAlignedSequence</code></a> — <span class="docstring-category">Type</span></header><section><div><p>A MIToS aligned sequence is an <code>AbstractMatrix{Residue}</code> with only 1 row/sequence.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MultipleSequenceAlignment.jl#L19-L21">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.AbstractMultipleSequenceAlignment" href="#MIToS.MSA.AbstractMultipleSequenceAlignment"><code>MIToS.MSA.AbstractMultipleSequenceAlignment</code></a> — <span class="docstring-category">Type</span></header><section><div><p>MSAs are stored as <code>Matrix{Residue}</code>. It&#39;s possible to use a <code>NamedResidueMatrix{Array{Residue,2}}</code> as the most simple MSA with sequence identifiers and column names.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MultipleSequenceAlignment.jl#L12-L16">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.AbstractSequence" href="#MIToS.MSA.AbstractSequence"><code>MIToS.MSA.AbstractSequence</code></a> — <span class="docstring-category">Type</span></header><section><div><p>A MIToS (unaligned) sequence is an <code>AbstractMatrix{Residue}</code> with only 1 row/sequence.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MultipleSequenceAlignment.jl#L24-L26">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.AlignedSequence" href="#MIToS.MSA.AlignedSequence"><code>MIToS.MSA.AlignedSequence</code></a> — <span class="docstring-category">Type</span></header><section><div><p>An <code>AlignedSequence</code> wraps a <code>NamedResidueMatrix{Array{Residue,2}}</code> with only 1 row/sequence. The <code>NamedArray</code> stores the sequence name and original column numbers as <code>String</code>s.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MultipleSequenceAlignment.jl#L94-L97">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.AnnotatedAlignedSequence" href="#MIToS.MSA.AnnotatedAlignedSequence"><code>MIToS.MSA.AnnotatedAlignedSequence</code></a> — <span class="docstring-category">Type</span></header><section><div><p>This type represent an aligned sequence, similar to <code>AlignedSequence</code>, but It also stores its <code>Annotations</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MultipleSequenceAlignment.jl#L108-L111">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.AnnotatedMultipleSequenceAlignment" href="#MIToS.MSA.AnnotatedMultipleSequenceAlignment"><code>MIToS.MSA.AnnotatedMultipleSequenceAlignment</code></a> — <span class="docstring-category">Type</span></header><section><div><p>This type represent an MSA, similar to <code>MultipleSequenceAlignment</code>, but It also stores <code>Annotations</code>. This annotations are used to store residue coordinates (i.e. mapping to UniProt residue numbers).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MultipleSequenceAlignment.jl#L51-L55">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.AnnotatedSequence" href="#MIToS.MSA.AnnotatedSequence"><code>MIToS.MSA.AnnotatedSequence</code></a> — <span class="docstring-category">Type</span></header><section><div><p>An <code>AnnotationSequence</code> wraps a <code>NamedResidueMatrix{Array{Residue,2}}</code> with only 1 row/sequence and its <code>Annotations</code>. The <code>NamedArray</code> stores the sequence name and original position numbers as <code>String</code>s.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MultipleSequenceAlignment.jl#L129-L133">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.Annotations" href="#MIToS.MSA.Annotations"><code>MIToS.MSA.Annotations</code></a> — <span class="docstring-category">Type</span></header><section><div><p>The <code>Annotations</code> type is basically a container for <code>Dict</code>s with the annotations of a multiple sequence alignment. <code>Annotations</code> was designed for storage of annotations of the <strong>Stockholm format</strong>.</p><p>MIToS also uses MSA annotations to keep track of:</p><ul><li><strong>Modifications</strong> of the MSA (<code>MIToS_...</code>) as deletion of sequences or columns.</li><li>Positions numbers in the original MSA file (<strong>column mapping:</strong> <code>ColMap</code>)</li><li>Position of the residues in the sequence (<strong>sequence mapping:</strong> <code>SeqMap</code>)</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Annotations.jl#L4-L14">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.Clusters" href="#MIToS.MSA.Clusters"><code>MIToS.MSA.Clusters</code></a> — <span class="docstring-category">Type</span></header><section><div><p>Data structure to represent sequence clusters. The sequence data itself is not included.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Clusters.jl#L11-L13">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.GappedAlphabet" href="#MIToS.MSA.GappedAlphabet"><code>MIToS.MSA.GappedAlphabet</code></a> — <span class="docstring-category">Type</span></header><section><div><p>This type defines the usual alphabet of the 20 natural residues and a gap character.</p><pre><code class="language-julia-repl hljs">julia&gt; using MIToS.MSA

julia&gt; GappedAlphabet()
GappedAlphabet of length 21. Residues : res&quot;ARNDCQEGHILKMFPSTWYV-&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Alphabet.jl#L6-L15">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.MultipleSequenceAlignment" href="#MIToS.MSA.MultipleSequenceAlignment"><code>MIToS.MSA.MultipleSequenceAlignment</code></a> — <span class="docstring-category">Type</span></header><section><div><p>This MSA type include a <code>NamedArray</code> wrapping a <code>Matrix</code> of <code>Residue</code>s. The use of <code>NamedArray</code> allows to store sequence names and original column numbers as <code>String</code>s, and fast indexing using them.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MultipleSequenceAlignment.jl#L35-L39">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.NoClustering" href="#MIToS.MSA.NoClustering"><code>MIToS.MSA.NoClustering</code></a> — <span class="docstring-category">Type</span></header><section><div><p>Use <code>NoClustering()</code> to avoid the use of clustering where a <code>Clusters</code> type is needed.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Clusters.jl#L6-L8">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.ReducedAlphabet" href="#MIToS.MSA.ReducedAlphabet"><code>MIToS.MSA.ReducedAlphabet</code></a> — <span class="docstring-category">Type</span></header><section><div><p><code>ReducedAlphabet</code> allows the construction of reduced residue alphabets, where residues inside parenthesis belong to the same group.</p><pre><code class="language-julia-repl hljs">julia&gt; using MIToS.MSA

julia&gt; ab = ReducedAlphabet(&quot;(AILMV)(RHK)(NQST)(DE)(FWY)CGP&quot;)
ReducedAlphabet of length 8 : &quot;(AILMV)(RHK)(NQST)(DE)(FWY)CGP&quot;

julia&gt; ab[Residue(&#39;K&#39;)]
2</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Alphabet.jl#L30-L43">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.Residue" href="#MIToS.MSA.Residue"><code>MIToS.MSA.Residue</code></a> — <span class="docstring-category">Type</span></header><section><div><p>Most of the <strong>MIToS</strong> design is created around the <code>Residue</code> bitstype. It has representations for the 20 natural amino acids, a value representing insertions and deletions (<code>GAP</code>, <code>&#39;-&#39;</code>) and one representing unknown, ambiguous and non standard residues (<code>XAA</code>, <code>&#39;X&#39;</code>). Each <code>Residue</code> is encoded as an integer number, with the same bit representation and size than a <code>Int</code>. This allows fast indexing operation of probability or frequency matrices.</p><p><strong>Residue creation and conversion</strong></p><p>Creation and conversion of <code>Residue</code>s should be treated carefully. <code>Residue</code> is encoded as a 32 or 64 bits type similar to <code>Int</code>, to get fast indexing using <code>Int(x::Residue)</code>. <code>Int</code> simply calls <code>reinterpret</code> without checking if the residue is valid. Valid residues have integer values in the closed interval [1,22]. <code>convert</code> from <code>Int</code>  and <code>Char</code> always returns valid residues, however it&#39;s possible to find invalid residues (they are shown using the character <code>&#39;�&#39;</code>) after the creation of uninitialized <code>Residue</code> arrays (i.e. using <code>Array</code>). You can use <code>zeros</code>, <code>ones</code> or <code>rand</code> to get initialized <code>Residue</code> arrays with valid residues. Conversions to and from <code>Char</code>s changes the bit representation and allows the use of the usual character representation of residues and amino acids. This conversions are used in IO operations and always return valid residues. In conversions from <code>Char</code>, lowercase letters, <code>&#39;*&#39;</code>, <code>&#39;-&#39;</code> and <code>&#39;.&#39;</code> are translated to <code>GAP</code>, letters representing the 20 natural amino (ARNDCQEGHILKMFPSTWYV) acids are translated to their corresponding <code>Residue</code> and any other character is translated to <code>XAA</code>. Since lowercase letters and dots are translated to gaps, Pfam MSA insert columns are converted to columns full of gaps.</p><pre><code class="language-julia-repl hljs">julia&gt; using MIToS.MSA

julia&gt; alanine = Residue(&#39;A&#39;)
A

julia&gt; Char(alanine)
&#39;A&#39;: ASCII/Unicode U+0041 (category Lu: Letter, uppercase)

julia&gt; for residue in res&quot;ARNDCQEGHILKMFPSTWYV-X&quot;
           println(residue, &quot; &quot;, Int(residue))
       end
A 1
R 2
N 3
D 4
C 5
Q 6
E 7
G 8
H 9
I 10
L 11
K 12
M 13
F 14
P 15
S 16
T 17
W 18
Y 19
V 20
- 21
X 22
</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Residues.jl#L10-L72">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.ResidueAlphabet" href="#MIToS.MSA.ResidueAlphabet"><code>MIToS.MSA.ResidueAlphabet</code></a> — <span class="docstring-category">Type</span></header><section><div><p>Abstract type to define residue alphabet types.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Alphabet.jl#L1-L3">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.UngappedAlphabet" href="#MIToS.MSA.UngappedAlphabet"><code>MIToS.MSA.UngappedAlphabet</code></a> — <span class="docstring-category">Type</span></header><section><div><p>This type defines the usual alphabet of the 20 natural residues, without the gap character.</p><pre><code class="language-julia-repl hljs">julia&gt; using MIToS.MSA

julia&gt; UngappedAlphabet()
UngappedAlphabet of length 20. Residues : res&quot;ARNDCQEGHILKMFPSTWYV&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Alphabet.jl#L18-L27">source</a></section></article><h2 id="Constants"><a class="docs-heading-anchor" href="#Constants">Constants</a><a id="Constants-1"></a><a class="docs-heading-anchor-permalink" href="#Constants" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.GAP" href="#MIToS.MSA.GAP"><code>MIToS.MSA.GAP</code></a> — <span class="docstring-category">Constant</span></header><section><div><p><code>GAP</code> is the <code>Residue</code> representation on MIToS for gaps (<code>&#39;-&#39;</code>, insertions and deletions). Lowercase residue characters, dots and <code>&#39;*&#39;</code> are encoded as <code>GAP</code> in conversion from <code>String</code>s and <code>Char</code>s. This <code>Residue</code> constant is encoded as <code>Residue(21)</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Residues.jl#L108-L112">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.WeightTypes" href="#MIToS.MSA.WeightTypes"><code>MIToS.MSA.WeightTypes</code></a> — <span class="docstring-category">Type</span></header><section><div><p>The <code>WeightTypes</code> type is the same as <code>Union{Weights,NoClustering,Clusters}</code>. This type is used to represent weights. Most of the functions taking the <code>weights</code> kerword argument in the <code>Information</code> module accept instances of <code>WeightTypes</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Clusters.jl#L27-L31">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.XAA" href="#MIToS.MSA.XAA"><code>MIToS.MSA.XAA</code></a> — <span class="docstring-category">Constant</span></header><section><div><p><code>XAA</code> is the <code>Residue</code> representation for unknown, ambiguous and non standard residues. This <code>Residue</code> constant is encoded as <code>Residue(22)</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Residues.jl#L115-L118">source</a></section></article><h2 id="Macros"><a class="docs-heading-anchor" href="#Macros">Macros</a><a id="Macros-1"></a><a class="docs-heading-anchor-permalink" href="#Macros" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.@res_str-Tuple{Any}" href="#MIToS.MSA.@res_str-Tuple{Any}"><code>MIToS.MSA.@res_str</code></a> — <span class="docstring-category">Macro</span></header><section><div><p>The MIToS macro <code>@res_str</code> takes a string and returns a <code>Vector</code> of <code>Residues</code> (sequence).</p><pre><code class="language-julia-repl hljs">julia&gt; using MIToS.MSA

julia&gt; res&quot;MIToS&quot;
5-element Vector{Residue}:
 M
 I
 T
 -
 S</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Residues.jl#L310-L324">source</a></section></article><h2 id="Methods-and-functions"><a class="docs-heading-anchor" href="#Methods-and-functions">Methods and functions</a><a id="Methods-and-functions-1"></a><a class="docs-heading-anchor-permalink" href="#Methods-and-functions" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.isvalid-Tuple{Type{MIToS.MSA.Residue}, MIToS.MSA.Residue}" href="#Base.isvalid-Tuple{Type{MIToS.MSA.Residue}, MIToS.MSA.Residue}"><code>Base.isvalid</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>isvalid(res::Residue)</code></p><p>It returns <code>true</code> if the encoded integer is in the closed interval [1,22].</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Residues.jl#L124-L128">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.names-Tuple{MIToS.MSA.ReducedAlphabet}" href="#Base.names-Tuple{MIToS.MSA.ReducedAlphabet}"><code>Base.names</code></a> — <span class="docstring-category">Method</span></header><section><div><p>It returns the name of each group. The name is a string with the one letter code of each residue that belong to the group.</p><pre><code class="language-julia-repl hljs">julia&gt; using MIToS.MSA

julia&gt; ab = ReducedAlphabet(&quot;(AILMV)(RHK)(NQST)(DE)(FWY)CGP&quot;)
ReducedAlphabet of length 8 : &quot;(AILMV)(RHK)(NQST)(DE)(FWY)CGP&quot;

julia&gt; names(ab)
8-element Vector{String}:
 &quot;AILMV&quot;
 &quot;RHK&quot;
 &quot;NQST&quot;
 &quot;DE&quot;
 &quot;FWY&quot;
 &quot;C&quot;
 &quot;G&quot;
 &quot;P&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Alphabet.jl#L169-L190">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.rand-Tuple{Random.AbstractRNG, Random.SamplerType{MIToS.MSA.Residue}}" href="#Base.rand-Tuple{Random.AbstractRNG, Random.SamplerType{MIToS.MSA.Residue}}"><code>Base.rand</code></a> — <span class="docstring-category">Method</span></header><section><div><p>It chooses from the 20 natural residues (it doesn&#39;t generate gaps).</p><pre><code class="language-julia-repl hljs">julia&gt; using MIToS.MSA

julia&gt; using Random

julia&gt; Random.seed!(1); # Reseed the random number generator.

julia&gt; rand(Residue)
R

julia&gt; rand(Residue, 4, 4)
4×4 Matrix{Residue}:
 E  D  D  A
 F  S  K  K
 M  S  I  M
 Y  F  E  D</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Residues.jl#L340-L360">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.adjustreference" href="#MIToS.MSA.adjustreference"><code>MIToS.MSA.adjustreference</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Creates a new matrix of residues. This function deletes positions/columns of the MSA with gaps in the reference (first) sequence.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MSAEditing.jl#L290-L293">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.adjustreference!" href="#MIToS.MSA.adjustreference!"><code>MIToS.MSA.adjustreference!</code></a> — <span class="docstring-category">Function</span></header><section><div><p>It removes positions/columns of the MSA with gaps in the reference (first) sequence.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MSAEditing.jl#L299-L301">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.annotate_modification!-Tuple{MIToS.MSA.Annotations, String}" href="#MIToS.MSA.annotate_modification!-Tuple{MIToS.MSA.Annotations, String}"><code>MIToS.MSA.annotate_modification!</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Annotates on file annotations the modifications realized by MIToS on the MSA. It always returns <code>true</code>, so It can be used in a boolean context.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Annotations.jl#L350-L353">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.annotations-Tuple{MIToS.MSA.AnnotatedMultipleSequenceAlignment}" href="#MIToS.MSA.annotations-Tuple{MIToS.MSA.AnnotatedMultipleSequenceAlignment}"><code>MIToS.MSA.annotations</code></a> — <span class="docstring-category">Method</span></header><section><div><p>The <code>annotations</code> function returns the <code>Annotations</code> of an annotated MSA or aligned sequence. If the object is not annotated, it returns an empty <code>Annotations</code> object.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MultipleSequenceAlignment.jl#L298-L301">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.column_index-Tuple{NamedArrays.NamedMatrix{MIToS.MSA.Residue, AT, Tuple{OrderedCollections.OrderedDict{String, Int64}, OrderedCollections.OrderedDict{String, Int64}}} where AT, AbstractString}" href="#MIToS.MSA.column_index-Tuple{NamedArrays.NamedMatrix{MIToS.MSA.Residue, AT, Tuple{OrderedCollections.OrderedDict{String, Int64}, OrderedCollections.OrderedDict{String, Int64}}} where AT, AbstractString}"><code>MIToS.MSA.column_index</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">column_index(msa, col_name)</code></pre><p>Return the index (integer position) of the column with name <code>col_name</code> in the MSA <code>msa</code>. A <code>KeyError</code> is thrown if the column name does not exist. If <code>col_name</code> is an integer, the same integer is returned without checking if it is a valid index.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/GetIndex.jl#L25-L31">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.columngapfraction-Tuple{AbstractMatrix{MIToS.MSA.Residue}}" href="#MIToS.MSA.columngapfraction-Tuple{AbstractMatrix{MIToS.MSA.Residue}}"><code>MIToS.MSA.columngapfraction</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Fraction of gaps per column/position on the MSA</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MSAStats.jl#L100-L102">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.columnname_iterator-Union{Tuple{NamedArrays.NamedMatrix{MIToS.MSA.Residue, AT, Tuple{OrderedCollections.OrderedDict{String, Int64}, OrderedCollections.OrderedDict{String, Int64}}}}, Tuple{AT}} where AT" href="#MIToS.MSA.columnname_iterator-Union{Tuple{NamedArrays.NamedMatrix{MIToS.MSA.Residue, AT, Tuple{OrderedCollections.OrderedDict{String, Int64}, OrderedCollections.OrderedDict{String, Int64}}}}, Tuple{AT}} where AT"><code>MIToS.MSA.columnname_iterator</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>columnname_iterator(msa)</code></p><p>It returns an iterator that returns the column names of the <code>msa</code>. If the <code>msa</code> is a <code>Matrix{Residue}</code> this function returns the actual column numbers as strings. Otherwise it returns the column number of the original MSA through the wrapped <code>NamedArray</code> column names.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MultipleSequenceAlignment.jl#L592-L598">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.columnnames-Union{Tuple{NamedArrays.NamedMatrix{MIToS.MSA.Residue, AT, Tuple{OrderedCollections.OrderedDict{String, Int64}, OrderedCollections.OrderedDict{String, Int64}}}}, Tuple{AT}} where AT" href="#MIToS.MSA.columnnames-Union{Tuple{NamedArrays.NamedMatrix{MIToS.MSA.Residue, AT, Tuple{OrderedCollections.OrderedDict{String, Int64}, OrderedCollections.OrderedDict{String, Int64}}}}, Tuple{AT}} where AT"><code>MIToS.MSA.columnnames</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>columnnames(msa)</code></p><p>It returns a <code>Vector{String}</code> with the sequence names/identifiers. If the <code>msa</code> is a <code>Matrix{Residue}</code> this function returns the actual column numbers as strings. Otherwise it returns the column number of the original MSA through the wrapped <code>NamedArray</code> column names.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MultipleSequenceAlignment.jl#L555-L561">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.columnpairsmatrix-Union{Tuple{diagonal}, Tuple{T}, Tuple{AbstractMatrix{MIToS.MSA.Residue}, Type{T}, Type{Val{diagonal}}, T}} where {T, diagonal}" href="#MIToS.MSA.columnpairsmatrix-Union{Tuple{diagonal}, Tuple{T}, Tuple{AbstractMatrix{MIToS.MSA.Residue}, Type{T}, Type{Val{diagonal}}, T}} where {T, diagonal}"><code>MIToS.MSA.columnpairsmatrix</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Initialize an empty <code>PairwiseListMatrix</code> for a pairwise measure in sequence pairs. It uses the sequence names if they are available, otherwise it uses the actual sequence numbers. You can use the positional argument to indicate the number <code>Type</code> (default: <code>Float64</code>), if the <code>PairwiseListMatrix</code> should store the diagonal values on the list (default: <code>false</code>) and a default value for the diagonal (default: <code>NaN</code>).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/PLM.jl#L31-L37">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.coverage-Tuple{AbstractMatrix{MIToS.MSA.Residue}}" href="#MIToS.MSA.coverage-Tuple{AbstractMatrix{MIToS.MSA.Residue}}"><code>MIToS.MSA.coverage</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Coverage of the sequences with respect of the number of positions on the MSA</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MSAStats.jl#L87-L89">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.delete_annotated_modifications!-Tuple{MIToS.MSA.Annotations}" href="#MIToS.MSA.delete_annotated_modifications!-Tuple{MIToS.MSA.Annotations}"><code>MIToS.MSA.delete_annotated_modifications!</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Deletes all the MIToS annotated modifications</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Annotations.jl#L359-L361">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.deletefullgapcolumns!" href="#MIToS.MSA.deletefullgapcolumns!"><code>MIToS.MSA.deletefullgapcolumns!</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Deletes columns with 100% gaps, this columns are generated by inserts.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/GeneralParserMethods.jl#L301-L303">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.filtercolumns!" href="#MIToS.MSA.filtercolumns!"><code>MIToS.MSA.filtercolumns!</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>filtercolumns!(msa, mask[, annotate::Bool=true])</code></p><p>It allows to filter MSA or aligned sequence columns/positions using a <code>AbstractVector{Bool}</code> <code>mask</code>. Annotations are updated if <code>annotate</code> is <code>true</code> (default).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MSAEditing.jl#L151-L156">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.filtercolumns!-Tuple{MIToS.MSA.Annotations, Any}" href="#MIToS.MSA.filtercolumns!-Tuple{MIToS.MSA.Annotations, Any}"><code>MIToS.MSA.filtercolumns!</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>filtercolumns!(data::Annotations, mask)</code></p><p>It is useful for deleting column annotations (creating a subset in place).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Annotations.jl#L99-L103">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.filtercolumns-Tuple{AbstractMatrix{MIToS.MSA.Residue}, Any}" href="#MIToS.MSA.filtercolumns-Tuple{AbstractMatrix{MIToS.MSA.Residue}, Any}"><code>MIToS.MSA.filtercolumns</code></a> — <span class="docstring-category">Method</span></header><section><div><p>It&#39;s similar to <code>filtercolumns!</code> but for an <code>AbstractMatrix{Residue}</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MSAEditing.jl#L144-L146">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.filtersequences!" href="#MIToS.MSA.filtersequences!"><code>MIToS.MSA.filtersequences!</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>filtersequences!(msa, mask[, annotate::Bool=true])</code></p><p>It allows to filter <code>msa</code> sequences using a <code>AbstractVector{Bool}</code> <code>mask</code> (It removes sequences with <code>false</code> values). <code>AnnotatedMultipleSequenceAlignment</code> annotations are updated if <code>annotate</code> is <code>true</code> (default).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MSAEditing.jl#L106-L112">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.filtersequences!-Tuple{MIToS.MSA.Annotations, Vector{String}, AbstractVector{Bool}}" href="#MIToS.MSA.filtersequences!-Tuple{MIToS.MSA.Annotations, Vector{String}, AbstractVector{Bool}}"><code>MIToS.MSA.filtersequences!</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>filtersequences!(data::Annotations, ids::Vector{String}, mask::AbstractArray{Bool,1})</code></p><p>It is useful for deleting sequence annotations. <code>ids</code> should be a list of the sequence names and <code>mask</code> should be a logical vector.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Annotations.jl#L63-L68">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.filtersequences-Tuple{AbstractMatrix{MIToS.MSA.Residue}, Any}" href="#MIToS.MSA.filtersequences-Tuple{AbstractMatrix{MIToS.MSA.Residue}, Any}"><code>MIToS.MSA.filtersequences</code></a> — <span class="docstring-category">Method</span></header><section><div><p>It&#39;s similar to <code>filtersequences!</code> but for an <code>AbstractMatrix{Residue}</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MSAEditing.jl#L101-L103">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.gapfraction-Tuple{AbstractArray{MIToS.MSA.Residue}}" href="#MIToS.MSA.gapfraction-Tuple{AbstractArray{MIToS.MSA.Residue}}"><code>MIToS.MSA.gapfraction</code></a> — <span class="docstring-category">Method</span></header><section><div><p>It calculates the fraction of gaps on the <code>Array</code> (alignment, sequence, column, etc.). This function can take an extra <code>dimension</code> argument for calculation of the gap fraction over the given dimension.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MSAStats.jl#L5-L9">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.gapstrip!" href="#MIToS.MSA.gapstrip!"><code>MIToS.MSA.gapstrip!</code></a> — <span class="docstring-category">Function</span></header><section><div><p>This functions deletes/filters sequences and columns/positions on the MSA on the following order:</p><ol><li>Removes all the columns/position on the MSA with gaps on the reference (first) sequence.</li><li>Removes all the sequences with a coverage with respect to the number of columns/positions on the MSA <strong>less</strong> than a <code>coveragelimit</code> (default to <code>0.75</code>: sequences with 25% of gaps).</li><li>Removes all the columns/position on the MSA with <strong>more</strong> than a <code>gaplimit</code> (default to <code>0.5</code>: 50% of gaps).</li></ol></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MSAEditing.jl#L337-L347">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.gapstrip-Tuple{AbstractMatrix{MIToS.MSA.Residue}}" href="#MIToS.MSA.gapstrip-Tuple{AbstractMatrix{MIToS.MSA.Residue}}"><code>MIToS.MSA.gapstrip</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Creates a new matrix of <code>Residue</code>s (MSA) with deleted sequences and columns/positions. The MSA is edited in the following way:</p><ol><li>Removes all the columns/position on the MSA with gaps on the reference (first) sequence</li><li>Removes all the sequences with a coverage with respect to the number of columns/positions on the MSA <strong>less</strong> than a <code>coveragelimit</code> (default to <code>0.75</code>: sequences with 25% of gaps)</li><li>Removes all the columns/position on the MSA with <strong>more</strong> than a <code>gaplimit</code> (default to <code>0.5</code>: 50% of gaps)</li></ol></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MSAEditing.jl#L306-L316">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.getannotcolumn" href="#MIToS.MSA.getannotcolumn"><code>MIToS.MSA.getannotcolumn</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>getannotcolumn(ann[, feature[,default]])</code></p><p>It returns per column annotation for <code>feature</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Annotations.jl#L245-L249">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.getannotfile" href="#MIToS.MSA.getannotfile"><code>MIToS.MSA.getannotfile</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>getannotfile(ann[, feature[,default]])</code></p><p>It returns per file annotation for <code>feature</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Annotations.jl#L240-L244">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.getannotresidue" href="#MIToS.MSA.getannotresidue"><code>MIToS.MSA.getannotresidue</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>getannotresidue(ann[, seqname, feature[,default]])</code></p><p>It returns per residue annotation for <code>(seqname, feature)</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Annotations.jl#L255-L259">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.getannotsequence" href="#MIToS.MSA.getannotsequence"><code>MIToS.MSA.getannotsequence</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>getannotsequence(ann[, seqname, feature[,default]])</code></p><p>It returns per sequence annotation for <code>(seqname, feature)</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Annotations.jl#L250-L254">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.getcolumnmapping-Tuple{MIToS.MSA.AnnotatedMultipleSequenceAlignment}" href="#MIToS.MSA.getcolumnmapping-Tuple{MIToS.MSA.AnnotatedMultipleSequenceAlignment}"><code>MIToS.MSA.getcolumnmapping</code></a> — <span class="docstring-category">Method</span></header><section><div><p>It returns a <code>Vector{Int}</code> with the original column number of each column on the actual MSA. The mapping is annotated in the <code>ColMap</code> file annotation of an <code>AnnotatedMultipleSequenceAlignment</code> or in the column names of an <code>NamedArray</code> or <code>MultipleSequenceAlignment</code>.</p><p>NOTE: When the MSA results from vertically concatenating MSAs using <code>vcat</code>, the column map annotations from the constituent MSAs (such as <code>1_ColMap</code>, <code>2_ColMap</code>, etc.) are not returned. Instead, the column numbers referenced in the column names are provided. To access the original annotations, utilize the <code>getannotfile</code> function.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MultipleSequenceAlignment.jl#L775-L785">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.gethcatmapping-Tuple{MIToS.MSA.AnnotatedMultipleSequenceAlignment}" href="#MIToS.MSA.gethcatmapping-Tuple{MIToS.MSA.AnnotatedMultipleSequenceAlignment}"><code>MIToS.MSA.gethcatmapping</code></a> — <span class="docstring-category">Method</span></header><section><div><p>It returns a vector of numbers from <code>1</code> to N for each column that indicates the source MSA. The mapping is annotated in the <code>&quot;HCat&quot;</code> file annotation of an <code>AnnotatedMultipleSequenceAlignment</code> or in the column names of an <code>NamedArray</code> or <code>MultipleSequenceAlignment</code>.</p><p>NOTE: When the MSA results from vertically concatenating MSAs using <code>vcat</code>, the <code>&quot;HCat&quot;</code> annotations from the constituent MSAs are renamed as <code>&quot;1_HCat&quot;</code>, <code>&quot;2_HCat&quot;</code>, etc. In that case, the MSA numbers referenced in the column names are provided. To access the original annotations, utilize the <code>getannotfile</code> function.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Concatenation.jl#L255-L265">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.getnamedict-Tuple{MIToS.MSA.ReducedAlphabet}" href="#MIToS.MSA.getnamedict-Tuple{MIToS.MSA.ReducedAlphabet}"><code>MIToS.MSA.getnamedict</code></a> — <span class="docstring-category">Method</span></header><section><div><p>It takes a <code>ResidueAlphabet</code> and returns a dictionary from group name to group position.</p><pre><code class="language-julia-repl hljs">julia&gt; using MIToS.MSA

julia&gt; ab = ReducedAlphabet(&quot;(AILMV)(RHK)(NQST)(DE)(FWY)CGP&quot;)
ReducedAlphabet of length 8 : &quot;(AILMV)(RHK)(NQST)(DE)(FWY)CGP&quot;

julia&gt; getnamedict(ab)
OrderedCollections.OrderedDict{String, Int64} with 8 entries:
  &quot;AILMV&quot; =&gt; 1
  &quot;RHK&quot;   =&gt; 2
  &quot;NQST&quot;  =&gt; 3
  &quot;DE&quot;    =&gt; 4
  &quot;FWY&quot;   =&gt; 5
  &quot;C&quot;     =&gt; 6
  &quot;G&quot;     =&gt; 7
  &quot;P&quot;     =&gt; 8</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Alphabet.jl#L207-L227">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.getresidues-Tuple{Matrix{MIToS.MSA.Residue}}" href="#MIToS.MSA.getresidues-Tuple{Matrix{MIToS.MSA.Residue}}"><code>MIToS.MSA.getresidues</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>getresidues</code> allows you to access the residues stored inside an MSA or aligned sequence as a <code>Matrix{Residue}</code> without annotations nor column/row names.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MultipleSequenceAlignment.jl#L445-L448">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.getresiduesequences-Tuple{Matrix{MIToS.MSA.Residue}}" href="#MIToS.MSA.getresiduesequences-Tuple{Matrix{MIToS.MSA.Residue}}"><code>MIToS.MSA.getresiduesequences</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>getresiduesequences</code> returns a <code>Vector{Vector{Residue}}</code> with all the MSA sequences without annotations nor column/sequence names.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MultipleSequenceAlignment.jl#L463-L466">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.getsequence" href="#MIToS.MSA.getsequence"><code>MIToS.MSA.getsequence</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>getsequence</code> takes an MSA and a sequence number or identifier and returns an aligned sequence object. If the MSA is an <code>AnnotatedMultipleSequenceAlignment</code>, it returns an <code>AnnotatedAlignedSequence</code> with the sequence annotations. From a <code>MultipleSequenceAlignment</code>, It returns an <code>AlignedSequence</code> object. If an <code>Annotations</code> object and a sequence identifier are used, this function returns the annotations related to the sequence.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MultipleSequenceAlignment.jl#L507-L514">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.getsequencemapping-Tuple{MIToS.MSA.AnnotatedMultipleSequenceAlignment, String}" href="#MIToS.MSA.getsequencemapping-Tuple{MIToS.MSA.AnnotatedMultipleSequenceAlignment, String}"><code>MIToS.MSA.getsequencemapping</code></a> — <span class="docstring-category">Method</span></header><section><div><p>It returns the sequence coordinates as a <code>Vector{Int}</code> for an MSA sequence. That vector has one element for each MSA column. If the number if <code>0</code> in the mapping, there is a gap in that column for that sequence.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MultipleSequenceAlignment.jl#L811-L815">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.getweight-Tuple{MIToS.MSA.NoClustering, Int64}" href="#MIToS.MSA.getweight-Tuple{MIToS.MSA.NoClustering, Int64}"><code>MIToS.MSA.getweight</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>getweight(c[, i::Int])</code></p><p>This function returns the weight of the sequence number <code>i</code>. getweight should be defined for any type used for <code>frequencies!</code>/<code>frequencies</code> in order to use his weigths. If <code>i</code> isn&#39;t used, this function returns a vector with the weight of each sequence.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Clusters.jl#L34-L40">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.hobohmI-Tuple{AbstractMatrix{MIToS.MSA.Residue}, Any}" href="#MIToS.MSA.hobohmI-Tuple{AbstractMatrix{MIToS.MSA.Residue}, Any}"><code>MIToS.MSA.hobohmI</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Sequence clustering using the Hobohm I method from Hobohm et. al. 1992.</p><p><em>Hobohm, Uwe, et al. &quot;Selection of representative protein data sets.&quot; Protein Science 1.3 (1992): 409-417.</em></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Hobohm.jl#L53-L58">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.meanpercentidentity" href="#MIToS.MSA.meanpercentidentity"><code>MIToS.MSA.meanpercentidentity</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Returns the mean of the percent identity between the sequences of a MSA. If the MSA has 300 sequences or less, the mean is exact. If the MSA has more sequences and the <code>exact</code> keyword is <code>false</code> (defualt), 44850 random pairs of sequences are used for the estimation. The number of samples can be changed using the second argument. Use <code>exact=true</code> to perform all the pairwise comparison (the calculation could be slow).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Identity.jl#L132-L138">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.namedmatrix-Tuple{MIToS.MSA.AbstractResidueMatrix}" href="#MIToS.MSA.namedmatrix-Tuple{MIToS.MSA.AbstractResidueMatrix}"><code>MIToS.MSA.namedmatrix</code></a> — <span class="docstring-category">Method</span></header><section><div><p>The <code>namedmatrix</code> function returns the <code>NamedResidueMatrix{Array{Residue,2}}</code> stored in an MSA or aligned sequence.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MultipleSequenceAlignment.jl#L308-L311">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.ncolumns-Tuple{AbstractMatrix{MIToS.MSA.Residue}}" href="#MIToS.MSA.ncolumns-Tuple{AbstractMatrix{MIToS.MSA.Residue}}"><code>MIToS.MSA.ncolumns</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>ncolumns</code> returns the number of MSA columns or positions.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MultipleSequenceAlignment.jl#L458-L460">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.ncolumns-Tuple{MIToS.MSA.Annotations}" href="#MIToS.MSA.ncolumns-Tuple{MIToS.MSA.Annotations}"><code>MIToS.MSA.ncolumns</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>ncolumns(ann::Annotations)</code> returns the number of columns/residues with annotations. This function returns <code>-1</code> if there is not annotations per column/residue.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Annotations.jl#L203-L206">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.nsequences-Tuple{AbstractMatrix{MIToS.MSA.Residue}}" href="#MIToS.MSA.nsequences-Tuple{AbstractMatrix{MIToS.MSA.Residue}}"><code>MIToS.MSA.nsequences</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>nsequences</code> returns the number of sequences on the MSA.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MultipleSequenceAlignment.jl#L453-L455">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.percentidentity-Tuple{Any, Any, Any}" href="#MIToS.MSA.percentidentity-Tuple{Any, Any, Any}"><code>MIToS.MSA.percentidentity</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>percentidentity(seq1, seq2, threshold)</code></p><p>Computes quickly if two aligned sequences have a identity value greater than a given <code>threshold</code> value. Returns a boolean value. Positions with gaps in both sequences doesn&#39;t count to the length of the sequences. Positions with a <code>XAA</code> in at least one sequence aren&#39;t counted.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Identity.jl#L49-L56">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.percentidentity-Tuple{Any, Any}" href="#MIToS.MSA.percentidentity-Tuple{Any, Any}"><code>MIToS.MSA.percentidentity</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>percentidentity(seq1, seq2)</code></p><p>Calculates the fraction of identities between two aligned sequences. The identity value is calculated as the number of identical characters in the i-th position of both sequences divided by the length of both sequences. Positions with gaps in both sequences doesn&#39;t count to the length of the sequences. Positions with a <code>XAA</code> in at least one sequence aren&#39;t counted.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Identity.jl#L28-L36">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.percentidentity-Union{Tuple{AbstractMatrix{MIToS.MSA.Residue}}, Tuple{T}, Tuple{AbstractMatrix{MIToS.MSA.Residue}, Type{T}}} where T" href="#MIToS.MSA.percentidentity-Union{Tuple{AbstractMatrix{MIToS.MSA.Residue}}, Tuple{T}, Tuple{AbstractMatrix{MIToS.MSA.Residue}, Type{T}}} where T"><code>MIToS.MSA.percentidentity</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>percentidentity(msa[, out::Type=Float64])</code></p><p>Calculates the identity between all the sequences on a MSA. You can indicate the output element type with the last optional parameter (<code>Float64</code> by default). For a MSA with a lot of sequences, you can use <code>Float32</code> or <code>Flot16</code> in order to avoid the <code>OutOfMemoryError()</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Identity.jl#L113-L119">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.percentsimilarity" href="#MIToS.MSA.percentsimilarity"><code>MIToS.MSA.percentsimilarity</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Calculates the similarity percent between two aligned sequences. The 100% is the length of the aligned sequences minus the number of columns with gaps in both sequences and the number of columns with at least one residue outside the alphabet. So, columns with residues outside the alphabet (other than the specially treated <code>GAP</code>) aren&#39;t counted to the protein length. Two residues are considered similar if they below to the same group in a <code>ReducedAlphabet</code>. The <code>alphabet</code> (third positional argument) by default is:</p><pre><code class="language-julia hljs">ReducedAlphabet(&quot;(AILMV)(NQST)(RHK)(DE)(FWY)CGP&quot;)</code></pre><p>The first group is composed of the non polar residues <code>(AILMV)</code>, the second group is composed of polar residues, the third group are positive residues, the fourth group are negative residues, the fifth group is composed by the aromatic residues <code>(FWY)</code>. <code>C</code>, <code>G</code> and <code>P</code> are considered unique residues.</p><p><strong>Other residue groups/alphabets:</strong></p><p><strong>SMS (Sequence Manipulation Suite)</strong> Ident and Sim:</p><pre><code class="language-julia hljs">ReducedAlphabet(&quot;(GAVLI)(FYW)(ST)(KRH)(DENQ)P(CM)&quot;)</code></pre><p><em>Stothard P (2000) The Sequence Manipulation Suite: JavaScript programs for analyzing and formatting protein and DNA sequences. Biotechniques 28:1102-1104.</em></p><p><strong>Bio3D 2.2</strong> seqidentity:</p><pre><code class="language-julia hljs">ReducedAlphabet(&quot;(GA)(MVLI)(FYW)(ST)(KRH)(DE)(NQ)PC&quot;)</code></pre><p><em>Grant, B.J. et al. (2006) Bioinformatics 22, 2695–2696.</em></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Identity.jl#L170-L205">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.percentsimilarity-Tuple{AbstractMatrix{MIToS.MSA.Residue}, Vararg{Any}}" href="#MIToS.MSA.percentsimilarity-Tuple{AbstractMatrix{MIToS.MSA.Residue}, Vararg{Any}}"><code>MIToS.MSA.percentsimilarity</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Calculates the similarity percent between all the sequences on a MSA. You can indicate the output element type with the <code>out</code> keyword argument (<code>Float64</code> by default). For an MSA with a lot of sequences, you can use <code>out=Float32</code> or <code>out=Flot16</code> in order to avoid the <code>OutOfMemoryError()</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Identity.jl#L247-L252">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.printmodifications-Tuple{MIToS.MSA.Annotations}" href="#MIToS.MSA.printmodifications-Tuple{MIToS.MSA.Annotations}"><code>MIToS.MSA.printmodifications</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Prints MIToS annotated modifications</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Annotations.jl#L370-L372">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.rename_sequences!-Union{Tuple{AT}, Tuple{NamedArrays.NamedMatrix{MIToS.MSA.Residue, AT, Tuple{OrderedCollections.OrderedDict{String, Int64}, OrderedCollections.OrderedDict{String, Int64}}}, Any}} where AT" href="#MIToS.MSA.rename_sequences!-Union{Tuple{AT}, Tuple{NamedArrays.NamedMatrix{MIToS.MSA.Residue, AT, Tuple{OrderedCollections.OrderedDict{String, Int64}, OrderedCollections.OrderedDict{String, Int64}}}, Any}} where AT"><code>MIToS.MSA.rename_sequences!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">rename_sequences!(msa, newnames)
rename_sequences!(msa, old2new::AbstractDict)
rename_sequences!(msa, old2new::Pair...)</code></pre><p>Rename the sequences of an MSA given a vector of new names, a dictionary mapping old names  to new names, or one or more pairs going from old to new names. If the <code>msa</code> is an  <code>AnnotatedMultipleSequenceAlignment</code>, the annotations are also updated.  The function modifies the <code>msa</code> in place and returns it.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MultipleSequenceAlignment.jl#L864-L870">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.rename_sequences-Tuple{Any, Any}" href="#MIToS.MSA.rename_sequences-Tuple{Any, Any}"><code>MIToS.MSA.rename_sequences</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">rename_sequences(msa, newnames)
rename_sequences(msa, old2new::AbstractDict)
rename_sequences(msa, old2new::Pair...)</code></pre><p>Rename the sequences of an MSA given a vector of new names, a dictionary mapping old names  to new names, or one or more pairs going from old to new names. If the <code>msa</code> is an  <code>AnnotatedMultipleSequenceAlignment</code>, the annotations are also updated.  The function returns a new MSA with the sequences renamed without  modifying the original MSA.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MultipleSequenceAlignment.jl#L892-L899">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.residue2three-Tuple{MIToS.MSA.Residue}" href="#MIToS.MSA.residue2three-Tuple{MIToS.MSA.Residue}"><code>MIToS.MSA.residue2three</code></a> — <span class="docstring-category">Method</span></header><section><div><p>This function returns the three letter name of the <code>Residue</code>.</p><pre><code class="language-julia-repl hljs">julia&gt; using MIToS.MSA

julia&gt; residue2three(Residue(&#39;G&#39;))
&quot;GLY&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/ThreeLetters.jl#L26-L35">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.residuefraction-Tuple{AbstractArray{MIToS.MSA.Residue}}" href="#MIToS.MSA.residuefraction-Tuple{AbstractArray{MIToS.MSA.Residue}}"><code>MIToS.MSA.residuefraction</code></a> — <span class="docstring-category">Method</span></header><section><div><p>It calculates the fraction of residues (no gaps) on the <code>Array</code> (alignment, sequence, column, etc.). This function can take an extra <code>dimension</code> argument for calculation of the residue fraction over the given dimension</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MSAStats.jl#L24-L28">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.sequence_id-Tuple{Union{MIToS.MSA.AbstractAlignedSequence, MIToS.MSA.AbstractSequence}}" href="#MIToS.MSA.sequence_id-Tuple{Union{MIToS.MSA.AbstractAlignedSequence, MIToS.MSA.AbstractSequence}}"><code>MIToS.MSA.sequence_id</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">sequence_id(seq::Union{AbstractSequence,AbstractAlignedSequence})</code></pre><p>It returns the sequence identifier of a sequence object.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MultipleSequenceAlignment.jl#L568-L572">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.sequence_index-Tuple{NamedArrays.NamedMatrix{MIToS.MSA.Residue, AT, Tuple{OrderedCollections.OrderedDict{String, Int64}, OrderedCollections.OrderedDict{String, Int64}}} where AT, AbstractString}" href="#MIToS.MSA.sequence_index-Tuple{NamedArrays.NamedMatrix{MIToS.MSA.Residue, AT, Tuple{OrderedCollections.OrderedDict{String, Int64}, OrderedCollections.OrderedDict{String, Int64}}} where AT, AbstractString}"><code>MIToS.MSA.sequence_index</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">sequence_index(msa, seq_name)</code></pre><p>Return the index (integer position) of the sequence with name <code>seq_name</code> in the MSA <code>msa</code>. A <code>KeyError</code> is thrown if the sequence name does not exist. If <code>seq_name</code> is an integer, the same integer is returned without checking if it is a valid index.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/GetIndex.jl#L16-L22">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.sequencename_iterator-Union{Tuple{NamedArrays.NamedMatrix{MIToS.MSA.Residue, AT, Tuple{OrderedCollections.OrderedDict{String, Int64}, OrderedCollections.OrderedDict{String, Int64}}}}, Tuple{AT}} where AT" href="#MIToS.MSA.sequencename_iterator-Union{Tuple{NamedArrays.NamedMatrix{MIToS.MSA.Residue, AT, Tuple{OrderedCollections.OrderedDict{String, Int64}, OrderedCollections.OrderedDict{String, Int64}}}}, Tuple{AT}} where AT"><code>MIToS.MSA.sequencename_iterator</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>sequencename_iterator(msa)</code></p><p>It returns an iterator that returns the sequence names/identifiers of the <code>msa</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MultipleSequenceAlignment.jl#L581-L585">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.sequencenames-Union{Tuple{NamedArrays.NamedMatrix{MIToS.MSA.Residue, AT, Tuple{OrderedCollections.OrderedDict{String, Int64}, OrderedCollections.OrderedDict{String, Int64}}}}, Tuple{AT}} where AT&lt;:AbstractArray" href="#MIToS.MSA.sequencenames-Union{Tuple{NamedArrays.NamedMatrix{MIToS.MSA.Residue, AT, Tuple{OrderedCollections.OrderedDict{String, Int64}, OrderedCollections.OrderedDict{String, Int64}}}}, Tuple{AT}} where AT&lt;:AbstractArray"><code>MIToS.MSA.sequencenames</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>sequencenames(msa)</code></p><p>It returns a <code>Vector{String}</code> with the sequence names/identifiers.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MultipleSequenceAlignment.jl#L544-L548">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.sequencepairsmatrix-Union{Tuple{diagonal}, Tuple{T}, Tuple{AbstractMatrix{MIToS.MSA.Residue}, Type{T}, Type{Val{diagonal}}, T}} where {T, diagonal}" href="#MIToS.MSA.sequencepairsmatrix-Union{Tuple{diagonal}, Tuple{T}, Tuple{AbstractMatrix{MIToS.MSA.Residue}, Type{T}, Type{Val{diagonal}}, T}} where {T, diagonal}"><code>MIToS.MSA.sequencepairsmatrix</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Initialize an empty <code>PairwiseListMatrix</code> for a pairwise measure in column pairs. It uses the column mapping (column number in the input MSA file) if it’s available, otherwise it uses the actual column numbers. You can use the positional argument to indicate the number <code>Type</code> (default: <code>Float64</code>), if the <code>PairwiseListMatrix</code> should store the diagonal values on the list (default: <code>false</code>) and a default value for the diagonal (default: <code>NaN</code>).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/PLM.jl#L3-L9">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.setannotcolumn!" href="#MIToS.MSA.setannotcolumn!"><code>MIToS.MSA.setannotcolumn!</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>setannotcolumn!(ann, feature, annotation)</code></p><p>It stores per column <code>annotation</code> (1 char per column) for <code>feature</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Annotations.jl#L332-L336">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.setannotfile!" href="#MIToS.MSA.setannotfile!"><code>MIToS.MSA.setannotfile!</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>setannotfile!(ann, feature, annotation)</code></p><p>It stores per file <code>annotation</code> for <code>feature</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Annotations.jl#L327-L331">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.setannotresidue!" href="#MIToS.MSA.setannotresidue!"><code>MIToS.MSA.setannotresidue!</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>setannotresidue!(ann, seqname, feature, annotation)</code></p><p>It stores per residue <code>annotation</code> (1 char per residue) for <code>(seqname, feature)</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Annotations.jl#L342-L346">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.setannotsequence!" href="#MIToS.MSA.setannotsequence!"><code>MIToS.MSA.setannotsequence!</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>setannotsequence!(ann, seqname, feature, annotation)</code></p><p>It stores per sequence <code>annotation</code> for <code>(seqname, feature)</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Annotations.jl#L337-L341">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.setreference!" href="#MIToS.MSA.setreference!"><code>MIToS.MSA.setreference!</code></a> — <span class="docstring-category">Function</span></header><section><div><p>It puts the sequence <code>i</code> (name or position) as reference (first sequence) of the MSA. This function swaps the sequences 1 and <code>i</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MSAEditing.jl#L229-L232">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.shuffle_msa!-Tuple{AbstractMatrix{MIToS.MSA.Residue}, Vararg{Any}}" href="#MIToS.MSA.shuffle_msa!-Tuple{AbstractMatrix{MIToS.MSA.Residue}, Vararg{Any}}"><code>MIToS.MSA.shuffle_msa!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">shuffle_msa!([rng=default_rng(),] msa::AbstractMatrix{Residue}, subset=Colon(); dims=2, fixedgaps=true, fixed_reference=false)</code></pre><p>In-place version of <a href="#MIToS.MSA.shuffle_msa-Tuple{AbstractMatrix{MIToS.MSA.Residue}, Vararg{Any}}"><code>shuffle_msa</code></a>. It randomly permute residues in the MSA <code>msa</code> along sequences (<code>dims=1</code>) or columns  (<code>dims=2</code>, the default). The optional positional argument <code>subset</code> allows to shuffle only  a subset of the sequences or columns. The optional keyword argument <code>fixedgaps</code> indicates  if the gaps should remain their positions (<code>true</code> by default). The optional keyword argument <code>fixed_reference</code> indicates if the residues in the first sequence should remain in their positions (<code>false</code> by default).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Shuffle.jl#L136-L140">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.shuffle_msa-Tuple{AbstractMatrix{MIToS.MSA.Residue}, Vararg{Any}}" href="#MIToS.MSA.shuffle_msa-Tuple{AbstractMatrix{MIToS.MSA.Residue}, Vararg{Any}}"><code>MIToS.MSA.shuffle_msa</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">shuffle_msa([rng=default_rng(),] msa::AbstractMatrix{Residue}, subset=Colon(); dims=2, fixedgaps=true, fixed_reference=false)</code></pre><p>It randomly permute residues in the MSA <code>msa</code> along sequences (<code>dims=1</code>) or columns  (<code>dims=2</code>, the default). The optional positional argument <code>subset</code> allows to shuffle only  a subset of the sequences or columns. The optional keyword argument <code>fixedgaps</code> indicates  if the gaps should remain their positions (<code>true</code> by default). The optional keyword argument <code>fixed_reference</code> indicates if the residues in the first sequence should remain in their positions (<code>false</code> by default).  To shuffle in-place, see <a href="#MIToS.MSA.shuffle_msa!-Tuple{AbstractMatrix{MIToS.MSA.Residue}, Vararg{Any}}"><code>shuffle_msa!</code></a>.</p><pre><code class="language-julia-repl hljs">julia&gt; using MIToS.MSA

julia&gt; using Random

julia&gt; msa = hcat(res&quot;RRE&quot;,res&quot;DDK&quot;, res&quot;G--&quot;)
3×3 Matrix{Residue}:
 R  D  G
 R  D  -
 E  K  -

julia&gt; Random.seed!(42);

julia&gt; shuffle_msa(msa, dims=1, fixedgaps=true)
3×3 Matrix{Residue}:
 G  D  R
 R  D  -
 E  K  -

julia&gt; Random.seed!(42);

julia&gt; shuffle_msa(msa, dims=1, fixedgaps=false)
3×3 Matrix{Residue}:
 G  D  R
 R  -  D
 E  K  -
</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Shuffle.jl#L149-L182">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.stringsequence-Tuple{AbstractMatrix{MIToS.MSA.Residue}, Any}" href="#MIToS.MSA.stringsequence-Tuple{AbstractMatrix{MIToS.MSA.Residue}, Any}"><code>MIToS.MSA.stringsequence</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">stringsequence(seq)
stringsequence(msa, i::Int)
stringsequence(msa, id::String)</code></pre><p>It returns the selected sequence as a <code>String</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MultipleSequenceAlignment.jl#L832-L840">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.swapsequences!-Tuple{Matrix{MIToS.MSA.Residue}, Int64, Int64}" href="#MIToS.MSA.swapsequences!-Tuple{Matrix{MIToS.MSA.Residue}, Int64, Int64}"><code>MIToS.MSA.swapsequences!</code></a> — <span class="docstring-category">Method</span></header><section><div><p>It swaps the sequences on the positions <code>i</code> and <code>j</code> of an MSA. Also it&#39;s possible to swap sequences using their sequence names/identifiers when the MSA object as names.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/MSAEditing.jl#L207-L210">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.MSA.three2residue-Tuple{String}" href="#MIToS.MSA.three2residue-Tuple{String}"><code>MIToS.MSA.three2residue</code></a> — <span class="docstring-category">Method</span></header><section><div><p>It takes a three letter residue name and returns the corresponding <code>Residue</code>. If the name isn&#39;t in the MIToS dictionary, a <code>XAA</code> is returned.</p><pre><code class="language-julia-repl hljs">julia&gt; using MIToS.MSA

julia&gt; three2residue(&quot;ALA&quot;)
A</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/ThreeLetters.jl#L45-L55">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.Utils.parse_file" href="#MIToS.Utils.parse_file"><code>MIToS.Utils.parse_file</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>parse_file(io, format[, output; generatemapping, useidcoordinates, deletefullgaps])</code></p><p>The keyword argument <code>generatemapping</code> (<code>false</code> by default) indicates if the mapping of the sequences (&quot;SeqMap&quot;) and columns (&quot;ColMap&quot;) and the number of columns in the original MSA (&quot;NCol&quot;) should be generated and saved in the annotations. If <code>useidcoordinates</code> is <code>true</code> (default: <code>false</code>) the sequence IDs of the form &quot;ID/start-end&quot; are parsed and used for determining the start and end positions when the mappings are generated. <code>deletefullgaps</code> (<code>true</code> by default) indicates if columns 100% gaps (generally inserts from a HMM) must be removed from the MSA.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/GeneralParserMethods.jl#L338-L348">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Random.shuffle" href="#Random.shuffle"><code>Random.shuffle</code></a> — <span class="docstring-category">Function</span></header><section><div><p>It&#39;s like <code>shuffle</code> but in-place. When a <code>Matrix{Residue}</code> or a <code>AbstractAlignedObject</code> (sequence or MSA) is used, you can indicate if the gaps should remain their positions using the last boolean argument.</p><p><strong>DEPRECATED:</strong> This method is deprecated. Use <a href="#MIToS.MSA.shuffle_msa-Tuple{AbstractMatrix{MIToS.MSA.Residue}, Vararg{Any}}"><code>shuffle_msa</code></a> instead.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Shuffle.jl#L213-L219">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Random.shuffle!" href="#Random.shuffle!"><code>Random.shuffle!</code></a> — <span class="docstring-category">Function</span></header><section><div><p>It&#39;s like <code>Random.shuffle</code>. When a <code>Matrix{Residue}</code> is used, you can indicate if the gaps should remain their positions using the last boolean argument. The previous argument should be the dimension to shuffle, 1 for shuffling residues in a sequence (row) or 2 for shuffling residues in a column.</p><p><strong>DEPRECATED:</strong> This method is deprecated. Use <a href="#MIToS.MSA.shuffle_msa!-Tuple{AbstractMatrix{MIToS.MSA.Residue}, Vararg{Any}}"><code>shuffle_msa!</code></a> instead.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/bb23d29f14ca0e3d4b3a3b74997d6c40251f7919/src/MSA/Shuffle.jl#L187-L194">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../03_RMSF/">« Root Mean Squared Fluctuation (RMSF)</a><a class="docs-footer-nextpage" href="../Information_API/">Information »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.5.0 on <span class="colophon-date" title="Friday 5 July 2024 10:34">Friday 5 July 2024</span>. Using Julia version 1.10.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
