
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="MSA-API">MSA API<a class="anchor-link" href="#MSA-API">&#182;</a></h1><ul>
<li><a href="#Types">Types</a><ul>
<li><a href="#Residues">Residues</a></li>
<li><a href="#Annotations">Annotations</a></li>
<li><a href="#Multiple-Sequence-Alignments">Multiple Sequence Alignments</a></li>
<li><a href="#Sequences">Sequences</a></li>
<li><a href="#MSA-formats">MSA formats</a></li>
<li><a href="#Clustering">Clustering</a></li>
</ul>
</li>
<li><a href="#Constants">Constants</a></li>
<li><a href="#Macros">Macros</a></li>
<li><a href="#Methods">Methods</a><ul>
<li><a href="#Imported-from-Base">Imported from Base</a></li>
<li><a href="#Imported-from-Clustering.jl">Imported from Clustering.jl</a></li>
</ul>
</li>
</ul>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[2]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="k">using</span> <span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[3]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[3]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>The MSA module of MIToS has utilities for working with Multiple Sequence Alignments of protein Sequences (MSA).</p>
<p><strong>Features</strong></p>
<ul>
<li>Read and write MSAs in <code>Stockholm</code>, <code>FASTA</code> or <code>Raw</code> format</li>
<li>Handle MSA annotations</li>
<li>Edit the MSA, e.g. delete columns or sequences, change sequence order, shuffling...</li>
<li>Keep track of positions and annotations after modifications on the MSA</li>
<li>Describe a MSA, e.g. mean percent identity, sequence coverage, gap percentage...</li>
</ul>
<div class="highlight"><pre><span class="k">using</span> <span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span>
</pre></div>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<div class="panel panel-info">
    <div class="panel-heading">
        <strong>Julia help mode</strong>
    </div>
    <div class="panel-body">
        <p>If you type <code>?</code> at the beginning of the Julia REPL line, you will enter in the Julia help mode. In this mode, Julia prints the help or <strong>documentation</strong> of the entered element. This is a nice way of getting information about MIToS functions, types, etc. from Julia.</p>
    </div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p><a href="#"><i class="fa fa-arrow-up"></i></a></p>
<h2 id="Types">Types<a class="anchor-link" href="#Types">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Residues">Residues<a class="anchor-link" href="#Residues">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[4]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">Residue</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[4]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Most of the <strong>MIToS</strong> design is created around the <code>Residue</code> bitstype. It represents the 20 natural amino acids and a GAP value to represent insertion, deletion but also missing data: ambiguous residues and non natural amino acids. Each residue is encoded as an integer number, this allows fast indexing operation using Residues of probability or frequency matrices.</p>
<p><strong>Residue creation and conversion</strong></p>
<p>Creation and <code>convert</code>ion of <code>Residue</code>s should be treated carefully. <code>Residue</code> is encoded as an 8 bits type similar to <code>Int8</code>, to get faster indexing using <code>Int(x::Residue)</code>. In this way, <code>Int</code>, <code>Int8</code> and other signed integers returns the integer value encoded by the residue. Conversions to and from <code>Char</code>s and <code>Uint8</code> are different, to use the <code>Char</code>acter representation in IO operations.</p>
<div class="highlight"><pre><span class="n">julia</span><span class="o">&gt;</span> <span class="n">alanine</span> <span class="o">=</span> <span class="n">Residue</span><span class="p">(</span><span class="sc">&#39;A&#39;</span><span class="p">)</span>
<span class="n">A</span>

<span class="n">julia</span><span class="o">&gt;</span> <span class="kt">Int</span><span class="p">(</span><span class="n">alanine</span><span class="p">)</span>
<span class="mi">1</span>

<span class="n">julia</span><span class="o">&gt;</span> <span class="n">Char</span><span class="p">(</span><span class="n">alanine</span><span class="p">)</span>
<span class="sc">&#39;A&#39;</span>

<span class="n">julia</span><span class="o">&gt;</span> <span class="n">UInt8</span><span class="p">(</span><span class="n">alanine</span><span class="p">)</span> <span class="c"># 0x41 == 65 == &#39;A&#39;</span>
<span class="mh">0x41</span>

<span class="n">julia</span><span class="o">&gt;</span> <span class="k">for</span> <span class="n">residue</span> <span class="k">in</span> <span class="n">res</span><span class="s">&quot;ARNDCQEGHILKMFPSTWYV-&quot;</span>
           <span class="n">println</span><span class="p">(</span><span class="n">residue</span><span class="p">,</span> <span class="s">&quot; &quot;</span><span class="p">,</span> <span class="kt">Int</span><span class="p">(</span><span class="n">residue</span><span class="p">))</span>
       <span class="k">end</span>
<span class="n">A</span> <span class="mi">1</span>
<span class="n">R</span> <span class="mi">2</span>
<span class="n">N</span> <span class="mi">3</span>
<span class="n">D</span> <span class="mi">4</span>
<span class="n">C</span> <span class="mi">5</span>
<span class="n">Q</span> <span class="mi">6</span>
<span class="n">E</span> <span class="mi">7</span>
<span class="n">G</span> <span class="mi">8</span>
<span class="n">H</span> <span class="mi">9</span>
<span class="n">I</span> <span class="mi">10</span>
<span class="n">L</span> <span class="mi">11</span>
<span class="n">K</span> <span class="mi">12</span>
<span class="n">M</span> <span class="mi">13</span>
<span class="n">F</span> <span class="mi">14</span>
<span class="n">P</span> <span class="mi">15</span>
<span class="n">S</span> <span class="mi">16</span>
<span class="n">T</span> <span class="mi">17</span>
<span class="n">W</span> <span class="mi">18</span>
<span class="n">Y</span> <span class="mi">19</span>
<span class="n">V</span> <span class="mi">20</span>
<span class="o">-</span> <span class="mi">21</span>
</pre></div>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p><a href="#"><i class="fa fa-arrow-up"></i></a></p>
<h3 id="Annotations">Annotations<a class="anchor-link" href="#Annotations">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[5]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">Annotations</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[5]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>The <code>Annotations</code> type is basically a container for <code>Dict</code>s with the annotations of a multiple sequence alignment. <code>Annotations</code> was designed for storage of annotations of the <strong>Stockholm format</strong>.</p>
<p>MIToS also uses MSA annotations to keep track of:</p>
<ul>
<li><strong>Modifications</strong> of the MSA (<code>MIToS_...</code>) as deletion of sequences or columns.</li>
<li>Positions numbers in the original MSA file (<strong>column mapping:</strong> <code>ColMap</code>)</li>
<li>Position of the residues in the sequence (<strong>sequence mapping:</strong> <code>SeqMap</code>)</li>
</ul>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p><a href="#"><i class="fa fa-arrow-up"></i></a></p>
<h3 id="Multiple-Sequence-Alignments">Multiple Sequence Alignments<a class="anchor-link" href="#Multiple-Sequence-Alignments">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[6]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">AbstractMultipleSequenceAlignment</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[6]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>MIToS MSAs are subtypes of <code>AbstractMatrix{Residue}</code>, because the most basic implementation of a MIToS MSA is a <code>Matrix</code> of <code>Residue</code>s.</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[7]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">MultipleSequenceAlignment</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[7]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>This MSA type include the <code>Matrix</code> of <code>Residue</code>s and the sequence names. To allow fast indexing of MSAs using <strong>sequence identifiers</strong>, they are saved as an <code>IndexedArray</code>.</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[8]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">AnnotatedMultipleSequenceAlignment</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[8]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>...</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p><a href="#"><i class="fa fa-arrow-up"></i></a></p>
<h3 id="Sequences">Sequences<a class="anchor-link" href="#Sequences">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[9]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">AbstractAlignedSequence</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[9]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>MIToS sequences are subtypes of <code>AbstractVector{Residue}</code>.</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[12]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">AlignedSequence</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[12]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>No documentation found.</p>
<p><strong>Summary:</strong></p>
<div class="highlight"><pre><span class="k">type</span><span class="nc"> MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">AlignedSequence</span> <span class="o">&lt;:</span> <span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">AbstractAlignedSequence</span>
</pre></div>
<p><strong>Fields:</strong></p>
<div class="highlight"><pre><span class="n">id</span>       <span class="p">::</span> <span class="n">ASCIIString</span>
<span class="n">index</span>    <span class="p">::</span> <span class="kt">Int64</span>
<span class="n">sequence</span> <span class="p">::</span> <span class="n">Array</span><span class="p">{</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">Residue</span><span class="p">,</span><span class="mi">1</span><span class="p">}</span>
</pre></div>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[13]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">AnnotatedAlignedSequence</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[13]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>No documentation found.</p>
<p><strong>Summary:</strong></p>
<div class="highlight"><pre><span class="k">type</span><span class="nc"> MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">AnnotatedAlignedSequence</span> <span class="o">&lt;:</span> <span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">AbstractAlignedSequence</span>
</pre></div>
<p><strong>Fields:</strong></p>
<div class="highlight"><pre><span class="n">id</span>          <span class="p">::</span> <span class="n">ASCIIString</span>
<span class="n">index</span>       <span class="p">::</span> <span class="kt">Int64</span>
<span class="n">sequence</span>    <span class="p">::</span> <span class="n">Array</span><span class="p">{</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">Residue</span><span class="p">,</span><span class="mi">1</span><span class="p">}</span>
<span class="n">annotations</span> <span class="p">::</span> <span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">Annotations</span>
</pre></div>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p><a href="#"><i class="fa fa-arrow-up"></i></a></p>
<h3 id="MSA-formats">MSA formats<a class="anchor-link" href="#MSA-formats">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[14]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">Stockholm</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[14]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>No documentation found.</p>
<p><strong>Summary:</strong></p>
<div class="highlight"><pre><span class="n">immutable</span> <span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">Stockholm</span> <span class="o">&lt;:</span> <span class="n">MIToS</span><span class="o">.</span><span class="n">Utils</span><span class="o">.</span><span class="n">Format</span>
</pre></div>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[15]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">FASTA</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[15]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>No documentation found.</p>
<p><strong>Summary:</strong></p>
<div class="highlight"><pre><span class="n">immutable</span> <span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">FASTA</span> <span class="o">&lt;:</span> <span class="n">MIToS</span><span class="o">.</span><span class="n">Utils</span><span class="o">.</span><span class="n">Format</span>
</pre></div>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[16]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">Raw</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[16]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>No documentation found.</p>
<p><strong>Summary:</strong></p>
<div class="highlight"><pre><span class="n">immutable</span> <span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">Raw</span> <span class="o">&lt;:</span> <span class="n">MIToS</span><span class="o">.</span><span class="n">Utils</span><span class="o">.</span><span class="n">Format</span>
</pre></div>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p><a href="#"><i class="fa fa-arrow-up"></i></a></p>
<h3 id="Clustering">Clustering<a class="anchor-link" href="#Clustering">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[17]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="n">Docs</span><span class="o">.</span><span class="n">typesummary</span><span class="p">(</span> <span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">ClusteringResult</span> <span class="p">)</span> <span class="c"># imported from Clustering.jl</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[17]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p><strong>Summary:</strong></p>
<div class="highlight"><pre><span class="k">abstract</span><span class="nc"> Clustering</span><span class="o">.</span><span class="n">ClusteringResult</span> <span class="o">&lt;:</span> <span class="kt">Any</span>
</pre></div>
<p><strong>Subtypes:</strong></p>
<div class="highlight"><pre><span class="n">Clustering</span><span class="o">.</span><span class="n">AffinityPropResult</span>
<span class="n">Clustering</span><span class="o">.</span><span class="n">DbscanResult</span>
<span class="n">Clustering</span><span class="o">.</span><span class="n">KmeansResult</span><span class="p">{</span><span class="n">T</span><span class="o">&lt;:</span><span class="n">AbstractFloat</span><span class="p">}</span>
<span class="n">Clustering</span><span class="o">.</span><span class="n">KmedoidsResult</span><span class="p">{</span><span class="n">T</span><span class="p">}</span>
<span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">NoClustering</span>
<span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">SequenceClusters</span>
</pre></div>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[18]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">NoClustering</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[18]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>No documentation found.</p>
<p><strong>Summary:</strong></p>
<div class="highlight"><pre><span class="n">immutable</span> <span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">NoClustering</span> <span class="o">&lt;:</span> <span class="n">Clustering</span><span class="o">.</span><span class="n">ClusteringResult</span>
</pre></div>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[19]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">SequenceClusters</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[19]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Data structure to represent sequence clusters. The sequence data itself is not included.</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p><a href="#"><i class="fa fa-arrow-up"></i></a></p>
<h2 id="Constants">Constants<a class="anchor-link" href="#Constants">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[20]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">GAP</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[20]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p><code>GAP</code> is the character/number representation on <strong>MIToS</strong> for gaps (also for non standard residues). Lowercase characters and dots are also encoded as <code>GAP</code> in conversion from <code>String</code>s and <code>Char</code>s. This <code>Residue</code> constant is encoded as <code>Residue(21)</code>.</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p><a href="#"><i class="fa fa-arrow-up"></i></a></p>
<h2 id="Macros">Macros<a class="anchor-link" href="#Macros">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[21]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="p">@</span><span class="n">res_str</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[21]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>The MIToS macro <code>@res_str</code> takes a string and returns a <code>Vector</code> of <code>Residues</code> (sequence).</p>
<div class="highlight"><pre><span class="n">julia</span><span class="o">&gt;</span> <span class="n">res</span><span class="s">&quot;MIToS&quot;</span>
<span class="mi">5</span><span class="o">-</span><span class="n">element</span> <span class="n">Array</span><span class="p">{</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">Residue</span><span class="p">,</span><span class="mi">1</span><span class="p">}:</span>
 <span class="n">M</span>
 <span class="n">I</span>
 <span class="n">T</span>
 <span class="o">-</span>
 <span class="n">S</span>
</pre></div>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p><a href="#"><i class="fa fa-arrow-up"></i></a></p>
<h2 id="Methods">Methods<a class="anchor-link" href="#Methods">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[22]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">swap</span><span class="o">!</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[22]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p><code>swap!(ia::IndexedArray, to::Int, from::Int)</code> interchange/swap the values on the indices <code>to</code> and <code>from</code> in the <code>IndexedArray</code></p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[23]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">annotations</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[23]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Returns the annotations of a MSA or a sequence.</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[24]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">filtersequences</span><span class="o">!</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[24]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p><code>filtersequences!(msa, mask[, annotate::Bool=true])</code></p>
<p>Allows to filter <code>msa</code> sequences using a <code>AbstractVector{Bool}</code> <code>mask</code> (removes <code>false</code>s). <code>AnnotatedMultipleSequenceAlignment</code> annotations are updated if <code>annotate</code> is <code>true</code> (default).</p>
<p><code>filtersequences!(data::Annotations, ids::IndexedArray, mask::AbstractArray{Bool,1})</code> is useful for deleting sequence annotations. <code>ids</code> should be an <code>IndexedArray</code> with the <code>seqname</code>s of the annotated sequences and <code>mask</code> should be a logical vector.</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[25]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">filtercolumns</span><span class="o">!</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[25]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p><code>filtercolumns!(msa, mask[, annotate::Bool=true])</code></p>
<p>Allows to filter MSA columns/sequence positions using a <code>AbstractVector{Bool}</code> <code>mask</code>. <code>AnnotatedMultipleSequenceAlignment</code>s or <code>AnnotatedAlignedSequence</code>s annotations are updated.</p>
<p><code>filtercolumns!(data::Annotations, mask)</code> is useful for deleting column annotations (creating a subset in place).</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[26]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">empty</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[26]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Creates empty MSA <code>Annotations</code> of length 0 using <code>sizehint!</code></p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[27]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">getannotfile</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[27]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p><code>getannotfile(ann[, feature[,default]])</code> returns per file annotation for <code>feature</code></p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[28]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">getannotcolumn</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[28]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p><code>getannotcolumn(ann[, feature[,default]])</code> returns per column annotation for <code>feature</code></p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[29]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">getannotsequence</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[29]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p><code>getannotsequence(ann[, seqname, feature[,default]])</code> returns per sequence annotation for <code>(seqname, feature)</code></p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[30]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">getannotresidue</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[30]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p><code>getannotresidue(ann[, seqname, feature[,default]])</code> returns per residue annotation for <code>(seqname, feature)</code></p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[31]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">setannotfile</span><span class="o">!</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[31]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p><code>setannotfile!(ann, feature, annotation)</code> stores per file <code>annotation</code> for <code>feature</code></p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[32]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">setannotcolumn</span><span class="o">!</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[32]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p><code>setannotcolumn!(ann, feature, annotation)</code> stores per column <code>annotation</code> (1 char per column) for <code>feature</code></p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[33]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">setannotsequence</span><span class="o">!</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[33]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p><code>setannotsequence!(ann, seqname, feature, annotation)</code> stores per sequence <code>annotation</code> for <code>(seqname, feature)</code></p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[34]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">setannotresidue</span><span class="o">!</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[34]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p><code>setannotresidue!(ann, seqname, feature, annotation)</code> stores per residue <code>annotation</code> (1 char per residue) for <code>(seqname, feature)</code></p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[35]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">annotate_modification</span><span class="o">!</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[35]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Annotates on file annotations the modifications realized by MIToS on the MSA</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[36]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">delete_annotated_modifications</span><span class="o">!</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[36]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Deletes all the MIToS annotated modifications</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[37]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">printmodifications</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[37]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Prints MIToS annotated modifications</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[38]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">getresidues</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[38]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Allows you to access the residues in a <code>Matrix{Residues}</code>/<code>Vector{Residues}</code> without annotations.</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[39]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">getsequence</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[39]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p><code>getsequence</code> takes an MSA and a sequence number or identifier to return a sequence. If the MSA is n <code>AnnotatedMultipleSequenceAlignment</code>, it returns an <code>AnnotatedAlignedSequence</code> with sequence annotations. From a <code>MultipleSequenceAlignment</code>, It returns an <code>AlignedSequence</code> object. If an <code>Annotations</code> object  and a sequence identifier are used, this function returns the annotations related to the sequence.</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[40]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">getresiduesequences</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[40]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Gives you a <code>Vector{Vector{Residue}}</code> with all the sequences of the MSA without Annotations</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[41]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">nsequences</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[41]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Gives you the number of sequences on the <code>MultipleSequenceAlignment</code></p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[42]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">ncolumns</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[42]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p><code>ncolumns</code> returns the MSA columns or sequence positions number.</p>
<p><code>ncolumns(ann::Annotations)</code> returns the number of columns/residues with annotations. This function returns <code>-1</code> if there is not annotations per column/residue.</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[43]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">gapfraction</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[43]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Calculates the fraction of gaps on the <code>Array</code> (alignment, sequence, column, etc.). This function can take an extra <code>dim</code> argument for calculation of the gap fraction over the given dimension</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[44]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">residuefraction</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[44]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Calculates the fraction of residues (no gaps) on the <code>Array</code> (alignment, sequence, column, etc.) This function can take an extra <code>dim</code> argument for calculation of the residue fraction over the given dimension</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[45]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">coverage</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[45]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Coverage of the sequences with respect of the number of positions on the MSA</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[46]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">columngapfraction</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[46]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Fraction of gaps per column/position on the MSA</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[47]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">setreference</span><span class="o">!</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[47]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Puts the sequence <code>i</code> as reference (as the first sequence) of the MSA. This function swaps the sequences 1 and <code>i</code>, also an <code>id</code> can be used to select the sequence.</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[48]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">gapstrip</span><span class="o">!</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[48]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>This functions deletes/filters sequences and columns/positions on the MSA on the following order:</p>
<ul>
<li>Removes all the columns/position on the MSA with gaps on the reference sequence (first sequence)</li>
<li>Removes all the sequences with a coverage (with respect to the number of columns/positions on the MSA) <strong>less</strong> than a <code>coveragelimit</code> (default to <code>0.75</code>)</li>
<li>Removes all the columns/position on the MSA with <strong>more</strong> than a <code>gaplimit</code> (default to <code>0.5</code>: 50% of gaps)</li>
</ul>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[49]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">adjustreference</span><span class="o">!</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[49]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Removes positions/columns of the MSA with gaps in the reference (first) sequence</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[50]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">asciisequence</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[50]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Gives an ASCIIString with the sequence number <code>seq</code> of the MSA</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[51]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">gapstrip</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[51]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Creates a new <code>Matrix{Residue}</code> with deleted sequences and columns/positions on the MSA:</p>
<ul>
<li>Removes all the columns/position on the MSA with gaps on the reference sequence (first sequence)</li>
<li>Removes all the sequences with a coverage with respect to the number of columns/positions on the MSA <strong>less</strong> than a <code>coveragelimit</code> (default to <code>0.75</code>: sequences with 25% of gaps)</li>
<li>Removes all the columns/position on the MSA with <strong>more</strong> than a <code>gaplimit</code> (default to <code>0.5</code>: 50% of gaps)</li>
</ul>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[52]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">adjustreference</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[52]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Creates a new Matrix{Residue}. This function deletes positions/columns of the MSA with gaps in the reference (first) sequence</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[53]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">filtersequences</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[53]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>It's similar to <code>filtersequences!</code> but for a <code>Matrix</code> of <code>Residue</code>s</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[54]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">filtercolumns</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[54]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>It's similar to <code>filtercolumns!</code> but for a <code>Matrix</code> or <code>Vector</code> of <code>Residue</code>s</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[55]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">getcolumnmapping</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[55]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>No documentation found.</p>
<p><code>MIToS.MSA.getcolumnmapping</code> is a generic <code>Function</code>.</p>
<div class="highlight"><pre><span class="c"># 1 method for generic function &quot;getcolumnmapping&quot;:</span>
<span class="n">getcolumnmapping</span><span class="p">(</span><span class="n">msa</span><span class="p">::</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">AnnotatedMultipleSequenceAlignment</span><span class="p">)</span> <span class="n">at</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">dzea</span><span class="o">/.</span><span class="n">julia</span><span class="o">/</span><span class="n">v0</span><span class="o">.</span><span class="mi">4</span><span class="o">/</span><span class="n">MIToS</span><span class="o">/</span><span class="n">src</span><span class="o">/</span><span class="n">MSA</span><span class="o">/</span><span class="n">MultipleSequenceAlignment</span><span class="o">.</span><span class="n">jl</span><span class="p">:</span><span class="mi">473</span>
</pre></div>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[56]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">getsequencemapping</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[56]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Returns the sequence coordinates as a <code>Vector{Int}</code> for an MSA sequence. That vector has one element for each MSA column. If the number if <code>0</code> in the mapping, there is a gap in that column for that sequence.</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[57]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">shuffle_columnwise</span><span class="o">!</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[57]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Shuffles the residues in each column</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[58]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">shuffle_sequencewise</span><span class="o">!</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[58]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Shuffles the residues in each sequence</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[59]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">shuffle_residues_sequencewise</span><span class="o">!</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[59]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Shuffles the residues in each sequence, keeping fixed the gap positions</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[60]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">shuffle_residues_columnwise</span><span class="o">!</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[60]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Shuffles the residues in each column, keeping fixed the gap positions</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[61]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">percentidentity</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[61]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p><code>percentidentity(seq1, seq2, threshold)</code></p>
<p>Computes quickly if two aligned sequences have a identity value greater than a given <code>threshold</code> value. Returns a boolean value.</p>
<p><code>percentidentity(seq1, seq2)</code></p>
<p>Calculates the fraction of identities between two aligned sequences. The identity value is calculated as the number of identical characters in the i-th position of both sequences divided by the length of both sequences. Positions with gaps in both sequences are not counted in the length of the sequence.</p>
<p><code>percentidentity(msa[, out::Type=Float64])</code></p>
<p>Calculates the identity between all the sequences on a MSA. You can indicate the output element type with the last optional parameter (<code>Float64</code> by default). For a MSA with a lot of sequences, you can use <code>Float32</code> or <code>Flot16</code> in order to avoid the <code>OutOfMemoryError()</code>.</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[62]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">meanpercentidentity</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[62]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Returns the mean of the percent identity between the sequences of a MSA. If the MSA has 300 sequences or less, the mean is exact. If the MSA has more sequences and the <code>exact</code> keyword is <code>false</code> (defualt), 44850 random pairs of sequences are used for the estimation. The number of samples can be changed using the second argument. Use <code>exact=true</code> to perform all the pairwise comparison (the calculation could be slow).</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[63]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">percentsimilarity</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[63]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Calculates the similarity percent between all the sequences on a MSA. You can indicate the output element type with the <code>out</code> keyword argument (<code>Float64</code> by default). For a MSA with a lot of sequences, you can use <code>out=Float32</code> or <code>out=Flot16</code> in order to avoid the <code>OutOfMemoryError()</code>.</p>
<p>Calculates the similarity percent between two aligned sequences. The 100% is the length of the aligned sequences minus the number of columns with gaps in both sequences. Two residues are considered similar if they below to the same group. The <code>groups</code> (third positional argument) can be indicated with a vector of length 20, having the group labels of each <code>Residue</code> in the following order (mandatory): A, R, N, D, C, Q, E, G, H, I, L, K, M, F, P, S, T, W, Y, V</p>
<p><strong>MIToS default groups</strong> are:</p>
<div class="highlight"><pre><span class="p">[</span>
<span class="p">:</span><span class="n">nonpolar</span><span class="p">,</span> <span class="c"># A</span>
<span class="p">:</span><span class="n">positive</span><span class="p">,</span> <span class="c"># R</span>
<span class="p">:</span><span class="n">polar</span><span class="p">,</span>    <span class="c"># N</span>
<span class="p">:</span><span class="n">negative</span><span class="p">,</span> <span class="c"># D</span>
<span class="p">:</span><span class="n">cysteine</span><span class="p">,</span> <span class="c"># C</span>
<span class="p">:</span><span class="n">polar</span><span class="p">,</span>    <span class="c"># Q</span>
<span class="p">:</span><span class="n">negative</span><span class="p">,</span> <span class="c"># E</span>
<span class="p">:</span><span class="n">glycine</span><span class="p">,</span>  <span class="c"># G</span>
<span class="p">:</span><span class="n">positive</span><span class="p">,</span> <span class="c"># H</span>
<span class="p">:</span><span class="n">nonpolar</span><span class="p">,</span> <span class="c"># I</span>
<span class="p">:</span><span class="n">nonpolar</span><span class="p">,</span> <span class="c"># L</span>
<span class="p">:</span><span class="n">positive</span><span class="p">,</span> <span class="c"># K</span>
<span class="p">:</span><span class="n">nonpolar</span><span class="p">,</span> <span class="c"># M</span>
<span class="p">:</span><span class="n">aromatic</span><span class="p">,</span> <span class="c"># F</span>
<span class="p">:</span><span class="n">proline</span><span class="p">,</span>  <span class="c"># P</span>
<span class="p">:</span><span class="n">polar</span><span class="p">,</span>    <span class="c"># S</span>
<span class="p">:</span><span class="n">polar</span><span class="p">,</span>    <span class="c"># T</span>
<span class="p">:</span><span class="n">aromatic</span><span class="p">,</span> <span class="c"># W</span>
<span class="p">:</span><span class="n">aromatic</span><span class="p">,</span> <span class="c"># Y</span>
<span class="p">:</span><span class="n">nonpolar</span>  <span class="c"># V</span>
<span class="p">]</span>
</pre></div>
<p>Other residue groups:</p>
<p><strong>SMS (Sequence Manipulation Suite)</strong> Ident and Sim: GAVLI, FYW, ST, KRH, DENQ, P, CM</p>
<p><em>Stothard P (2000) The Sequence Manipulation Suite: JavaScript programs for analyzing and formatting protein and DNA sequences. Biotechniques 28:1102-1104.</em></p>
<div class="highlight"><pre><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
</pre></div>
<p><strong>Bio3D 2.2</strong> seqidentity: GA, MVLI, FYW, ST, KRH, DE, NQ, P, C</p>
<p><em>Grant, B.J. et al. (2006) Bioinformatics 22, 26952696.</em></p>
<div class="highlight"><pre><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">]</span>
</pre></div>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[64]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">getweight</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[64]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p><code>getweight(c[, i::Int])</code></p>
<p>This function returns the weight of the sequence number <code>i</code>. getweight should be defined for any type used for <code>count!</code>/<code>count</code> in order to use his weigths. If <code>i</code> isn't used, this function returns a vector with the weight of each sequence.</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[65]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">hobohmI</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[65]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Sequence clustering using the Hobohm I method from Hobohm et. al. 1992.</p>
<p><em>Hobohm, Uwe, et al. "Selection of representative protein data sets." Protein Science 1.3 (1992): 409-417.</em></p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[66]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">sequencepairsmatrix</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[66]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Initialize an empty <code>PairwiseListMatrix</code> for a pairwise measure in column pairs. This function uses the column mapping (column number in the input MSA file) if its available, otherwise it uses the actual column numbers. You can use the positional argument to indicate the number <code>Type</code> (default: <code>Float64</code>), if the <code>PairwiseListMatrix</code> should store the diagonal values on the list (default: <code>false</code>) and a default value for the diagonal (default: <code>NaN</code>).</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[67]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">columnpairsmatrix</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[67]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Initialize an empty <code>PairwiseListMatrix</code> for a pairwise measure in sequence pairs. This function uses the sequence names if they are available, otherwise it uses the actual sequence numbers. You can use the positional argument to indicate the number <code>Type</code> (default: <code>Float64</code>), if the <code>PairwiseListMatrix</code> should store the diagonal values on the list (default: <code>false</code>) and a default value for the diagonal (default: <code>NaN</code>).</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[68]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">columnlabels</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[68]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Returns the column labels: use the column mapping (column number in the input MSA file) if its available, otherwise it uses the actual column numbers.</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[69]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">sequencelabels</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[69]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Returns the sequence labels: use the sequence names if they are available, otherwise it uses the actual sequence numbers.</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[70]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">three2residue</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[70]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>It takes a three letter residue name and returns the <code>Residue</code>.</p>
<div class="highlight"><pre><span class="n">julia</span><span class="o">&gt;</span> <span class="n">three2residue</span><span class="p">(</span><span class="s">&quot;ALA&quot;</span><span class="p">)</span>
<span class="n">A</span>
</pre></div>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[71]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">residue2three</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[71]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>This function returns the three letter name of the <code>Residue</code>.</p>
<div class="highlight"><pre><span class="n">julia</span><span class="o">&gt;</span> <span class="n">residue2three</span><span class="p">(</span><span class="n">Residue</span><span class="p">(</span><span class="sc">&#39;G&#39;</span><span class="p">))</span>
<span class="s">&quot;GLY&quot;</span>
</pre></div>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p><a href="#"><i class="fa fa-arrow-up"></i></a></p>
<h3 id="Imported-from-Base">Imported from Base<a class="anchor-link" href="#Imported-from-Base">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[72]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">names</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[72]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">

<pre><code>names(x::Module[, all=false[, imported=false]])</code></pre>
<p>Get an array of the names exported by a <code>Module</code>, with optionally more <code>Module</code> globals according to the additional parameters.</p>
<p><code>names(msa)</code></p>
<p>Returns the <code>MSA</code> sequences names or identifiers as an <code>IndexedArray</code>.</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p><a href="#"><i class="fa fa-arrow-up"></i></a></p>
<h3 id="Imported-from-Clustering.jl">Imported from Clustering.jl<a class="anchor-link" href="#Imported-from-Clustering.jl">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[73]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">nclusters</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[73]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Get the number of clusters in a <code>SequenceClusters</code> object.</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[74]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">counts</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[74]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Get sample counts of clusters as a <code>Vector</code>. Each <code>k</code> value is the number of samples assigned to the k-th cluster.</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[75]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-julia"><pre><span class="o">?</span><span class="n">MIToS</span><span class="o">.</span><span class="n">MSA</span><span class="o">.</span><span class="n">assignments</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area"><div class="prompt output_prompt">Out[75]:</div>

<div class="output_markdown rendered_html output_subarea output_execute_result">
<p>Get a vector of assignments, where the <code>i</code> value is the index/number of the cluster to which the i-th sequence is assigned.</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p><a href="#"><i class="fa fa-arrow-up"></i></a></p>

</div>
</div>
</div>