<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>SIFTS · MIToS</title><meta name="title" content="SIFTS · MIToS"/><meta property="og:title" content="SIFTS · MIToS"/><meta property="twitter:title" content="SIFTS · MIToS"/><meta name="description" content="Documentation for MIToS."/><meta property="og:description" content="Documentation for MIToS."/><meta property="twitter:description" content="Documentation for MIToS."/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">MIToS</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../Installation/">Installation</a></li><li><a class="tocitem" href="../Example/">Example</a></li><li><span class="tocitem">Modules</span><ul><li><a class="tocitem" href="../MSA/">MSA</a></li><li><a class="tocitem" href="../Information/">Information</a></li><li><a class="tocitem" href="../SIFTS/">SIFTS</a></li><li><a class="tocitem" href="../PDB/">PDB</a></li><li><a class="tocitem" href="../Pfam/">Pfam</a></li></ul></li><li><span class="tocitem">Cookbook</span><ul><li><a class="tocitem" href="../01_Change_B_factors/">Change B-factors</a></li><li><a class="tocitem" href="../02_Linking_structural_and_evolutionary_information/">Linking structural and evolutionary information</a></li><li><a class="tocitem" href="../03_RMSF/">Root Mean Squared Fluctuation (RMSF)</a></li></ul></li><li><span class="tocitem">API</span><ul><li><a class="tocitem" href="../MSA_API/">MSA</a></li><li><a class="tocitem" href="../Information_API/">Information</a></li><li class="is-active"><a class="tocitem" href>SIFTS</a><ul class="internal"><li><a class="tocitem" href="#Contents"><span>Contents</span></a></li><li><a class="tocitem" href="#Types"><span>Types</span></a></li><li><a class="tocitem" href="#Constants"><span>Constants</span></a></li><li><a class="tocitem" href="#Macros"><span>Macros</span></a></li><li><a class="tocitem" href="#Methods-and-functions"><span>Methods and functions</span></a></li></ul></li><li><a class="tocitem" href="../PDB_API/">PDB</a></li><li><a class="tocitem" href="../Pfam_API/">Pfam</a></li><li><a class="tocitem" href="../Utils_API/">Utils</a></li></ul></li><li><a class="tocitem" href="../Scripts/">Scripts</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">API</a></li><li class="is-active"><a href>SIFTS</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>SIFTS</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/diegozea/MIToS.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/diegozea/MIToS.jl/blob/master/docs/src/SIFTS_API.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="SIFTS"><a class="docs-heading-anchor" href="#SIFTS">SIFTS</a><a id="SIFTS-1"></a><a class="docs-heading-anchor-permalink" href="#SIFTS" title="Permalink"></a></h1><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.SIFTS" href="#MIToS.SIFTS"><code>MIToS.SIFTS</code></a> — <span class="docstring-category">Module</span></header><section><div><p>The <code>SIFTS</code> module of MIToS allows to obtain the residue-level mapping between databases stored in the SIFTS XML files. It makes easy to assign PDB residues to UniProt/Pfam positions. Given the fact that pairwise alignments can lead to misleading association between residues in both sequences, SIFTS offers  more reliable association between sequence and structure residue numbers.</p><p><strong>Features</strong></p><ul><li>Download and parse SIFTS XML files</li><li>Store residue-level mapping in Julia</li><li>Easy generation of <code>OrderedDict</code>s between residues numbers</li></ul><pre><code class="language-julia hljs">using MIToS.SIFTS</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/09af65a7583627f53bcd1ee131646a8ee62cc0e7/src/SIFTS/SIFTS.jl#L1-L18">source</a></section></article><h2 id="Contents"><a class="docs-heading-anchor" href="#Contents">Contents</a><a id="Contents-1"></a><a class="docs-heading-anchor-permalink" href="#Contents" title="Permalink"></a></h2><ul><li><a href="#SIFTS">SIFTS</a></li><li class="no-marker"><ul><li><a href="#Contents">Contents</a></li><li><a href="#Types">Types</a></li><li><a href="#Constants">Constants</a></li><li><a href="#Macros">Macros</a></li><li><a href="#Methods-and-functions">Methods and functions</a></li></ul></li></ul><h2 id="Types"><a class="docs-heading-anchor" href="#Types">Types</a><a id="Types-1"></a><a class="docs-heading-anchor-permalink" href="#Types" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.SIFTS.SIFTSResidue" href="#MIToS.SIFTS.SIFTSResidue"><code>MIToS.SIFTS.SIFTSResidue</code></a> — <span class="docstring-category">Type</span></header><section><div><p>A <code>SIFTSResidue</code> object stores the SIFTS residue level mapping for a residue. It has the following fields that you can access at any moment for query purposes:</p><pre><code class="nohighlight hljs">- `PDBe` : A `dbPDBe` object, it&#39;s present in all the `SIFTSResidue`s.
- `UniProt` : A `dbUniProt` object or `missing`.
- `Pfam` : A `dbPfam` object or `missing`.
- `NCBI` : A `dbNCBI` object or `missing`.
- `InterPro` : An array of `dbInterPro` objects.
- `PDB` : A `dbPDB` object or `missing`.
- `SCOP` : A `dbSCOP` object or `missing`.
- `SCOP2` : An array of `dbSCOP2` objects.
- `SCOP2B` : A `dbSCOP2B` object or `missing`.
- `CATH` : A `dbCATH` object or `missing`.
- `Ensembl` : An array of `dbEnsembl` objects.
- `missing` : It&#39;s `true` if the residue is missing, i.e. not observed, in the structure.
- `sscode` : A string with the secondary structure code of the residue.
- `ssname` : A string with the secondary structure name of the residue.</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/09af65a7583627f53bcd1ee131646a8ee62cc0e7/src/SIFTS/ResidueMapping.jl#L175-L193">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.SIFTS.dbCATH" href="#MIToS.SIFTS.dbCATH"><code>MIToS.SIFTS.dbCATH</code></a> — <span class="docstring-category">Type</span></header><section><div><p><code>dbCATH</code> stores the residue <code>id</code>, <code>number</code>, <code>name</code> and <code>chain</code> in CATH as  strings.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/09af65a7583627f53bcd1ee131646a8ee62cc0e7/src/SIFTS/ResidueMapping.jl#L76-L79">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.SIFTS.dbEnsembl" href="#MIToS.SIFTS.dbEnsembl"><code>MIToS.SIFTS.dbEnsembl</code></a> — <span class="docstring-category">Type</span></header><section><div><p><code>dbEnsembl</code> stores the residue (gene) accession <code>id</code>, the <code>transcript</code>, <code>translation</code> and <code>exon</code> ids in Ensembl as strings, together with the residue <code>number</code> and <code>name</code> using the UniProt coordinates.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/09af65a7583627f53bcd1ee131646a8ee62cc0e7/src/SIFTS/ResidueMapping.jl#L21-L25">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.SIFTS.dbInterPro" href="#MIToS.SIFTS.dbInterPro"><code>MIToS.SIFTS.dbInterPro</code></a> — <span class="docstring-category">Type</span></header><section><div><p><code>dbInterPro</code> stores the residue <code>id</code>, <code>number</code>, <code>name</code> and <code>evidence</code> in InterPro as strings.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/09af65a7583627f53bcd1ee131646a8ee62cc0e7/src/SIFTS/ResidueMapping.jl#L11-L13">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.SIFTS.dbNCBI" href="#MIToS.SIFTS.dbNCBI"><code>MIToS.SIFTS.dbNCBI</code></a> — <span class="docstring-category">Type</span></header><section><div><p><code>dbNCBI</code> stores the residue <code>id</code>, <code>number</code> and <code>name</code> in NCBI as strings.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/09af65a7583627f53bcd1ee131646a8ee62cc0e7/src/SIFTS/ResidueMapping.jl#L55-L57">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.SIFTS.dbPDB" href="#MIToS.SIFTS.dbPDB"><code>MIToS.SIFTS.dbPDB</code></a> — <span class="docstring-category">Type</span></header><section><div><p><code>dbPDB</code> stores the residue <code>id</code>, <code>number</code>, <code>name</code> and <code>chain</code> in PDB as strings.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/09af65a7583627f53bcd1ee131646a8ee62cc0e7/src/SIFTS/ResidueMapping.jl#L72-L74">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.SIFTS.dbPDBe" href="#MIToS.SIFTS.dbPDBe"><code>MIToS.SIFTS.dbPDBe</code></a> — <span class="docstring-category">Type</span></header><section><div><p><code>dbPDBe</code> stores the residue <code>number</code> and <code>name</code> in PDBe as strings.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/09af65a7583627f53bcd1ee131646a8ee62cc0e7/src/SIFTS/ResidueMapping.jl#L3-L5">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.SIFTS.dbPfam" href="#MIToS.SIFTS.dbPfam"><code>MIToS.SIFTS.dbPfam</code></a> — <span class="docstring-category">Type</span></header><section><div><p><code>dbPfam</code> stores the residue <code>id</code>, <code>number</code> and <code>name</code> in Pfam as strings.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/09af65a7583627f53bcd1ee131646a8ee62cc0e7/src/SIFTS/ResidueMapping.jl#L51-L53">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.SIFTS.dbSCOP" href="#MIToS.SIFTS.dbSCOP"><code>MIToS.SIFTS.dbSCOP</code></a> — <span class="docstring-category">Type</span></header><section><div><p><code>dbSCOP</code> stores the residue <code>id</code>, <code>number</code>, <code>name</code> and <code>chain</code> in SCOP as  strings.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/09af65a7583627f53bcd1ee131646a8ee62cc0e7/src/SIFTS/ResidueMapping.jl#L81-L84">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.SIFTS.dbSCOP2" href="#MIToS.SIFTS.dbSCOP2"><code>MIToS.SIFTS.dbSCOP2</code></a> — <span class="docstring-category">Type</span></header><section><div><p><code>dbSCOP2</code> stores the residue <code>id</code>, <code>number</code>, <code>name</code> and <code>chain</code> in SCOP2 as  strings.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/09af65a7583627f53bcd1ee131646a8ee62cc0e7/src/SIFTS/ResidueMapping.jl#L86-L89">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.SIFTS.dbSCOP2B" href="#MIToS.SIFTS.dbSCOP2B"><code>MIToS.SIFTS.dbSCOP2B</code></a> — <span class="docstring-category">Type</span></header><section><div><p><code>dbSCOP2B</code> stores the residue <code>id</code>, <code>number</code>, <code>name</code> and <code>chain</code> in SCOP2B as  strings. <em>SCOP2B</em> is expansion of <em>SCOP2</em> domain annotations at superfamily  level to every <em>PDB</em> with same <em>UniProt</em> accession having at least 80% <em>SCOP2</em>  domain coverage.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/09af65a7583627f53bcd1ee131646a8ee62cc0e7/src/SIFTS/ResidueMapping.jl#L91-L96">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.SIFTS.dbUniProt" href="#MIToS.SIFTS.dbUniProt"><code>MIToS.SIFTS.dbUniProt</code></a> — <span class="docstring-category">Type</span></header><section><div><p><code>dbUniProt</code> stores the residue <code>id</code>, <code>number</code> and <code>name</code> in UniProt as strings.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/09af65a7583627f53bcd1ee131646a8ee62cc0e7/src/SIFTS/ResidueMapping.jl#L47-L49">source</a></section></article><h2 id="Constants"><a class="docs-heading-anchor" href="#Constants">Constants</a><a id="Constants-1"></a><a class="docs-heading-anchor-permalink" href="#Constants" title="Permalink"></a></h2><h2 id="Macros"><a class="docs-heading-anchor" href="#Macros">Macros</a><a id="Macros-1"></a><a class="docs-heading-anchor-permalink" href="#Macros" title="Permalink"></a></h2><h2 id="Methods-and-functions"><a class="docs-heading-anchor" href="#Methods-and-functions">Methods and functions</a><a id="Methods-and-functions-1"></a><a class="docs-heading-anchor-permalink" href="#Methods-and-functions" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.parse-Tuple{LightXML.XMLDocument, Type{MIToS.SIFTS.SIFTSXML}}" href="#Base.parse-Tuple{LightXML.XMLDocument, Type{MIToS.SIFTS.SIFTSXML}}"><code>Base.parse</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>parse(document::LightXML.XMLDocument, ::Type{SIFTSXML}; chain=All, missings::Bool=true)</code></p><p>Returns a <code>Vector{SIFTSResidue}</code> parsed from a <code>SIFTSXML</code> file. By default, parses all the <code>chain</code>s and includes missing residues.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/09af65a7583627f53bcd1ee131646a8ee62cc0e7/src/SIFTS/ResidueMapping.jl#L401-L406">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.SIFTS.downloadsifts-Tuple{String}" href="#MIToS.SIFTS.downloadsifts-Tuple{String}"><code>MIToS.SIFTS.downloadsifts</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">downloadsifts(pdbcode::String; filename::String, source::String=&quot;https&quot;)</code></pre><p>Download the gzipped SIFTS XML file for the provided <code>pdbcode</code>.  The downloaded file will have the default extension <code>.xml.gz</code>.  While you can change the <code>filename</code>, it must include the <code>.xml.gz</code> ending.  The <code>source</code> keyword argument is set to <code>&quot;https&quot;</code> by default.  Alternatively, you can choose <code>&quot;ftp&quot;</code> as the <code>source</code>, which will retrieve the file from  the EBI FTP server at ftp://ftp.ebi.ac.uk/pub/databases/msd/sifts/.  However, please note that using <code>&quot;https&quot;</code> is highly recommended.  This option will download the file from the  EBI PDBe server at https://www.ebi.ac.uk/pdbe/files/sifts/.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/09af65a7583627f53bcd1ee131646a8ee62cc0e7/src/SIFTS/XMLParser.jl#L6-L18">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MIToS.SIFTS.siftsmapping-Union{Tuple{T}, Tuple{F}, Tuple{String, Type{F}, String, Type{T}, String}} where {F, T}" href="#MIToS.SIFTS.siftsmapping-Union{Tuple{T}, Tuple{F}, Tuple{String, Type{F}, String, Type{T}, String}} where {F, T}"><code>MIToS.SIFTS.siftsmapping</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Parses a SIFTS XML file and returns a <code>OrderedDict</code> between residue numbers of  two <code>DataBase</code>s with the given identifiers. A <code>chain</code> could be specified  (<code>All</code> by default). If <code>missings</code> is <code>true</code> (default) all the residues are  used, even if they haven’t coordinates in the PDB file.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/diegozea/MIToS.jl/blob/09af65a7583627f53bcd1ee131646a8ee62cc0e7/src/SIFTS/ResidueMapping.jl#L344-L349">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../Information_API/">« Information</a><a class="docs-footer-nextpage" href="../PDB_API/">PDB »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="auto">Automatic (OS)</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.2.1 on <span class="colophon-date" title="Tuesday 19 December 2023 12:27">Tuesday 19 December 2023</span>. Using Julia version 1.9.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
